{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SearchIO\n",
    "from Bio import SeqIO\n",
    "from Bio import Phylo\n",
    "import os\n",
    "import re\n",
    "import glob\n",
    "import regex\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "\n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "import numpy as np\n",
    "\n",
    "# from check_seq import check_seq\n",
    "# from check_seq import conditional_check_seq\n",
    "\n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore', category=FutureWarning)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "date = \"20260114\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse HMM output and grab hits; hits must meet the following criteria\n",
    "- positive bit score\n",
    "- full sequence evalue must be significant (<0.01)\n",
    "- best domain evalue should be significant (<0.01)\n",
    "    - otherwise flagged for manual review to check if it is distant homolog or just short repeats\n",
    "\n",
    "### Store hits in feather file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio import SearchIO\n",
    "import re\n",
    "import glob\n",
    "import sys\n",
    "\n",
    "# get folder\n",
    "dir = sys.argv[1]\n",
    "\n",
    "# Get taxonomy from GTDB, https://data.ace.uq.edu.au/public/gtdb/data/releases/latest/\n",
    "    # NCBI taxonomy from metadata files\n",
    "\n",
    "GTDB_taxonomy = pd.read_csv('GTDB_taxonomy.gz', header = None, sep = '\\t', names=['GenomeID', 'GTDB'])\n",
    "\n",
    "\n",
    "'''Parse HMM output and grab hits; hits must meet the following criteria\n",
    "- positive bit score\n",
    "- full sequence evalue must be significant (<0.01)\n",
    "- best domain evalue should be significant (<0.01)\n",
    "    - otherwise flagged for manual review to check if it is distant homolog or just short repeats\n",
    "\n",
    "Store hits in feather file'''\n",
    "\n",
    "# Create empty lists\n",
    "result_target = []\n",
    "query_id = []\n",
    "hit_id = []\n",
    "evalue = []\n",
    "best_domain_evalue = []\n",
    "bitscore = []\n",
    "bias = []\n",
    "location = []\n",
    "alength = []\n",
    "slength = []\n",
    "flag1 = []\n",
    "flag2 = []\n",
    "\n",
    "def append_hit(genomeID, gene, item):\n",
    "    result_target.append(genomeID)\n",
    "    query_id.append(gene)\n",
    "    hit_id.append(item.id)\n",
    "    evalue.append(item.evalue)\n",
    "    best_domain_evalue.append(item.hsps[0].evalue)\n",
    "    bitscore.append(item.bitscore)\n",
    "    bias.append(item.bias)\n",
    "    s = r'# ([0-9]+) # ([0-9]+)'\n",
    "    location.append(re.match(s, item.description).group(1) + \"-\" + re.match(str, item.description).group(2))\n",
    "    alength.append(item.hsps[0].aln_span)\n",
    "    slength.append(int(re.match(s, item.description).group(2))-int(re.match(str, item.description).group(1)))\n",
    "\n",
    "# Parse through files in output directory\n",
    "\n",
    "for file in glob.glob(f'../results/{dir}/hmmsearch_results/*.out'):\n",
    "\n",
    "    # RegEx for the GenomeID (double checking that file is really a genome)\n",
    "    try:\n",
    "        s = r'([\\w]+_[\\w]+_[\\d]+\\.[\\d])'\n",
    "        genomeID = re.search(s, file).group()\n",
    "    except:\n",
    "        continue\n",
    "\n",
    "    # Parse file using SearchIO/HmmerIO\n",
    "    for result in SearchIO.parse(file, 'hmmer3-text'):\n",
    "        for item in result.hits:\n",
    "\n",
    "            # grab gene name\n",
    "            s = r'([a-zA-Z]+)' # ex. nifHDK, pchlide\n",
    "            gene = re.findall(s, result.id)[0]\n",
    "\n",
    "            # Check for positive bitscore and append the data to the corresponding lists\n",
    "            if item.bitscore > 0 and item.evalue < 0.01:\n",
    "                # append hits\n",
    "                append_hit(genomeID, gene, item)\n",
    "\n",
    "                # check if full seq and best domain e-val are significant\n",
    "                if item.hsps[0].evalue < 0.01:\n",
    "                    flag1.append(0)\n",
    "                else:\n",
    "                    # check if \"full sequence Eval is sig but best domain is not, keep only if the target sequence \"a multidomain remote homolog; but be wary, and watch out for the case where itâ€™s just a repetitive sequence\"\n",
    "                    flag1.append(1)\n",
    "\n",
    "                # check if bitscore >> bias (same order of magnitude) as bitscore\n",
    "                if item.bias != 0 and item.bitscore/item.bias > 10:\n",
    "                    flag2.append(0)\n",
    "                else:\n",
    "                    flag2.append(1)\n",
    "                \n",
    "# create and store dataframe\n",
    "hits = pd.DataFrame({'GenomeID': result_target, 'Gene': query_id, 'Hit': hit_id, \n",
    "                     'E-value': evalue, 'Best Domain E-value': best_domain_evalue, 'Bit Score': bitscore, 'Bias': bias,\n",
    "                     'Location': location, 'Alignment Length': alength, 'Sequence Length': slength, \n",
    "                     'Flag_Eval': flag1, 'Flag_Bias': flag2})\n",
    "\n",
    "# add taxonomy info\n",
    "hits = pd.merge(hits, GTDB_taxonomy, on = \"GenomeID\", how = \"left\")\n",
    "\n",
    "# save as feather file\n",
    "hits.to_feather(f'../results/{dir}/hits.feather')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Parse hits to find best gene match\n",
    "- Since nifDK are distantly related, they almost always both appear as hits for each other (how can I word this better)\n",
    "- To properly assign the correct gene, select the hit with the highest bitscore/lowest evalue\n",
    "- then consider operon organization\n",
    "    - if more than 10 genes away (need a reference pt for this number, look for other hits)\n",
    "- Keep genes from genomes that contain all nifHDK\n",
    "- Store other less promising hits, for later validation\n",
    "\n",
    "### Output top hits as feather file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "warnings.filterwarnings('ignore', category=FutureWarning)\n",
    "\n",
    "from itertools import combinations\n",
    "from cluster_pos import cluster_pos\n",
    "\n",
    "# get parsed hmm results\n",
    "hits = pd.read_feather('archaea/hits.feather')\n",
    "\n",
    "# save \"contig\" as col\n",
    "hits['contig'] = hits['Hit'].str.split('_').str[:-1].str.join('_') \n",
    "\n",
    "# multi-index to cluster by genome, contig\n",
    "hits.set_index(['GenomeID', 'contig'], inplace = True)\n",
    "hits.sort_index(inplace = True)\n",
    "hits.drop_duplicates(inplace = True)\n",
    "\n",
    "# filter for genome, contig with at least 3 unique genes (nifHDKENB)\n",
    "filtered_df = hits.groupby(level=['GenomeID', 'contig']).filter(lambda x: x['Gene'].nunique() >= 3)\n",
    "\n",
    "# # make sure these 3 unique genes are not the same hit (i.e. not the same gene in reference genome)\n",
    "filtered_df2 = filtered_df.groupby(level=['GenomeID', 'contig']).filter(lambda x: x['Hit'].nunique() >= 3)\n",
    "\n",
    "genomes_to_keep = pd.DataFrame(columns = filtered_df2.columns)\n",
    "genomes_to_keep['gene_cluster'] = 0\n",
    "\n",
    "# iterate through each genome and contig\n",
    "for genome in filtered_df2.index.get_level_values(0).unique(): # iterate through each genome\n",
    "    for contig in filtered_df2.loc[genome].index.get_level_values(0).unique(): # iterate through each contig\n",
    "\n",
    "        tmp = filtered_df2.loc[(genome, contig)]\n",
    "\n",
    "        # only keep numbers that have clusters >= 3\n",
    "        pos_clusters = cluster_pos(tmp.Hit.unique())\n",
    "\n",
    "        # for each cluster\n",
    "        for ind, cl in enumerate(pos_clusters):\n",
    "            pos = [contig + '_' + str(p) for p in cl]\n",
    "            no_pos = len(pos)\n",
    "            \n",
    "            # need at least 3 genes to continue\n",
    "            if no_pos < 3:\n",
    "                continue\n",
    "\n",
    "            # only keep hits that are in the cluster\n",
    "            tmp2 = tmp[tmp.Hit.isin(pos)]\n",
    "\n",
    "            # only keep hits that are in the cluster\n",
    "            tmp2 = tmp[tmp.Hit.isin(pos)].reset_index()\n",
    "\n",
    "            tmp2['gene_cluster'] = ind + 1\n",
    "\n",
    "            # save index\n",
    "            genomes_to_keep = pd.concat([genomes_to_keep, tmp2])\n",
    "\n",
    "# filter for genomes to keep\n",
    "genomes_to_keep.to_feather('archaea/nif.feather')\n",
    "genomes_to_keep.to_csv('archaea/nif.csv')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "how many genomes only have nifDK?\n",
    "How many genomes/contigs have multiple nifHDK motifs?\n",
    "How many genes have both nifDK together --> i.e. no stop codon --> do they have two separate high scoring domains?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Validate hits as correct nifHDK matches by checking for the presence of CDD's conserved features\n",
    "- For nifD check for P-cluster and FeMoCo binding residues (https://www.ncbi.nlm.nih.gov/Structure/cdd/cddsrv.cgi?uid=cd01976)\n",
    "- For nifK check for P-cluster binding residues (https://www.ncbi.nlm.nih.gov/Structure/cdd/cddsrv.cgi?uid=cd01974)\n",
    "- For nifH check for active site, Fe-S cluster (https://www.ncbi.nlm.nih.gov/Structure/cdd/cddsrv.cgi?uid=349759)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "reload_fasta = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nifH\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "nifD\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "nifK\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "nifB\n",
      "1\n",
      "2\n",
      "3\n",
      "nifE\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "nifN\n",
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n"
     ]
    }
   ],
   "source": [
    "# align hits\n",
    "\n",
    "from Bio import AlignIO\n",
    "\n",
    "# grab both archaea + bacteria hits\n",
    "nif_archaea = pd.read_feather('results/archaea/nif.feather')\n",
    "nif = nif_archaea.copy()\n",
    "#nif_bacteria = pd.read_feather('results/bacteria/nif.feather')\n",
    "#nif = pd.concat([nif_archaea, nif_bacteria])\n",
    "\n",
    "nif.reset_index(inplace = True)\n",
    "nif.set_index(['GenomeID'], inplace = True)\n",
    "nif['Seq'] = ''\n",
    "\n",
    "# separate by annotation\n",
    "nifH = nif[nif.Gene == 'nifH']\n",
    "nifD = nif[nif.Gene == 'nifD']\n",
    "nifK = nif[nif.Gene == 'nifK']\n",
    "nifB = nif[nif.Gene == 'nifB']\n",
    "nifE = nif[nif.Gene == 'nifE']\n",
    "nifN = nif[nif.Gene == 'nifN']\n",
    "\n",
    "gene_list = [nifH, nifD, nifK, nifB, nifE, nifN]\n",
    "gene_names = ['nifH', 'nifD', 'nifK', 'nifB', 'nifE', 'nifN']\n",
    "\n",
    "if reload_fasta: # should skip if fasta sequences have already been extracted\n",
    "    # get fasta sequences for each gene & export to fasta\n",
    "    for gene, name in zip(gene_list, gene_names):\n",
    "        print(name)\n",
    "        records = []\n",
    "        for genome,hit in gene.iterrows():\n",
    "            hit = hit.Hit\n",
    "\n",
    "            file = glob.glob(f\"all_rep_proteins_aa/*/{genome}_protein.faa\")[0]\n",
    "            \n",
    "            for result in SeqIO.parse(file, \"fasta\"):\n",
    "                if result.id == hit:\n",
    "                    # store seq\n",
    "                    gene.loc[genome, 'Seq'] = str(result.seq)\n",
    "                    # convert to seqrecord\n",
    "                    record = SeqRecord(Seq(result.seq), id=genome, description=hit)\n",
    "                    records.append(record)\n",
    "                    # exit loop once sequence is found\n",
    "                    break\n",
    "            \n",
    "        # Write the records to a FASTA file\n",
    "        with open(\"results/\" + name + f\"_{date}.fasta\", \"w\") as output_handle:\n",
    "            SeqIO.write(records, output_handle, \"fasta\")`\n",
    "\n",
    "# align fasta files\n",
    "for gene in gene_names:\n",
    "    print(\"aligning \"+ gene, flush=True)\n",
    "    num = eval(f\"int({gene}.shape[0]/200)+1\") # how many splits\n",
    "    os.system(f\"seqtk split -n {num} results/fasta_splits/{gene}_split results/{gene}_{date}.fasta\") # split fasta file\n",
    "    for i in range(num):\n",
    "        print(i+1)\n",
    "        os.system(f\"seqtk subseq results/{gene}_{date}.fasta results/ref_seq.ids >> results/fasta_splits/{gene}_split.{str(i+1).zfill(5)}.fa\") # add reference sequences\n",
    "        os.system(f\"mafft --auto --quiet --thread 4 results/fasta_splits/{gene}_split.{str(i+1).zfill(5)}.fa > results/fasta_splits/{gene}_split.{str(i+1).zfill(5)}.aln\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# nif_archaea = pd.read_feather('archaea/nif.feather')\n",
    "# # nif = nif_archaea.copy()\n",
    "# nif_bacteria = pd.read_feather('bacteria/nif.feather')\n",
    "# nif = pd.concat([nif_archaea, nif_bacteria])\n",
    "\n",
    "nif.reset_index(inplace = True)\n",
    "nif.set_index(['GenomeID', 'Hit','Gene'], inplace = True)\n",
    "nif['Seq'] = ''\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check for conserved residues\n",
    "from Bio import AlignIO\n",
    "def check_gene(gene, ref_seq, important_residues, passing_score, p=False):\n",
    "    \n",
    "    alignment = AlignIO.read(f\"{gene}.aln\", \"fasta\")\n",
    "\n",
    "    for record in alignment:\n",
    "        print(record.description)\n",
    "        if ref_seq in record.description:\n",
    "            print()\n",
    "            aln = record.seq\n",
    "            break\n",
    "    \n",
    "    # map important cols in aln to ref_seq\n",
    "    residue_to_alignment = {}\n",
    "    residue_idx = 0  # index in original (ungapped) sequence\n",
    "    col2res = {}\n",
    "\n",
    "    for aln_idx, char in enumerate(aln):\n",
    "        if char != '-':\n",
    "            residue_idx += 1\n",
    "            if residue_idx in important_residues:\n",
    "                residue_to_alignment[residue_idx] = aln_idx\n",
    "                col2res[aln_idx] = aln[aln_idx]\n",
    "                \n",
    "            # early stop if we've found everything\n",
    "            if len(residue_to_alignment) == len(important_residues):\n",
    "                break\n",
    "\n",
    "    # print the corresponding residues in the original sequence\n",
    "    if p:\n",
    "        for residue, aln_idx in residue_to_alignment.items():\n",
    "            print(f\"Residue {residue} corresponds to alignment index {aln_idx}: {aln[aln_idx]}\", flush = True)\n",
    "\n",
    "    # store in dataframe        \n",
    "    acc = [result.id for result in alignment]\n",
    "    seqs = [list(str(result.seq)) for result in alignment]\n",
    "    hits = [result.description.split(' ')[-1] for result in alignment]\n",
    "\n",
    "    pssm = pd.DataFrame(seqs, index = acc)\n",
    "\n",
    "    pssm = pssm.iloc[:, list(residue_to_alignment.values())]\n",
    "    pssm['hit'] = hits\n",
    "    pssm['contig'] = pssm['hit'].str.split('_').str[:-1].str.join('_')\n",
    "    \n",
    "    # check if cols contain correct residue for function\n",
    "    def check_res(row):\n",
    "        score = 0\n",
    "        for col in residue_to_alignment.values():\n",
    "            if row[col] == aln[col]:\n",
    "                score += 2\n",
    "                if aln[col] == 'C': # higher weight for correct C\n",
    "                    score += 1\n",
    "            elif aln[col] == 'C': # if ref seq is C, must also be C\n",
    "                continue\n",
    "            elif row[col] != '-': # greater penalty for gap than incorrect residue\n",
    "                score += 1\n",
    "        if score >= passing_score:\n",
    "            return score\n",
    "        else:\n",
    "            return np.nan\n",
    "        #return score\n",
    "\n",
    "    pssm['score'] = pssm.apply(check_res, axis = 1)\n",
    "    pssm.dropna(subset = ['score'], inplace = True)\n",
    "    # rename cols\n",
    "    pssm.rename(columns = col2res, inplace = True)\n",
    "\n",
    "    return pssm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GB_GCA_000306725.1 CP003083.1_1272\n",
      "GB_GCA_000499765.1 HG425166.1_1663\n",
      "GB_GCA_001509375.1 LGFT01000081.1_1\n",
      "GB_GCA_001602645.1 LUYE01000021.1_25\n",
      "GB_GCA_001723855.1 MCBC01000014.1_13\n",
      "GB_GCA_002011165.2 MTMI02000009.1_227\n",
      "GB_GCA_002067095.1 MVQD01000106.1_245\n",
      "GB_GCA_002067705.1 MVRL01000088.1_88\n",
      "GB_GCA_002068435.1 MVZQ01000082.1_8\n",
      "GB_GCA_002495065.1 DADT01000005.1_34\n",
      "GB_GCA_002495625.1 DAGC01000013.1_55\n",
      "GB_GCA_002497075.1 DAML01000010.1_53\n",
      "GB_GCA_002499455.1 DALN01000141.1_15\n",
      "GB_GCA_002503135.1 DARA01000054.1_10\n",
      "GB_GCA_002504405.1 DARO01000004.1_9\n",
      "GB_GCA_002506015.1 DAWR01000075.1_9\n",
      "GB_GCA_002507205.1 DAZF01000050.1_6\n",
      "GB_GCA_002508465.1 DAQQ01000257.1_20\n",
      "GB_GCA_002509495.1 DAEV01000020.1_8\n",
      "GB_GCA_002839605.1 PGYK01000018.1_18\n",
      "GB_GCA_002926195.1 MZXQ01000108.1_2\n",
      "GB_GCA_003151535.1 PLXT01000078.1_8\n",
      "GB_GCA_003158115.1 PMGM01000048.1_30\n",
      "GB_GCA_003164755.1 PLUB01000010.1_109\n",
      "GB_GCA_003194445.1 PQXF01000004.1_158\n",
      "GB_GCA_003336485.1 QENH01000192.1_26\n",
      "GB_GCA_003601795.1 PIXU01000248.1_4\n",
      "GB_GCA_003661125.1 QMVO01000304.1_1\n",
      "GB_GCA_004193545.1 RPGO01000033.1_46\n",
      "GB_GCA_004332335.1 SISS01000008.1_42\n",
      "GB_GCA_009712615.1 WASR01000091.1_1\n",
      "GB_GCA_009780575.1 WQZU01000009.1_34\n",
      "GB_GCA_011620785.1 WOYM01000080.1_17\n",
      "GB_GCA_012329085.1 JAAXQB010000082.1_3\n",
      "GB_GCA_012520845.1 JAAYNL010000015.1_85\n",
      "GB_GCA_012728895.1 JAAYEG010000080.1_49\n",
      "GB_GCA_013180605.1 QBUL01000078.1_3\n",
      "GB_GCA_013329575.1 DUJX01000033.1_11\n",
      "GB_GCA_013403005.1 JACBOA010000001.1_34\n",
      "GB_GCA_013572355.1 QBUV01000135.1_7\n",
      "GB_GCA_014361295.1 JACIVX010000001.1_209\n",
      "GB_GCA_015063085.1 SUTD01000003.1_78\n",
      "GB_GCA_015660885.1 JACTUB010000001.1_326\n",
      "GB_GCA_015661405.1 JACWMK010000035.1_23\n",
      "GB_GCA_016699325.1 CP064971.1_295\n",
      "GB_GCA_016841955.1 JAENYE010000002.1_372\n",
      "GB_GCA_016928355.1 JAFGKW010000034.1_10\n",
      "GB_GCA_017399555.1 JAFQBW010000002.1_46\n",
      "GB_GCA_017468935.1 JAFUHJ010000137.1_8\n",
      "GB_GCA_017608725.1 JAFCFG010000001.1_10\n",
      "GB_GCA_017883485.1 JADGNM010000057.1_1\n",
      "GB_GCA_018262855.1 JADFDO010000011.1_52\n",
      "GB_GCA_018687785.1 LTBL01000001.1_1391\n",
      "GB_GCA_019262145.1 CP076745.1_439\n",
      "GB_GCA_020854755.1 JAHKLN010000051.1_25\n",
      "GB_GCA_021409285.1 JAGFND010000012.1_6\n",
      "GB_GCA_023228295.1 JALOBP010000043.1_12\n",
      "GB_GCA_023396065.1 JAEWXF010000005.1_26\n",
      "GB_GCA_023409695.1 JAEYQZ010000253.1_11\n",
      "GB_GCA_023436175.1 JAEZKE010000018.1_31\n",
      "GB_GCA_023955635.1 JAMXHR010000001.1_451\n",
      "GB_GCA_024699715.1 JANPRD010000007.1_84\n",
      "GB_GCA_025360805.1 JANYBK010000146.1_3\n",
      "GB_GCA_025464325.1 JAODGH010000033.1_12\n",
      "GB_GCA_025464645.1 JAODGF010000166.1_2\n",
      "GB_GCA_026014465.1 JAOZID010000010.1_980\n",
      "GB_GCA_026016205.1 CP061209.1_926\n",
      "GB_GCA_026394605.1 JAPKXY010000040.1_19\n",
      "GB_GCA_026552855.1 JAPMTX010000018.1_2\n",
      "GB_GCA_027028145.1 JALVJV010000011.1_97\n",
      "GB_GCA_027064745.1 JAMOPD010000204.1_1\n",
      "GB_GCA_027460525.1 JAPZAS010000036.1_122\n",
      "GB_GCA_027460605.1 JAPZAN010000216.1_15\n",
      "GB_GCA_028678635.1 JAJRDC010000201.1_13\n",
      "GB_GCA_028678965.1 JAJRCM010000107.1_2\n",
      "GB_GCA_028679185.1 JAJRBZ010000012.1_5\n",
      "GB_GCA_028679325.1 JAJRBS010000032.1_3\n",
      "GB_GCA_028682845.1 JAJRAT010000124.1_2\n",
      "GB_GCA_028688535.1 JAQTXY010000449.1_2\n",
      "GB_GCA_028708055.1 JAQVZT010000001.1_38\n",
      "GB_GCA_028717345.1 JAQUNZ010000003.1_71\n",
      "GB_GCA_029556335.1 JAKQFA010000028.1_9\n",
      "GB_GCA_029762625.1 JARVGI010000005.1_39\n",
      "GB_GCA_029887785.1 JARXPG010000001.1_1372\n",
      "GB_GCA_029907705.1 JARYLJ010000007.1_38\n",
      "GB_GCA_029987115.1 DALTWK010000016.1_13\n",
      "GB_GCA_030016715.1 JASEJD010000010.1_14\n",
      "GB_GCA_030019395.1 JASEES010000019.1_20\n",
      "GB_GCA_030156225.1 JASGMW010000001.1_41\n",
      "GB_GCA_030539595.1 JAUNXW010000013.1_10\n",
      "GB_GCA_030604085.1 JAUVWP010000061.1_4\n",
      "GB_GCA_030605385.1 JAUWIL010000101.1_4\n",
      "GB_GCA_030610815.1 JAUWRA010000008.1_25\n",
      "GB_GCA_030611265.1 JAUWQD010000002.1_19\n",
      "GB_GCA_030651815.1 JAUSBW010000216.1_26\n",
      "GB_GCA_030694985.1 JAUYTV010000004.1_13\n",
      "GB_GCA_031256615.1 JAIRQG010000076.1_1\n",
      "GB_GCA_031261915.1 JAISIT010000144.1_18\n",
      "GB_GCA_031271555.1 JAISVN010000013.1_4\n",
      "GB_GCA_031282205.1 JAITEE010000061.1_28\n",
      "GB_GCA_031288985.1 JAITQH010000039.1_2\n",
      "GB_GCA_031290095.1 JAITOF010000046.1_3\n",
      "GB_GCA_031312595.1 CP133592.1_1757\n",
      "GB_GCA_902158745.1 CABGHH010000072.1_4\n",
      "GB_GCA_902386255.1 CABMII010000067.1_3\n",
      "GB_GCA_903871465.1 CAIQHB010000022.1_9\n",
      "GB_GCA_905235495.1 CAJOMK010000074.1_19\n",
      "GB_GCA_937906725.1 CALCXE010000066.1_32\n",
      "RS_GCF_000007305.1 NC_003413.1_2063\n",
      "RS_GCF_000007345.1 NC_003552.1_4467\n",
      "RS_GCF_000008645.1 NC_000916.1_1422\n",
      "RS_GCF_000009965.1 NC_006624.1_2014\n",
      "RS_GCF_000013725.1 NC_007955.1_665\n",
      "RS_GCF_000015805.1 NC_009073.1_912\n",
      "RS_GCF_000017185.1 NC_009635.1_1450\n",
      "RS_GCF_000018365.1 NC_011529.1_926\n",
      "RS_GCF_000022365.1 NC_012804.1_232\n",
      "RS_GCF_000025285.1 NC_013741.1_1047\n",
      "RS_GCF_000063445.1 NC_009464.1_1403\n",
      "RS_GCF_000147875.1 NC_014507.1_2334\n",
      "RS_GCF_000191585.1 NC_015216.1_1812\n",
      "RS_GCF_000196655.1 NC_014253.1_2116\n",
      "RS_GCF_000214415.1 NC_015562.1_1756\n",
      "RS_GCF_000243255.1 NZ_CM001436.1_835\n",
      "RS_GCF_000251105.1 NC_017034.1_912\n",
      "RS_GCF_000275865.1 NZ_CM001555.1_1253\n",
      "RS_GCF_000308215.1 NZ_CAJE01000013.1_252\n",
      "RS_GCF_000328665.1 NC_019977.1_2060\n",
      "RS_GCF_000517445.1 NZ_CP006965.1_1132\n",
      "RS_GCF_000685155.1 NZ_JMIY01000001.1_759\n",
      "RS_GCF_000765475.1 NZ_JRHO01000014.1_196\n",
      "RS_GCF_000828575.1 NZ_AP011952.1_1083\n",
      "RS_GCF_000969885.1 NZ_CP009501.1_1188\n",
      "RS_GCF_000969905.1 NZ_CP009520.1_2025\n",
      "RS_GCF_000969985.1 NZ_CP009526.1_860\n",
      "RS_GCF_000970025.1 NZ_CP009528.1_1458\n",
      "RS_GCF_000970085.1 NZ_CP009506.1_3668\n",
      "RS_GCF_000970205.1 NZ_CP009512.1_329\n",
      "RS_GCF_000970285.1 NZ_CP009516.1_737\n",
      "RS_GCF_000970305.1 NZ_CP009517.1_2001\n",
      "RS_GCF_000970325.1 NZ_CP009518.1_1606\n",
      "RS_GCF_001316325.1 NZ_BBES01000002.1_184\n",
      "RS_GCF_001571405.1 NZ_BCNX01000007.1_244\n",
      "RS_GCF_001639295.1 NZ_LWMV01000018.1_7\n",
      "RS_GCF_002067275.1 NZ_MVQX01000060.1_21\n",
      "RS_GCF_002153915.1 NZ_MRZU01000002.1_162\n",
      "RS_GCF_002214485.1 NZ_CP015102.1_1672\n",
      "RS_GCF_002214525.1 NZ_CP015104.1_1428\n",
      "RS_GCF_002214585.1 NZ_CP014862.1_1849\n",
      "RS_GCF_002287175.1 NZ_LMVM01000041.1_88\n",
      "RS_GCF_002496565.1 NZ_DAKJ01000018.1_492\n",
      "RS_GCF_002502735.1 NZ_DAQC01000001.1_307\n",
      "RS_GCF_002506585.1 NZ_DAXW01000016.1_237\n",
      "RS_GCF_002813695.1 NZ_CP017768.1_1500\n",
      "RS_GCF_003170075.1 NZ_PMZU01000058.1_30\n",
      "RS_GCF_003351865.1 NZ_CP022242.1_323\n",
      "RS_GCF_004099695.1 NZ_SCHL01000008.1_93\n",
      "RS_GCF_004745425.1 NZ_PGGK01000004.1_118\n",
      "RS_GCF_006546655.1 NZ_VIAQ01000019.1_163\n",
      "RS_GCF_012027355.1 NZ_SNUH01000001.1_237\n",
      "RS_GCF_012027635.1 NZ_SNUS01000001.1_85\n",
      "RS_GCF_013377755.1 NZ_JABXWR010000001.1_700\n",
      "RS_GCF_013760955.1 NZ_JACDUL010000003.1_20\n",
      "RS_GCF_016217785.1 NZ_JACRGA010000005.1_372\n",
      "RS_GCF_016278385.1 NZ_JADQBY010000006.1_79\n",
      "RS_GCF_017873415.1 NZ_JAGGKD010000007.1_28\n",
      "RS_GCF_017874455.1 NZ_JAGGLN010000004.1_96\n",
      "RS_GCF_017875325.1 NZ_JAGGMN010000001.1_1167\n",
      "RS_GCF_018141185.1 NZ_CP058560.1_267\n",
      "RS_GCF_019263745.1 NZ_CP077107.1_396\n",
      "RS_GCF_019931065.1 NZ_JAIOUQ010000003.1_642\n",
      "RS_GCF_020804225.1 NZ_WJBI01000001.1_288\n",
      "RS_GCF_021023085.1 NZ_PGCK01000003.1_112\n",
      "RS_GCF_023016325.1 NZ_CP095400.1_985\n",
      "RS_GCF_023227715.1 NZ_JALKIF010000002.1_91\n",
      "RS_GCF_024022995.1 NZ_CP099582.1_1872\n",
      "RS_GCF_024372725.1 NZ_VOUA01000001.1_67\n",
      "RS_GCF_024662215.1 NZ_CP096115.1_1100\n",
      "RS_GCF_025384765.1 NZ_VHLL01000002.1_251\n",
      "RS_GCF_026684035.1 NZ_CP113361.1_823\n",
      "RS_GCF_027008145.1 NZ_JALTCZ010000060.1_2\n",
      "RS_GCF_027012695.1 NZ_JALUYR010000143.1_25\n",
      "RS_GCF_028679455.1 NZ_JAJRBM010000288.1_3\n",
      "RS_GCF_028712425.1 NZ_JAQTQD010000014.1_52\n",
      "RS_GCF_029633915.1 NZ_CP091092.1_452\n",
      "RS_GCF_029854155.1 NZ_CP122958.1_58\n",
      "RS_GCF_030296715.1 NZ_AP027738.1_300\n",
      "RS_GCF_030323845.1 NZ_CP040735.1_1280\n",
      "RS_GCF_030655665.1 NZ_JAUSPQ010000004.1_30\n",
      "RS_GCF_900100715.1 NZ_FNCA01000003.1_233\n",
      "RS_GCF_900114835.1 NZ_FOUJ01000005.1_126\n",
      "RS_GCF_902784195.1 NZ_CACZTX010000002.1_66\n",
      "RS_GCF_946463545.1 NZ_OX296583.1_1636\n"
     ]
    },
    {
     "ename": "UnboundLocalError",
     "evalue": "cannot access local variable 'aln' where it is not associated with a value",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mUnboundLocalError\u001b[39m                         Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[12]\u001b[39m\u001b[32m, line 13\u001b[39m\n\u001b[32m     10\u001b[39m passing_score = \u001b[32m25\u001b[39m \u001b[38;5;66;03m# try 28 (complete match?)\u001b[39;00m\n\u001b[32m     12\u001b[39m \u001b[38;5;66;03m# initialize dataframe\u001b[39;00m\n\u001b[32m---> \u001b[39m\u001b[32m13\u001b[39m nifH_checked = \u001b[43mcheck_gene\u001b[49m\u001b[43m(\u001b[49m\u001b[33;43m'\u001b[39;49m\u001b[33;43mresults/fasta_splits/nifH_split.00001\u001b[39;49m\u001b[33;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mref_seq_nifH\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mimportant_residues_nifH\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpassing_score\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mp\u001b[49m\u001b[43m=\u001b[49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[32m     15\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m file \u001b[38;5;129;01min\u001b[39;00m glob.glob(\u001b[33mf\u001b[39m\u001b[33m'\u001b[39m\u001b[33mresults/fasta_splits/nifH_split.000*.aln\u001b[39m\u001b[33m'\u001b[39m):\n\u001b[32m     16\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m file == \u001b[33m'\u001b[39m\u001b[33mresults/fasta_splits/nifH_split.00001.aln\u001b[39m\u001b[33m'\u001b[39m:\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[11]\u001b[39m\u001b[32m, line 19\u001b[39m, in \u001b[36mcheck_gene\u001b[39m\u001b[34m(gene, ref_seq, important_residues, passing_score, p)\u001b[39m\n\u001b[32m     16\u001b[39m residue_idx = \u001b[32m0\u001b[39m  \u001b[38;5;66;03m# index in original (ungapped) sequence\u001b[39;00m\n\u001b[32m     17\u001b[39m col2res = {}\n\u001b[32m---> \u001b[39m\u001b[32m19\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m aln_idx, char \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(\u001b[43maln\u001b[49m):\n\u001b[32m     20\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m char != \u001b[33m'\u001b[39m\u001b[33m-\u001b[39m\u001b[33m'\u001b[39m:\n\u001b[32m     21\u001b[39m         residue_idx += \u001b[32m1\u001b[39m\n",
      "\u001b[31mUnboundLocalError\u001b[39m: cannot access local variable 'aln' where it is not associated with a value"
     ]
    }
   ],
   "source": [
    "# nifH\n",
    "\n",
    "important_residues_nifH = [8, 9, 10, 11, 13, 14, 15, 38, 40, 42, 95, 127, 130] # methanosarcina\n",
    "#important_residues_nifH = [10, 11, 12, 13, 15, 16, 17, 40, 42, 44, 98, 130, 133] # azotobacter\n",
    "#GKGGGKS DKD CDC --> also returns BchL, ChlL, and BchX\n",
    "\n",
    "ref_seq = 'NC_003552.1_4100'\n",
    "ref_seq_nifH = 'NZ_BSFG01000004.1_46'\n",
    "\n",
    "passing_score = 25 # try 28 (complete match?)\n",
    "\n",
    "# initialize dataframe\n",
    "nifH_checked = check_gene('results/fasta_splits/nifH_split.00001', ref_seq_nifH, important_residues_nifH, passing_score, p=True)\n",
    "\n",
    "for file in glob.glob(f'results/fasta_splits/nifH_split.000*.aln'):\n",
    "    if file == 'results/fasta_splits/nifH_split.00001.aln':\n",
    "        continue\n",
    "    new = check_gene(file[:-4], ref_seq_nifH, important_residues_nifH, passing_score)\n",
    "    nifH_checked = pd.concat([nifH_checked, new])\n",
    "\n",
    "nifH_checked.drop_duplicates(subset = ['hit'], inplace = True)\n",
    "nifH_checked.set_index(['hit'], append= True, inplace = True)\n",
    "\n",
    "print(str(len(nifH_checked)) + \" nifH seqs\")\n",
    "\n",
    "# potentially also YAANN 159-162, if 162 = R than def Bch\n",
    "# might be easier to filter at nifEN step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Residue 62 corresponds to alignment index 93: C\n",
      "Residue 88 corresponds to alignment index 119: C\n",
      "Residue 154 corresponds to alignment index 194: C\n",
      "Residue 191 corresponds to alignment index 239: Q\n",
      "Residue 195 corresponds to alignment index 243: H\n",
      "Residue 275 corresponds to alignment index 356: C\n",
      "Residue 442 corresponds to alignment index 566: H\n",
      "14 nifD seqs\n"
     ]
    }
   ],
   "source": [
    "# nifD\n",
    "\n",
    "# grab aln numbering to map to ref seq (azoto, methano)\n",
    "#important_residues_nifD = [58, 84, 149, 186, 190, 267, 485] # methanosarcina\n",
    "important_residues_nifD = [61, 62, 77, 83, 84, 85, 87, 88, 154, 195, 275, 349, 442] \n",
    "# GCDHGPGCCHCKH\n",
    "# GC{A|S}DHGP{I|L|V}GCCHCKH\n",
    "# CCCQHCH\n",
    "# azotobacter --> removed Q/Gln at 191 b/c not group IV\n",
    "# Group IVA do not have H at 442, instead have GDE (LLIGDE) motif (July 23)\n",
    "    # but have H at 426\n",
    "    # P-cluster, nifK has 2nd Cys --> S/A GPGC at ~36\n",
    "# MarDK (GroupIVC) also do not have H at 442, also have GDE (PLGDE) motif but conserved region upstream which is different than Group IVA\n",
    "    # but have H at 429 that ligates FeMoco\n",
    "    # P-cluster, marK has 2nd Cys --> CGPGC at ~36\n",
    "# if MarDk does not have homocitrate, does GroupIVA --> what residues help coordinate homocitrate?\n",
    "# appears that some groups have H442R,K\n",
    "\n",
    "#ref_seq = 'NC_003552.1_4103'\n",
    "ref_seq_nifD = 'NZ_BSFG01000004.1_45' # azo\n",
    "\n",
    "passing_score = 15\n",
    "\n",
    "# initialize dataframe\n",
    "nifD_checked = check_gene('fasta_splits/nifD_split.00001', ref_seq_nifD, important_residues_nifD, passing_score, p=True)\n",
    "\n",
    "for file in glob.glob(f'fasta_splits/nifD_split.00*.aln'):\n",
    "    if file == 'fasta_splits/nifD_split.00001.aln':\n",
    "        continue\n",
    "    new = check_gene(file[:-4], ref_seq_nifD, important_residues_nifD, passing_score)\n",
    "    nifD_checked = pd.concat([nifD_checked, new])\n",
    "\n",
    "# for each genome, only keep the best hit per gene cluster\n",
    "nifD_checked.drop_duplicates(subset = ['hit'], inplace = True)\n",
    "nifD_checked.set_index(['hit'], append= True, inplace = True)\n",
    "nifD_checked['gene_cluster'] = nif.loc[(nifD_checked.index.get_level_values(0), nifD_checked.index.get_level_values(1), 'nifD'), \n",
    "                                       'gene_cluster'].values\n",
    "nifD_checked = nifD_checked.loc[nifD_checked.groupby(['contig', 'gene_cluster'])['score'].idxmax()]\n",
    "#nifD_checked.drop_duplicates(inplace = True) # --> was previously dropping hits across same contig even if not same gene\n",
    "\n",
    "print(str(len(nifD_checked)) + \" nifD seqs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Residue 25 corresponds to alignment index 64: C\n",
      "Residue 37 corresponds to alignment index 97: C\n",
      "Residue 62 corresponds to alignment index 123: C\n",
      "Residue 124 corresponds to alignment index 202: C\n",
      "12 nifE seqs\n"
     ]
    }
   ],
   "source": [
    "# nifE check\n",
    "\n",
    "important_residues_nifE = [25, 37, 62, 83, 84, 85, 124] # azo, known catalytic\n",
    "#important_residues_nifE = [25, 37, 62, 124, 46, 51, 155, 163, 324, 395]\n",
    "# important_residues = [25, 37, 62, 124] # azotobacter YF39, G31, G41, L46, D51, E90, G96, L102, P115, I129, D132, P147, G155, G163, G202, G221, P268, G276, G324, G347, D395, G427\n",
    "# CCCHGPC\n",
    "\n",
    "# ref_seq = 'NC_003552.1_4105' # methano\n",
    "ref_seq_nifE = 'NZ_BSFG01000004.1_39' # azo\n",
    "\n",
    "passing_score = 12\n",
    "\n",
    "nifE_checked = check_gene('fasta_splits/nifE_split.00001', ref_seq_nifE, important_residues_nifE, passing_score, p=True)\n",
    "\n",
    "for file in glob.glob(f'fasta_splits/nifE_split.000*.aln'):\n",
    "    if file == 'fasta_splits/nifE_split.00001.aln':\n",
    "        continue\n",
    "    new = check_gene(file[:-4], ref_seq_nifE, important_residues_nifE, passing_score)\n",
    "    nifE_checked = pd.concat([nifE_checked, new])\n",
    "\n",
    "nifE_checked.drop_duplicates(subset = ['hit'], inplace = True)\n",
    "nifE_checked.set_index(['hit'], append= True, inplace = True)\n",
    "\n",
    "#remove nifD hits from nifE\n",
    "nifE_checked = nifE_checked[~nifE_checked.index.get_level_values(1).isin(nifD_checked.index.get_level_values(1).to_list())]\n",
    "\n",
    "#for each genome, only keep the best hit per gene cluster\n",
    "nifE_checked['gene_cluster'] = nif.loc[(nifE_checked.index.get_level_values(0), nifE_checked.index.get_level_values(1), 'nifE'), \n",
    "                                       'gene_cluster'].values\n",
    "nifE_checked = nifE_checked.loc[nifE_checked.groupby(['contig', 'gene_cluster'])['score'].idxmax()]\n",
    "\n",
    "print(str(len(nifE_checked)) + \" nifE seqs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Residue 70 corresponds to alignment index 94: C\n",
      "Residue 95 corresponds to alignment index 120: C\n",
      "Residue 153 corresponds to alignment index 194: C\n",
      "Residue 188 corresponds to alignment index 233: S\n",
      "14 nifK seqs\n"
     ]
    }
   ],
   "source": [
    "# nifK\n",
    "\n",
    "# grab aln numbering to map to ref seq (azoto, methano)\n",
    "#important_residues = [23, 48, 106, 141] # methanosarcina\n",
    "important_residues_nifK = [70, 95, 153, 188] # azotobacter\n",
    "\n",
    "#ref_seq = 'NC_003552.1_4104'\n",
    "ref_seq_nifK = 'NZ_BSFG01000004.1_44'\n",
    "\n",
    "passing_score = 9 # S188 has not been verified to be conserved\n",
    "\n",
    "nifK_checked = check_gene('fasta_splits/nifK_split.00001', ref_seq_nifK, important_residues_nifK, passing_score, p=True)\n",
    "\n",
    "for file in glob.glob(f'fasta_splits/nifK_split.000*.aln'):\n",
    "    if file == 'fasta_splits/nifK_split.00001.aln':\n",
    "        continue\n",
    "    new = check_gene(file[:-4], ref_seq_nifK, important_residues_nifK, passing_score)\n",
    "    nifK_checked = pd.concat([nifK_checked, new])\n",
    "\n",
    "nifK_checked.drop_duplicates(subset = ['hit'], inplace = True)\n",
    "nifK_checked.set_index(['hit'], append= True, inplace = True)\n",
    "\n",
    "#remove nifD,E hits from nifK\n",
    "nifK_checked = nifK_checked[~nifK_checked.index.get_level_values(1).isin(nifD_checked.index.get_level_values(1).to_list())]\n",
    "nifK_checked = nifK_checked[~nifK_checked.index.get_level_values(1).isin(nifE_checked.index.get_level_values(1).to_list())]\n",
    "\n",
    "#for each genome, only keep the best hit per gene cluster\n",
    "nifK_checked['gene_cluster'] = nif.loc[(nifK_checked.index.get_level_values(0), nifK_checked.index.get_level_values(1), 'nifK'), \n",
    "                                       'gene_cluster'].values\n",
    "nifK_checked = nifK_checked.loc[nifK_checked.groupby(['contig', 'gene_cluster'])['score'].idxmax()]\n",
    "\n",
    "print(str(len(nifK_checked)) + \" nifK seqs\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Residue 44 corresponds to alignment index 1459: C\n",
      "11272 nifN seqs\n"
     ]
    }
   ],
   "source": [
    "# nifN --> any other residues?? also check that not nifNB --> new criteria\n",
    "\n",
    "# grab aln numbering to map to ref seq (azoto, methano)\n",
    "#important_residues = [] # methanosarcina\n",
    "important_residues_nifN = [44]#, 75] # azotobacter\n",
    "\n",
    "#ref_seq = 'NC_003552.1_4106'\n",
    "ref_seq_nifN = 'NZ_BSFG01000004.1_38'\n",
    "\n",
    "passing_score = 3\n",
    "\n",
    "nifN_checked = check_gene('fasta_splits/nifN_split.00001', ref_seq_nifN, important_residues_nifN, passing_score, p=True)\n",
    "for file in glob.glob(f'fasta_splits/nifN_split.000*.aln'):\n",
    "    if file == 'fasta_splits/nifN_split.00001.aln':\n",
    "        continue\n",
    "    new = check_gene(file[:-4], ref_seq_nifN, important_residues_nifN, passing_score)\n",
    "    nifN_checked = pd.concat([nifN_checked, new])\n",
    "\n",
    "nifN_checked.drop_duplicates(subset = ['hit'], inplace = True)\n",
    "nifN_checked.set_index(['hit'], append= True, inplace = True)\n",
    "\n",
    "# remove nifD,E,K hits from nifN\n",
    "nifN_checked = nifN_checked[~nifN_checked.index.get_level_values(1).isin(nifD_checked.index.get_level_values(1).to_list())]\n",
    "nifN_checked = nifN_checked[~nifN_checked.index.get_level_values(1).isin(nifE_checked.index.get_level_values(1).to_list())]\n",
    "nifN_checked = nifN_checked[~nifN_checked.index.get_level_values(1).isin(nifK_checked.index.get_level_values(1).to_list())]\n",
    "\n",
    "# for each genome, only keep the best hit per gene cluster\n",
    "nifN_checked['gene_cluster'] = nif.loc[(nifN_checked.index.get_level_values(0), nifN_checked.index.get_level_values(1), 'nifN'), \n",
    "                                        'gene_cluster'].values\n",
    "nifN_checked = nifN_checked.loc[nifN_checked.groupby(['contig', 'gene_cluster'])['score'].idxmax()]\n",
    "\n",
    "print(str(len(nifN_checked)) + \" nifN seqs\")\n",
    "\n",
    "## G at position (some have D, but BchB has gap)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 693,
   "metadata": {},
   "outputs": [
    {
     "ename": "UnboundLocalError",
     "evalue": "cannot access local variable 'aln' where it is not associated with a value",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[693], line 11\u001b[0m\n\u001b[1;32m      7\u001b[0m ref_seq_nifB \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mNC_003552.1_4106\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      9\u001b[0m passing_score \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m\n\u001b[0;32m---> 11\u001b[0m nifB_checked \u001b[38;5;241m=\u001b[39m \u001b[43mcheck_gene\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mnifB\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mref_seq_nifB\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mimportant_residues\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mpassing_score\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[687], line 16\u001b[0m, in \u001b[0;36mcheck_gene\u001b[0;34m(gene, ref_seq, important_residues, passing_score)\u001b[0m\n\u001b[1;32m     13\u001b[0m residue_to_alignment \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m     14\u001b[0m residue_idx \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m0\u001b[39m  \u001b[38;5;66;03m# index in original (ungapped) sequence\u001b[39;00m\n\u001b[0;32m---> 16\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m aln_idx, char \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28menumerate\u001b[39m(\u001b[43maln\u001b[49m):\n\u001b[1;32m     17\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m char \u001b[38;5;241m!=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[1;32m     18\u001b[0m         residue_idx \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n",
      "\u001b[0;31mUnboundLocalError\u001b[0m: cannot access local variable 'aln' where it is not associated with a value"
     ]
    }
   ],
   "source": [
    "# nifB--> not done\n",
    "\n",
    "# grab aln numbering to map to ref seq (azoto, methano)\n",
    "important_residues = [] # methanosarcina\n",
    "# important_residues = [] # azotobacter\n",
    "\n",
    "ref_seq_nifB = 'NC_003552.1_4106'\n",
    "\n",
    "passing_score = 0\n",
    "\n",
    "nifB_checked = check_gene('nifB', ref_seq_nifB, important_residues, passing_score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "131650 seqs failed nifDKEN check\n",
      "aligning failed sequences\n"
     ]
    }
   ],
   "source": [
    "#get seq that failed checks\n",
    "\n",
    "seqs = []\n",
    "\n",
    "for gene in 'DKEN':\n",
    "\n",
    "    # for each gene, get all hmm hit acc\n",
    "    result = list(SeqIO.parse(f\"nif{gene}.fasta\", \"fasta\"))\n",
    "    hit = [record.description.split(\" \")[-1] for record in result]\n",
    "\n",
    "    # get seq that failed check\n",
    "    for record, acc in zip(result, hit):\n",
    "        if acc not in eval(f\"list(nif{gene}_checked.index.get_level_values(1).unique())\"):\n",
    "            seq = SeqRecord(Seq(record.seq), id=record.id, description=acc)\n",
    "            seqs.append(record)\n",
    "\n",
    "\n",
    "print(str(len(seqs)) + \" seqs failed nifDKEN checks\")\n",
    "\n",
    "# Write the records to a FASTA file\n",
    "with open(f\"results/nifDKEN_{date}.fasta\", \"w\") as output_handle:\n",
    "    SeqIO.write(seqs, output_handle, \"fasta\")\n",
    "\n",
    "# add reference sequences\n",
    "for gene in 'DKEN':\n",
    "    os.system(f\"seqtk subseq results/nif{gene}.fasta results/ref_seq.ids >> results/nifDKEN_{date}.fasta\")\n",
    "\n",
    "print('aligning failed sequences', flush=True)\n",
    "\n",
    "# aln all seqs\n",
    "num = int(len(seqs)/200) +1  # how many splits\n",
    "os.system(f\"seqtk split -n {num} nifDKEN_split nifDKEN_{date}.fasta\") # split fasta file\n",
    "for i in range(num):\n",
    "    print(i+1)\n",
    "    os.system(f\"seqtk subseq results/nifDKEN_{date}.fasta results/ref_seq.ids >> results/nifDKEN_split.{str(i+1).zfill(5)}.fa\") # add reference sequences\n",
    "    os.system(f\"mafft --auto --quiet --thread 4 results/nifDKEN_split.{str(i+1).zfill(5)}.fa > results/nifDKEN_split.{str(i+1).zfill(5)}.aln\")\n",
    "   \n",
    "# takes ~ 12 min (how much does this change now?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "428 nifD seqs\n",
      "0 nifK seqs\n",
      "2870 nifE seqs\n",
      "2864 nifN seqs\n"
     ]
    }
   ],
   "source": [
    "nifD_backup = check_gene('fasta_splits/nifDKEN_split.00001', ref_seq_nifD, important_residues_nifD, 14)\n",
    "nifK_backup = check_gene('fasta_splits/nifDKEN_split.00001', ref_seq_nifK, important_residues_nifK, 12)\n",
    "nifE_backup = check_gene('fasta_splits/nifDKEN_split.00001', ref_seq_nifE, important_residues_nifE, 9)\n",
    "nifN_backup = check_gene('fasta_splits/nifDKEN_split.00001', ref_seq_nifN, important_residues_nifN, 3)\n",
    "\n",
    "for file in glob.glob(f'fasta_splits/nifDKEN_split.00*.aln'):\n",
    "    if file == 'fasta_splits/nifDKEN_split.00001.aln':\n",
    "        continue\n",
    "    new = check_gene(file[:-4], ref_seq_nifD, important_residues_nifD, passing_score)\n",
    "    nifD_backup = pd.concat([nifD_backup, new])\n",
    "\n",
    "    new = check_gene(file[:-4], ref_seq_nifK, important_residues_nifK, passing_score)\n",
    "    nifK_backup = pd.concat([nifK_backup, new])\n",
    "\n",
    "    new = check_gene(file[:-4], ref_seq_nifE, important_residues_nifE, passing_score)\n",
    "    nifE_backup = pd.concat([nifE_backup, new])\n",
    "\n",
    "    new = check_gene(file[:-4], ref_seq_nifN, important_residues_nifN, passing_score)\n",
    "    nifN_backup = pd.concat([nifN_backup, new])\n",
    "\n",
    "# set index as genome, hit\n",
    "nifD_backup.set_index(['hit'], append= True, inplace = True)\n",
    "\n",
    "# remove hits that are already in saved\n",
    "nifD_backup = nifD_backup[~nifD_backup.index.get_level_values(1).isin(nifD_checked.index.get_level_values(1).to_list())] # remove nifD\n",
    "nifD_backup = nifD_backup[~nifD_backup.index.get_level_values(1).isin(nifK_checked.index.get_level_values(1).to_list())] # remove nifK\n",
    "nifD_backup = nifD_backup[~nifD_backup.index.get_level_values(1).isin(nifE_checked.index.get_level_values(1).to_list())] # remove nifE\n",
    "nifD_backup = nifD_backup[~nifD_backup.index.get_level_values(1).isin(nifN_checked.index.get_level_values(1).to_list())] # remove nifN\n",
    "\n",
    "# for each genome, only keep the best hit per contig\n",
    "nifD_backup['gene_cluster'] = 0\n",
    "nifD_backup = nifD_backup.loc[nifD_backup.groupby(['contig'])['score'].idxmax()]\n",
    "nifD_backup.drop_duplicates(inplace = True)\n",
    "\n",
    "print(str(len(nifD_backup.index.unique())) + \" nifD seqs\", flush=True)\n",
    "nifD_checked = pd.concat([nifD_checked, nifD_backup])\n",
    "\n",
    "# set index as genome, hit\n",
    "nifK_backup.set_index(['hit'], append= True, inplace = True)\n",
    "\n",
    "# remove hits that are already in saved\n",
    "nifK_backup = nifK_backup[~nifK_backup.index.get_level_values(1).isin(nifD_checked.index.get_level_values(1).to_list())] # remove nifD\n",
    "nifK_backup = nifK_backup[~nifK_backup.index.get_level_values(1).isin(nifK_checked.index.get_level_values(1).to_list())] # remove nifK\n",
    "nifK_backup = nifK_backup[~nifK_backup.index.get_level_values(1).isin(nifE_checked.index.get_level_values(1).to_list())] # remove nifE\n",
    "nifK_backup = nifK_backup[~nifK_backup.index.get_level_values(1).isin(nifN_checked.index.get_level_values(1).to_list())] # remove nifN\n",
    "\n",
    "# for each genome, only keep the best hit per contig\n",
    "nifK_backup['gene_cluster'] = 0\n",
    "nifK_backup = nifK_backup.loc[nifK_backup.groupby(['contig'])['score'].idxmax()]\n",
    "nifK_backup.drop_duplicates(inplace = True)\n",
    "\n",
    "print(str(len(nifK_backup.index.unique())) + \" nifK seqs\", flush=True)\n",
    "nifK_checked = pd.concat([nifK_checked, nifK_backup])\n",
    "\n",
    "# set index as genome, hit\n",
    "nifE_backup.set_index(['hit'], append= True, inplace = True)\n",
    "\n",
    "# remove hits that are already in saved\n",
    "nifE_backup = nifE_backup[~nifE_backup.index.get_level_values(1).isin(nifD_checked.index.get_level_values(1).to_list())] # remove nifD\n",
    "nifE_backup = nifE_backup[~nifE_backup.index.get_level_values(1).isin(nifK_checked.index.get_level_values(1).to_list())] # remove nifK\n",
    "nifE_backup = nifE_backup[~nifE_backup.index.get_level_values(1).isin(nifE_checked.index.get_level_values(1).to_list())] # remove nifE\n",
    "nifE_backup = nifE_backup[~nifE_backup.index.get_level_values(1).isin(nifN_checked.index.get_level_values(1).to_list())] # remove nifN\n",
    "\n",
    "# for each genome, only keep the best hit per gene cluster\n",
    "nifE_backup['gene_cluster'] = 0\n",
    "nifE_backup = nifE_backup.loc[nifE_backup.groupby(['contig'])['score'].idxmax()]\n",
    "nifE_backup.drop_duplicates(inplace = True)\n",
    "\n",
    "print(str(len(nifE_backup.index.unique())) + \" nifE seqs\", flush=True)\n",
    "nifE_checked = pd.concat([nifE_checked, nifE_backup])\n",
    "\n",
    "# set index as genome, hit\n",
    "nifN_backup.set_index(['hit'], append= True, inplace = True)\n",
    "\n",
    "# remove hits that are already in saved\n",
    "nifN_backup = nifN_backup[~nifN_backup.index.get_level_values(1).isin(nifD_checked.index.get_level_values(1).to_list())] # remove nifD\n",
    "nifN_backup = nifN_backup[~nifN_backup.index.get_level_values(1).isin(nifK_checked.index.get_level_values(1).to_list())] # remove nifK\n",
    "nifN_backup = nifN_backup[~nifN_backup.index.get_level_values(1).isin(nifE_checked.index.get_level_values(1).to_list())] # remove nifE\n",
    "nifN_backup = nifN_backup[~nifN_backup.index.get_level_values(1).isin(nifN_checked.index.get_level_values(1).to_list())] # remove nifN\n",
    "\n",
    "# for each genome, only keep the best hit per gene cluster\n",
    "nifN_backup['gene_cluster'] = 0\n",
    "nifN_backup = nifN_backup.loc[nifN_backup.groupby(['contig'])['score'].idxmax()]\n",
    "nifN_backup.drop_duplicates(inplace = True)\n",
    "\n",
    "print(str(len(nifN_backup.index.unique())) + \" nifN seqs\", flush=True)\n",
    "nifN_checked = pd.concat([nifN_checked, nifN_backup])\n",
    "\n",
    "#342,0,4680,2735\n",
    "#428, 0,2870,2864\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {},
   "outputs": [],
   "source": [
    "nif_archaea = pd.read_feather('archaea/nif.feather')\n",
    "# nif = nif_archaea.copy()\n",
    "nif_bacteria = pd.read_feather('bacteria/nif.feather')\n",
    "nif = pd.concat([nif_archaea, nif_bacteria])\n",
    "\n",
    "nif.reset_index(inplace = True)\n",
    "nif.set_index(['GenomeID', 'Hit','Gene'], inplace = True)\n",
    "nif['Seq'] = ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# append updated annotation (based on conserved residue matching) to nif files\n",
    "nif['residue_match'] = ''\n",
    "nif['backup_match'] = ''\n",
    "\n",
    "nif.reset_index(level='Gene', inplace = True)\n",
    "nif.sort_index(inplace = True)\n",
    "\n",
    "# update residue match column in nif df\n",
    "for gene in 'HDKEN':\n",
    "    for genome, cols in eval(f\"nif{gene}_checked.iterrows()\"):\n",
    "        nif.loc[(genome[0], genome[1]), 'residue_match'] = \"nif\" + gene\n",
    "    \n",
    " # add backup check       \n",
    "for gene in 'DKEN':\n",
    "    for genome, cols in eval(f\"nif{gene}_backup.iterrows()\"):\n",
    "        nif.loc[(genome[0], genome[1]), 'backup_match'] = \"nif\" + gene\n",
    "\n",
    "# filter to get hits that passed residue matching\n",
    "nifH = nif[(nif.residue_match == 'nifH') & (nif.Gene == 'nifH') & (nif['Alignment Length'] > 200)]\n",
    "# nifB = nif[(nif.Gene == 'nifB')] # not done\n",
    "# nifB['residue_match'] = 'nifB'\n",
    "\n",
    "# only index OG matches\n",
    "nifD = nif[((nif.residue_match == 'nifD') & (nif.Gene == 'nifD') & (nif.backup_match != 'nifD') & (nif['Alignment Length'] > 300))]\n",
    "# add backup check\n",
    "nifD_backup = nif[(nif.backup_match == 'nifD') & (nif['Alignment Length'] > 300)].sort_values(by = 'E-value')\n",
    "nifD_backup = nifD_backup.groupby(['GenomeID', 'Hit']).first()\n",
    "nifD = pd.concat([nifD, nifD_backup])\n",
    "\n",
    "# only index OG matches\n",
    "nifK = nif[((nif.residue_match == 'nifK') & (nif.Gene == 'nifK') & (nif.backup_match != 'nifK') & (nif['Alignment Length'] > 300))]\n",
    "# add backup check\n",
    "nifK_backup = nif[(nif.backup_match == 'nifK') & (nif['Alignment Length'] > 300)].sort_values(by = 'E-value')\n",
    "nifK_backup = nifK_backup.groupby(['GenomeID', 'Hit']).first()\n",
    "nifK = pd.concat([nifK, nifK_backup])\n",
    "\n",
    "# only index OG matches\n",
    "nifE = nif[((nif.residue_match == 'nifE') & (nif.Gene == 'nifE') & (nif.backup_match != 'nifE') & (nif['Alignment Length'] > 300))]\n",
    "# add backup check\n",
    "nifE_backup = nif[(nif.backup_match == 'nifE') & (nif['Alignment Length'] > 300)].sort_values(by = 'E-value')\n",
    "nifE_backup = nifE_backup.groupby(['GenomeID', 'Hit']).first()\n",
    "nifE = pd.concat([nifE, nifE_backup])\n",
    "\n",
    "# only index OG matches\n",
    "nifN = nif[((nif.residue_match == 'nifN') & (nif.Gene == 'nifN') & (nif.backup_match != 'nifN') & (nif['Alignment Length'] > 300))]\n",
    "# add backup check\n",
    "nifN_backup = nif[(nif.backup_match == 'nifN') & (nif['Alignment Length'] > 300)].sort_values(by = 'E-value')\n",
    "nifN_backup = nifN_backup.groupby(['GenomeID', 'Hit']).first()\n",
    "nifN = pd.concat([nifN, nifN_backup])\n",
    "\n",
    "nif = pd.concat([nifH, nifD, nifK, nifE, nifN])\n",
    "nif.sort_index(inplace = True)\n",
    "\n",
    "nif.to_csv(f'nif_rescheck_nofilt_{date}.csv')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# want to only keep contigs w/ nifHDKEN\n",
    "\n",
    "from cluster_pos import cluster_pos\n",
    "\n",
    "# make sure gene clusters have at least nifHDK\n",
    "def gene_check(genes):\n",
    "    if genes.__contains__('nifH'):\n",
    "        if genes.__contains__('nifD'):\n",
    "            if genes.__contains__('nifK'):\n",
    "                return True\n",
    "\n",
    "# multi-index to cluster by genome, contig\n",
    "nif = pd.read_csv(f'nif_rescheck_nofilt_{date}.csv')\n",
    "nif.reset_index(inplace = True)\n",
    "nif.set_index(['GenomeID', 'contig'], inplace = True)\n",
    "nif.sort_index(inplace = True)\n",
    "nif.drop_duplicates(inplace = True)\n",
    "\n",
    "# filter for genome, contig with at least 3 unique genes (nifHDKENB)\n",
    "filtered_nif = nif.groupby(level=['GenomeID', 'contig']).filter(lambda x: x['Gene'].nunique() >= 3)\n",
    "\n",
    "# make sure these 3 unique genes are not the same hit (i.e. not the same gene in reference genome)\n",
    "filtered_nif2 = filtered_nif.groupby(level=['GenomeID', 'contig']).filter(lambda x: x['Hit'].nunique() >= 3)\n",
    "\n",
    "genomes_to_keep = []\n",
    "# iterate through each genome and contig\n",
    "for genome in filtered_nif2.index.get_level_values(0).unique(): # iterate through each genome\n",
    "    for contig in filtered_nif2.loc[genome].index.get_level_values(0).unique(): # iterate through each contig\n",
    "\n",
    "        tmp = filtered_nif2.loc[(genome, contig)]\n",
    "\n",
    "        # only keep numbers that have clusters >= 3\n",
    "        pos_clusters = cluster_pos(tmp.Hit.unique(), 20)\n",
    "\n",
    "        # for each cluster, find the best combination of genes (min e-value)\n",
    "        for ind, cl in enumerate(pos_clusters):\n",
    "            pos = [contig + '_' + str(p) for p in cl]\n",
    "            no_pos = len(pos)\n",
    "            \n",
    "            # need at least 3 genes to continue\n",
    "            if no_pos < 3:\n",
    "                continue\n",
    "\n",
    "            # only keep hits that are in the cluster\n",
    "            tmp2 = tmp[tmp.Hit.isin(pos)]\n",
    "\n",
    "            # only keep hits that are in the cluster\n",
    "            tmp2 = tmp[tmp.Hit.isin(pos)].reset_index()\n",
    "\n",
    "            # check if all genes are present\n",
    "            if gene_check(tmp2.residue_match.to_list()):\n",
    "                # get index\n",
    "                items = [(genome, contig, hit) for hit in tmp2.Hit]\n",
    "                genomes_to_keep.extend(items)\n",
    "\n",
    "# filter for genomes to keep\n",
    "filtered_nif2.set_index(['Hit'], append = True, inplace = True)\n",
    "filtered_nif2 = filtered_nif2.loc[genomes_to_keep]\n",
    "filtered_nif2.sort_index(inplace = True)\n",
    "\n",
    "#clean up cols\n",
    "filtered_nif2['Gene'] = filtered_nif2['residue_match']\n",
    "filtered_nif2 = filtered_nif2[['Gene', 'E-value', 'Bit Score', 'Location', 'Orientation', 'Alignment Length', 'Sequence Length', 'GTDB']]\n",
    "filtered_nif2.drop_duplicates(inplace = True)\n",
    "\n",
    "# # export \n",
    "# filtered_nif2.to_feather(f'nif_final_{date}.feather')\n",
    "# filtered_nif2.to_csv(f'nif_final_{date}.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>E-value</th>\n",
       "      <th>Bit Score</th>\n",
       "      <th>Location</th>\n",
       "      <th>Alignment Length</th>\n",
       "      <th>Sequence Length</th>\n",
       "      <th>GTDB</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>contig</th>\n",
       "      <th>Hit</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">JAKSTR010000008.1</th>\n",
       "      <th>JAKSTR010000008.1_4</th>\n",
       "      <td>nifH</td>\n",
       "      <td>3.700000e-143</td>\n",
       "      <td>473.8</td>\n",
       "      <td>2680-3531</td>\n",
       "      <td>275</td>\n",
       "      <td>851</td>\n",
       "      <td>d__Bacteria;p__Spirochaetota;c__Spirochaetia;o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JAKSTR010000008.1_6</th>\n",
       "      <td>nifD</td>\n",
       "      <td>1.100000e-61</td>\n",
       "      <td>205.9</td>\n",
       "      <td>3907-5382</td>\n",
       "      <td>437</td>\n",
       "      <td>1475</td>\n",
       "      <td>d__Bacteria;p__Spirochaetota;c__Spirochaetia;o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JAKSTR010000008.1_7</th>\n",
       "      <td>nifK</td>\n",
       "      <td>1.200000e-38</td>\n",
       "      <td>130.2</td>\n",
       "      <td>5384-6703</td>\n",
       "      <td>445</td>\n",
       "      <td>1319</td>\n",
       "      <td>d__Bacteria;p__Spirochaetota;c__Spirochaetia;o...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       Gene        E-value  Bit Score  \\\n",
       "contig            Hit                                                   \n",
       "JAKSTR010000008.1 JAKSTR010000008.1_4  nifH  3.700000e-143      473.8   \n",
       "                  JAKSTR010000008.1_6  nifD   1.100000e-61      205.9   \n",
       "                  JAKSTR010000008.1_7  nifK   1.200000e-38      130.2   \n",
       "\n",
       "                                        Location  Alignment Length  \\\n",
       "contig            Hit                                                \n",
       "JAKSTR010000008.1 JAKSTR010000008.1_4  2680-3531               275   \n",
       "                  JAKSTR010000008.1_6  3907-5382               437   \n",
       "                  JAKSTR010000008.1_7  5384-6703               445   \n",
       "\n",
       "                                       Sequence Length  \\\n",
       "contig            Hit                                    \n",
       "JAKSTR010000008.1 JAKSTR010000008.1_4              851   \n",
       "                  JAKSTR010000008.1_6             1475   \n",
       "                  JAKSTR010000008.1_7             1319   \n",
       "\n",
       "                                                                                    GTDB  \n",
       "contig            Hit                                                                     \n",
       "JAKSTR010000008.1 JAKSTR010000008.1_4  d__Bacteria;p__Spirochaetota;c__Spirochaetia;o...  \n",
       "                  JAKSTR010000008.1_6  d__Bacteria;p__Spirochaetota;c__Spirochaetia;o...  \n",
       "                  JAKSTR010000008.1_7  d__Bacteria;p__Spirochaetota;c__Spirochaetia;o...  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_nif2 = pd.read_feather(f'nif_final_{date}.feather')\n",
    "filtered_nif2.loc[('GB_GCA_024402495.1')]\n",
    "\n",
    "# GB_GCA_023439835.1 potential E for D\n",
    "# GB_GCA_023228295.1 potential N for E, D for N\n",
    "# RS_GCF_002934545.1 potential K fro E, E for K, D for N\n",
    "# GB_GCA_016183425.1_JACPEP010000042.1_138 --> unknown\n",
    "# GB_GCA_024698145.1 N for E, D for N --> both nifD in tree\n",
    "# GB_GCA_031263985.1, GB_GCA_024402495.1 --> weird nifD\n",
    "# GB_GCA_031273455.1_JAISRZ010000156.1_5 K for N\n",
    "# GB_GCA_031273455.1_JAISRZ010000065.1_5 --> weird order but has nifHDKEN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>E-value</th>\n",
       "      <th>Best Domain E-value</th>\n",
       "      <th>Bit Score</th>\n",
       "      <th>Bias</th>\n",
       "      <th>Location</th>\n",
       "      <th>Alignment Length</th>\n",
       "      <th>Sequence Length</th>\n",
       "      <th>Flag_Eval</th>\n",
       "      <th>Flag_Bias</th>\n",
       "      <th>GTDB</th>\n",
       "      <th>gene_cluster</th>\n",
       "      <th>contig</th>\n",
       "      <th>Seq</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Hit</th>\n",
       "      <th>Gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_8</th>\n",
       "      <th>nifH</th>\n",
       "      <td>0</td>\n",
       "      <td>2.200000e-77</td>\n",
       "      <td>2.700000e-77</td>\n",
       "      <td>257.1</td>\n",
       "      <td>0.2</td>\n",
       "      <td>8036-8863</td>\n",
       "      <td>272</td>\n",
       "      <td>827</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_12</th>\n",
       "      <th>nifD</th>\n",
       "      <td>1</td>\n",
       "      <td>2.000000e-131</td>\n",
       "      <td>2.700000e-131</td>\n",
       "      <td>435.4</td>\n",
       "      <td>0.1</td>\n",
       "      <td>12198-13559</td>\n",
       "      <td>451</td>\n",
       "      <td>1361</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_13</th>\n",
       "      <th>nifD</th>\n",
       "      <td>2</td>\n",
       "      <td>9.500000e-17</td>\n",
       "      <td>1.200000e-15</td>\n",
       "      <td>57.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13706-15025</td>\n",
       "      <td>388</td>\n",
       "      <td>1319</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_7</th>\n",
       "      <th>nifD</th>\n",
       "      <td>3</td>\n",
       "      <td>2.200000e-16</td>\n",
       "      <td>4.800000e-08</td>\n",
       "      <td>56.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6401-7999</td>\n",
       "      <td>328</td>\n",
       "      <td>1598</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_11</th>\n",
       "      <th>nifD</th>\n",
       "      <td>4</td>\n",
       "      <td>5.600000e-14</td>\n",
       "      <td>1.200000e-12</td>\n",
       "      <td>48.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10843-12201</td>\n",
       "      <td>260</td>\n",
       "      <td>1358</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_6</th>\n",
       "      <th>nifD</th>\n",
       "      <td>5</td>\n",
       "      <td>4.600000e-04</td>\n",
       "      <td>4.000000e+01</td>\n",
       "      <td>15.4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5102-6361</td>\n",
       "      <td>140</td>\n",
       "      <td>1259</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_13</th>\n",
       "      <th>nifK</th>\n",
       "      <td>6</td>\n",
       "      <td>1.000000e-188</td>\n",
       "      <td>1.200000e-188</td>\n",
       "      <td>624.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13706-15025</td>\n",
       "      <td>443</td>\n",
       "      <td>1319</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_11</th>\n",
       "      <th>nifK</th>\n",
       "      <td>7</td>\n",
       "      <td>2.000000e-122</td>\n",
       "      <td>2.600000e-122</td>\n",
       "      <td>406.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10843-12201</td>\n",
       "      <td>443</td>\n",
       "      <td>1358</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_12</th>\n",
       "      <th>nifK</th>\n",
       "      <td>8</td>\n",
       "      <td>4.500000e-28</td>\n",
       "      <td>4.800000e-22</td>\n",
       "      <td>94.8</td>\n",
       "      <td>0.1</td>\n",
       "      <td>12198-13559</td>\n",
       "      <td>331</td>\n",
       "      <td>1361</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_7</th>\n",
       "      <th>nifK</th>\n",
       "      <td>9</td>\n",
       "      <td>1.900000e-12</td>\n",
       "      <td>8.100000e-06</td>\n",
       "      <td>43.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6401-7999</td>\n",
       "      <td>263</td>\n",
       "      <td>1598</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_10</th>\n",
       "      <th>nifB</th>\n",
       "      <td>10</td>\n",
       "      <td>4.200000e-148</td>\n",
       "      <td>1.100000e-147</td>\n",
       "      <td>490.7</td>\n",
       "      <td>0.2</td>\n",
       "      <td>9560-10831</td>\n",
       "      <td>429</td>\n",
       "      <td>1271</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_12</th>\n",
       "      <th>nifE</th>\n",
       "      <td>11</td>\n",
       "      <td>1.700000e-186</td>\n",
       "      <td>2.100000e-186</td>\n",
       "      <td>617.7</td>\n",
       "      <td>0.1</td>\n",
       "      <td>12198-13559</td>\n",
       "      <td>445</td>\n",
       "      <td>1361</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_13</th>\n",
       "      <th>nifE</th>\n",
       "      <td>12</td>\n",
       "      <td>3.100000e-34</td>\n",
       "      <td>1.600000e-23</td>\n",
       "      <td>115.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13706-15025</td>\n",
       "      <td>411</td>\n",
       "      <td>1319</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_11</th>\n",
       "      <th>nifE</th>\n",
       "      <td>13</td>\n",
       "      <td>1.600000e-29</td>\n",
       "      <td>1.900000e-20</td>\n",
       "      <td>99.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10843-12201</td>\n",
       "      <td>388</td>\n",
       "      <td>1358</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_7</th>\n",
       "      <th>nifE</th>\n",
       "      <td>14</td>\n",
       "      <td>1.200000e-23</td>\n",
       "      <td>1.700000e-17</td>\n",
       "      <td>80.1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6401-7999</td>\n",
       "      <td>318</td>\n",
       "      <td>1598</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_6</th>\n",
       "      <th>nifE</th>\n",
       "      <td>15</td>\n",
       "      <td>5.700000e-11</td>\n",
       "      <td>7.100000e-11</td>\n",
       "      <td>38.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5102-6361</td>\n",
       "      <td>292</td>\n",
       "      <td>1259</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_11</th>\n",
       "      <th>nifN</th>\n",
       "      <td>16</td>\n",
       "      <td>4.600000e-162</td>\n",
       "      <td>5.400000e-162</td>\n",
       "      <td>537.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10843-12201</td>\n",
       "      <td>447</td>\n",
       "      <td>1358</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_13</th>\n",
       "      <th>nifN</th>\n",
       "      <td>17</td>\n",
       "      <td>4.400000e-131</td>\n",
       "      <td>5.300000e-131</td>\n",
       "      <td>435.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13706-15025</td>\n",
       "      <td>438</td>\n",
       "      <td>1319</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_12</th>\n",
       "      <th>nifN</th>\n",
       "      <td>18</td>\n",
       "      <td>4.700000e-34</td>\n",
       "      <td>1.000000e-33</td>\n",
       "      <td>114.6</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12198-13559</td>\n",
       "      <td>400</td>\n",
       "      <td>1361</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_10</th>\n",
       "      <th>nifN</th>\n",
       "      <td>19</td>\n",
       "      <td>6.100000e-21</td>\n",
       "      <td>7.100000e-21</td>\n",
       "      <td>71.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9560-10831</td>\n",
       "      <td>208</td>\n",
       "      <td>1271</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_7</th>\n",
       "      <th>nifN</th>\n",
       "      <td>20</td>\n",
       "      <td>2.700000e-13</td>\n",
       "      <td>1.900000e-12</td>\n",
       "      <td>46.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6401-7999</td>\n",
       "      <td>363</td>\n",
       "      <td>1598</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_6</th>\n",
       "      <th>nifN</th>\n",
       "      <td>21</td>\n",
       "      <td>4.100000e-06</td>\n",
       "      <td>1.700000e-03</td>\n",
       "      <td>22.2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5102-6361</td>\n",
       "      <td>134</td>\n",
       "      <td>1259</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_7</th>\n",
       "      <th>pchlide</th>\n",
       "      <td>22</td>\n",
       "      <td>4.900000e-205</td>\n",
       "      <td>5.500000e-205</td>\n",
       "      <td>678.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6401-7999</td>\n",
       "      <td>539</td>\n",
       "      <td>1598</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_12</th>\n",
       "      <th>pchlide</th>\n",
       "      <td>23</td>\n",
       "      <td>8.900000e-24</td>\n",
       "      <td>1.100000e-23</td>\n",
       "      <td>80.7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12198-13559</td>\n",
       "      <td>369</td>\n",
       "      <td>1361</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_13</th>\n",
       "      <th>pchlide</th>\n",
       "      <td>24</td>\n",
       "      <td>9.100000e-20</td>\n",
       "      <td>2.400000e-18</td>\n",
       "      <td>67.5</td>\n",
       "      <td>0.0</td>\n",
       "      <td>13706-15025</td>\n",
       "      <td>443</td>\n",
       "      <td>1319</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_11</th>\n",
       "      <th>pchlide</th>\n",
       "      <td>25</td>\n",
       "      <td>4.100000e-19</td>\n",
       "      <td>1.100000e-05</td>\n",
       "      <td>65.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10843-12201</td>\n",
       "      <td>353</td>\n",
       "      <td>1358</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SZYA01000008.1_6</th>\n",
       "      <th>pchlide</th>\n",
       "      <td>26</td>\n",
       "      <td>6.900000e-08</td>\n",
       "      <td>2.500000e-07</td>\n",
       "      <td>28.3</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5102-6361</td>\n",
       "      <td>327</td>\n",
       "      <td>1259</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...</td>\n",
       "      <td>1</td>\n",
       "      <td>SZYA01000008.1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           index        E-value  Best Domain E-value  \\\n",
       "Hit               Gene                                                 \n",
       "SZYA01000008.1_8  nifH         0   2.200000e-77         2.700000e-77   \n",
       "SZYA01000008.1_12 nifD         1  2.000000e-131        2.700000e-131   \n",
       "SZYA01000008.1_13 nifD         2   9.500000e-17         1.200000e-15   \n",
       "SZYA01000008.1_7  nifD         3   2.200000e-16         4.800000e-08   \n",
       "SZYA01000008.1_11 nifD         4   5.600000e-14         1.200000e-12   \n",
       "SZYA01000008.1_6  nifD         5   4.600000e-04         4.000000e+01   \n",
       "SZYA01000008.1_13 nifK         6  1.000000e-188        1.200000e-188   \n",
       "SZYA01000008.1_11 nifK         7  2.000000e-122        2.600000e-122   \n",
       "SZYA01000008.1_12 nifK         8   4.500000e-28         4.800000e-22   \n",
       "SZYA01000008.1_7  nifK         9   1.900000e-12         8.100000e-06   \n",
       "SZYA01000008.1_10 nifB        10  4.200000e-148        1.100000e-147   \n",
       "SZYA01000008.1_12 nifE        11  1.700000e-186        2.100000e-186   \n",
       "SZYA01000008.1_13 nifE        12   3.100000e-34         1.600000e-23   \n",
       "SZYA01000008.1_11 nifE        13   1.600000e-29         1.900000e-20   \n",
       "SZYA01000008.1_7  nifE        14   1.200000e-23         1.700000e-17   \n",
       "SZYA01000008.1_6  nifE        15   5.700000e-11         7.100000e-11   \n",
       "SZYA01000008.1_11 nifN        16  4.600000e-162        5.400000e-162   \n",
       "SZYA01000008.1_13 nifN        17  4.400000e-131        5.300000e-131   \n",
       "SZYA01000008.1_12 nifN        18   4.700000e-34         1.000000e-33   \n",
       "SZYA01000008.1_10 nifN        19   6.100000e-21         7.100000e-21   \n",
       "SZYA01000008.1_7  nifN        20   2.700000e-13         1.900000e-12   \n",
       "SZYA01000008.1_6  nifN        21   4.100000e-06         1.700000e-03   \n",
       "SZYA01000008.1_7  pchlide     22  4.900000e-205        5.500000e-205   \n",
       "SZYA01000008.1_12 pchlide     23   8.900000e-24         1.100000e-23   \n",
       "SZYA01000008.1_13 pchlide     24   9.100000e-20         2.400000e-18   \n",
       "SZYA01000008.1_11 pchlide     25   4.100000e-19         1.100000e-05   \n",
       "SZYA01000008.1_6  pchlide     26   6.900000e-08         2.500000e-07   \n",
       "\n",
       "                           Bit Score  Bias     Location  Alignment Length  \\\n",
       "Hit               Gene                                                      \n",
       "SZYA01000008.1_8  nifH         257.1   0.2    8036-8863               272   \n",
       "SZYA01000008.1_12 nifD         435.4   0.1  12198-13559               451   \n",
       "SZYA01000008.1_13 nifD          57.2   0.0  13706-15025               388   \n",
       "SZYA01000008.1_7  nifD          56.0   0.0    6401-7999               328   \n",
       "SZYA01000008.1_11 nifD          48.1   0.0  10843-12201               260   \n",
       "SZYA01000008.1_6  nifD          15.4   0.0    5102-6361               140   \n",
       "SZYA01000008.1_13 nifK         624.6   0.0  13706-15025               443   \n",
       "SZYA01000008.1_11 nifK         406.0   0.0  10843-12201               443   \n",
       "SZYA01000008.1_12 nifK          94.8   0.1  12198-13559               331   \n",
       "SZYA01000008.1_7  nifK          43.3   0.0    6401-7999               263   \n",
       "SZYA01000008.1_10 nifB         490.7   0.2   9560-10831               429   \n",
       "SZYA01000008.1_12 nifE         617.7   0.1  12198-13559               445   \n",
       "SZYA01000008.1_13 nifE         115.1   0.0  13706-15025               411   \n",
       "SZYA01000008.1_11 nifE          99.5   0.0  10843-12201               388   \n",
       "SZYA01000008.1_7  nifE          80.1   0.0    6401-7999               318   \n",
       "SZYA01000008.1_6  nifE          38.3   0.0    5102-6361               292   \n",
       "SZYA01000008.1_11 nifN         537.7   0.0  10843-12201               447   \n",
       "SZYA01000008.1_13 nifN         435.3   0.0  13706-15025               438   \n",
       "SZYA01000008.1_12 nifN         114.6   0.0  12198-13559               400   \n",
       "SZYA01000008.1_10 nifN          71.3   0.0   9560-10831               208   \n",
       "SZYA01000008.1_7  nifN          46.0   0.0    6401-7999               363   \n",
       "SZYA01000008.1_6  nifN          22.2   0.0    5102-6361               134   \n",
       "SZYA01000008.1_7  pchlide      678.7   0.0    6401-7999               539   \n",
       "SZYA01000008.1_12 pchlide       80.7   0.0  12198-13559               369   \n",
       "SZYA01000008.1_13 pchlide       67.5   0.0  13706-15025               443   \n",
       "SZYA01000008.1_11 pchlide       65.3   0.0  10843-12201               353   \n",
       "SZYA01000008.1_6  pchlide       28.3   0.0    5102-6361               327   \n",
       "\n",
       "                           Sequence Length  Flag_Eval  Flag_Bias  \\\n",
       "Hit               Gene                                             \n",
       "SZYA01000008.1_8  nifH                 827          0          0   \n",
       "SZYA01000008.1_12 nifD                1361          0          0   \n",
       "SZYA01000008.1_13 nifD                1319          0          0   \n",
       "SZYA01000008.1_7  nifD                1598          0          0   \n",
       "SZYA01000008.1_11 nifD                1358          0          0   \n",
       "SZYA01000008.1_6  nifD                1259          1          0   \n",
       "SZYA01000008.1_13 nifK                1319          0          0   \n",
       "SZYA01000008.1_11 nifK                1358          0          0   \n",
       "SZYA01000008.1_12 nifK                1361          0          0   \n",
       "SZYA01000008.1_7  nifK                1598          0          0   \n",
       "SZYA01000008.1_10 nifB                1271          0          0   \n",
       "SZYA01000008.1_12 nifE                1361          0          0   \n",
       "SZYA01000008.1_13 nifE                1319          0          0   \n",
       "SZYA01000008.1_11 nifE                1358          0          0   \n",
       "SZYA01000008.1_7  nifE                1598          0          0   \n",
       "SZYA01000008.1_6  nifE                1259          0          0   \n",
       "SZYA01000008.1_11 nifN                1358          0          0   \n",
       "SZYA01000008.1_13 nifN                1319          0          0   \n",
       "SZYA01000008.1_12 nifN                1361          0          0   \n",
       "SZYA01000008.1_10 nifN                1271          0          0   \n",
       "SZYA01000008.1_7  nifN                1598          0          0   \n",
       "SZYA01000008.1_6  nifN                1259          0          0   \n",
       "SZYA01000008.1_7  pchlide             1598          0          0   \n",
       "SZYA01000008.1_12 pchlide             1361          0          0   \n",
       "SZYA01000008.1_13 pchlide             1319          0          0   \n",
       "SZYA01000008.1_11 pchlide             1358          0          0   \n",
       "SZYA01000008.1_6  pchlide             1259          0          0   \n",
       "\n",
       "                                                                        GTDB  \\\n",
       "Hit               Gene                                                         \n",
       "SZYA01000008.1_8  nifH     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_12 nifD     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_13 nifD     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_7  nifD     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_11 nifD     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_6  nifD     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_13 nifK     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_11 nifK     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_12 nifK     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_7  nifK     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_10 nifB     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_12 nifE     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_13 nifE     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_11 nifE     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_7  nifE     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_6  nifE     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_11 nifN     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_13 nifN     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_12 nifN     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_10 nifN     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_7  nifN     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_6  nifN     d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_7  pchlide  d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_12 pchlide  d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_13 pchlide  d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_11 pchlide  d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "SZYA01000008.1_6  pchlide  d__Bacteria;p__Bacteroidota;c__Chlorobiia;o__C...   \n",
       "\n",
       "                           gene_cluster          contig Seq  \n",
       "Hit               Gene                                       \n",
       "SZYA01000008.1_8  nifH                1  SZYA01000008.1      \n",
       "SZYA01000008.1_12 nifD                1  SZYA01000008.1      \n",
       "SZYA01000008.1_13 nifD                1  SZYA01000008.1      \n",
       "SZYA01000008.1_7  nifD                1  SZYA01000008.1      \n",
       "SZYA01000008.1_11 nifD                1  SZYA01000008.1      \n",
       "SZYA01000008.1_6  nifD                1  SZYA01000008.1      \n",
       "SZYA01000008.1_13 nifK                1  SZYA01000008.1      \n",
       "SZYA01000008.1_11 nifK                1  SZYA01000008.1      \n",
       "SZYA01000008.1_12 nifK                1  SZYA01000008.1      \n",
       "SZYA01000008.1_7  nifK                1  SZYA01000008.1      \n",
       "SZYA01000008.1_10 nifB                1  SZYA01000008.1      \n",
       "SZYA01000008.1_12 nifE                1  SZYA01000008.1      \n",
       "SZYA01000008.1_13 nifE                1  SZYA01000008.1      \n",
       "SZYA01000008.1_11 nifE                1  SZYA01000008.1      \n",
       "SZYA01000008.1_7  nifE                1  SZYA01000008.1      \n",
       "SZYA01000008.1_6  nifE                1  SZYA01000008.1      \n",
       "SZYA01000008.1_11 nifN                1  SZYA01000008.1      \n",
       "SZYA01000008.1_13 nifN                1  SZYA01000008.1      \n",
       "SZYA01000008.1_12 nifN                1  SZYA01000008.1      \n",
       "SZYA01000008.1_10 nifN                1  SZYA01000008.1      \n",
       "SZYA01000008.1_7  nifN                1  SZYA01000008.1      \n",
       "SZYA01000008.1_6  nifN                1  SZYA01000008.1      \n",
       "SZYA01000008.1_7  pchlide             1  SZYA01000008.1      \n",
       "SZYA01000008.1_12 pchlide             1  SZYA01000008.1      \n",
       "SZYA01000008.1_13 pchlide             1  SZYA01000008.1      \n",
       "SZYA01000008.1_11 pchlide             1  SZYA01000008.1      \n",
       "SZYA01000008.1_6  pchlide             1  SZYA01000008.1      "
      ]
     },
     "execution_count": 270,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nif.loc[('GB_GCA_005843815.1')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Gene</th>\n",
       "      <th>E-value</th>\n",
       "      <th>Bit Score</th>\n",
       "      <th>Location</th>\n",
       "      <th>Alignment Length</th>\n",
       "      <th>Sequence Length</th>\n",
       "      <th>GTDB</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>contig</th>\n",
       "      <th>Hit</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">JAHEPY010000058.1</th>\n",
       "      <th>JAHEPY010000058.1_10</th>\n",
       "      <td>nifD</td>\n",
       "      <td>1.500000e-120</td>\n",
       "      <td>398.5</td>\n",
       "      <td>8612-9955</td>\n",
       "      <td>453</td>\n",
       "      <td>1343</td>\n",
       "      <td>d__Archaea;p__Methanobacteriota_A;c__Methanoco...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JAHEPY010000058.1_12</th>\n",
       "      <td>nifK</td>\n",
       "      <td>1.700000e-167</td>\n",
       "      <td>553.6</td>\n",
       "      <td>10549-11955</td>\n",
       "      <td>455</td>\n",
       "      <td>1406</td>\n",
       "      <td>d__Archaea;p__Methanobacteriota_A;c__Methanoco...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JAHEPY010000058.1_13</th>\n",
       "      <td>nifD</td>\n",
       "      <td>6.800000e-179</td>\n",
       "      <td>591.0</td>\n",
       "      <td>11948-13384</td>\n",
       "      <td>472</td>\n",
       "      <td>1436</td>\n",
       "      <td>d__Archaea;p__Methanobacteriota_A;c__Methanoco...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JAHEPY010000058.1_16</th>\n",
       "      <td>nifH</td>\n",
       "      <td>1.800000e-137</td>\n",
       "      <td>453.6</td>\n",
       "      <td>14379-15233</td>\n",
       "      <td>275</td>\n",
       "      <td>854</td>\n",
       "      <td>d__Archaea;p__Methanobacteriota_A;c__Methanoco...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JAHEPY010000058.1_9</th>\n",
       "      <td>nifN</td>\n",
       "      <td>1.200000e-119</td>\n",
       "      <td>396.4</td>\n",
       "      <td>7096-8415</td>\n",
       "      <td>437</td>\n",
       "      <td>1319</td>\n",
       "      <td>d__Archaea;p__Methanobacteriota_A;c__Methanoco...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Gene        E-value  Bit Score  \\\n",
       "contig            Hit                                                    \n",
       "JAHEPY010000058.1 JAHEPY010000058.1_10  nifD  1.500000e-120      398.5   \n",
       "                  JAHEPY010000058.1_12  nifK  1.700000e-167      553.6   \n",
       "                  JAHEPY010000058.1_13  nifD  6.800000e-179      591.0   \n",
       "                  JAHEPY010000058.1_16  nifH  1.800000e-137      453.6   \n",
       "                  JAHEPY010000058.1_9   nifN  1.200000e-119      396.4   \n",
       "\n",
       "                                           Location  Alignment Length  \\\n",
       "contig            Hit                                                   \n",
       "JAHEPY010000058.1 JAHEPY010000058.1_10    8612-9955               453   \n",
       "                  JAHEPY010000058.1_12  10549-11955               455   \n",
       "                  JAHEPY010000058.1_13  11948-13384               472   \n",
       "                  JAHEPY010000058.1_16  14379-15233               275   \n",
       "                  JAHEPY010000058.1_9     7096-8415               437   \n",
       "\n",
       "                                        Sequence Length  \\\n",
       "contig            Hit                                     \n",
       "JAHEPY010000058.1 JAHEPY010000058.1_10             1343   \n",
       "                  JAHEPY010000058.1_12             1406   \n",
       "                  JAHEPY010000058.1_13             1436   \n",
       "                  JAHEPY010000058.1_16              854   \n",
       "                  JAHEPY010000058.1_9              1319   \n",
       "\n",
       "                                                                                     GTDB  \n",
       "contig            Hit                                                                      \n",
       "JAHEPY010000058.1 JAHEPY010000058.1_10  d__Archaea;p__Methanobacteriota_A;c__Methanoco...  \n",
       "                  JAHEPY010000058.1_12  d__Archaea;p__Methanobacteriota_A;c__Methanoco...  \n",
       "                  JAHEPY010000058.1_13  d__Archaea;p__Methanobacteriota_A;c__Methanoco...  \n",
       "                  JAHEPY010000058.1_16  d__Archaea;p__Methanobacteriota_A;c__Methanoco...  \n",
       "                  JAHEPY010000058.1_9   d__Archaea;p__Methanobacteriota_A;c__Methanoco...  "
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "filtered_nif2.loc[('GB_GCA_019561565.1')]\n",
    "#nif[nif.GTDB.str.contains('Acidi')].head(40)#.index.get_level_values(0).unique().to_list()\n",
    "# nif[(nif.backup_match!=nif.residue_match) & (nif.backup_match == 'nifK')]\n",
    "\n",
    "# RS_GCF_027922225.1 Azotobacter vinelandii\n",
    "# RS_GCF_001941975.1 Sporomusa sphaeroides (1 nif cluster, 1 unknown cluster--should not be returned in search results)\n",
    "# RS_GCF_000214355.1 Leadbetter azotonutria ZAS-9\n",
    "# RS_GCF_000214375.1 Treponema primitia ZAS-2\n",
    "# RS_GCF_001027545.1 Endomicrobium proavitum\n",
    "# RS_GCF_000297095.1 Treponema primitia ZAS-1 (3 gene clusters--1 is potentially not nitrogenase but unknown fxn)\n",
    "# RS_GCF_000016665.1 Roseiflexus sp. RS-1 (1 nif cluster, 1 bch cluster--should not be returned in search results)\n",
    "# RS_GCF_001045675.1 Megasphaera cerevisiae (1 nif cluster, 1 unknown fxn (ruminant guts)-- should not be returned in search results) --> looks like only has altnerative nif\n",
    "# GB_GCA_016177775.1 Geobacter sp. likely only 1 true nif, 1 ambiguous (JACOUQ010000015.1)\n",
    "# RS_GCF_008830225.1 multiple copies of nifH,N (NZ_WAIE01000001.1_498 likely false--~12 away from center)\n",
    "# RS_GCF_009617575.1 Duplicate nifH needs to be dropped (NZ_WIXJ01000009.1_11)\n",
    "# GB_GCA_004347685.1 Duplicate nifH (SLTJ01000030.1_8), very close to true genes...how to remove?\n",
    "# RS_GCF_000025525.1 Methanocladococcus sp. strain FS406-22 (2nd K is E)\n",
    "# GCF_000021865.1 Rhodobacter capsulatus -- missing! needs nif, anf, vnf\n",
    "# RS_GCF_000007345.1 Methanosarcina acetivorans C2A, 1 nif, 1 anf, 1 vnf all on same contig\n",
    "# RS_GCF_000017625.1 Methanoregula boonei\n",
    "# GB_GCA_005843815.1 Chlorobium sp005843815 -- appears to have Bch + nif adjacent but only has one nifH/bchL SZYA01000008.1_8' -- COOL\n",
    "# RS_GCF_000092305.1 Methanocaldococcus infernus -- true nif gene cluster (hyperthermophilic methanogen)\n",
    "# RS_GCF_017912535.1 all out of order\n",
    "# RS_GCF_000179635.2 Ruminococcus albus 7 nifHDK (NC_014833.1_2465)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# final fasta export\n",
    "nif = pd.read_feather(f'nif_final_{date}.feather')\n",
    "\n",
    "# export fasta files\n",
    "nifH = nif[(nif.Gene == 'nifH')]\n",
    "nifD = nif[(nif.Gene == 'nifD')]\n",
    "nifK = nif[(nif.Gene == 'nifK')]\n",
    "nifB = nif[(nif.Gene == 'nifB')]\n",
    "nifE = nif[(nif.Gene == 'nifE')]\n",
    "nifN = nif[(nif.Gene == 'nifN')]\n",
    "\n",
    "gene_list = [nifH, nifD, nifK, nifB, nifE, nifN]\n",
    "gene_names = ['nifH', 'nifD', 'nifK', 'nifB', 'nifE', 'nifN']\n",
    "\n",
    "# get fasta sequences for each gene & export to fasta\n",
    "for gene, name in zip(gene_list, gene_names):\n",
    "    print(name)\n",
    "    records = []\n",
    "\n",
    "    for genome,hit in gene.iterrows():\n",
    "\n",
    "        file = glob.glob(f\"*/{genome[0]}_protein.faa\")[0]\n",
    "\n",
    "        for result in SeqIO.parse(file, \"fasta\"):\n",
    "            if result.id == genome[2]:\n",
    "                # store seq\n",
    "                gene.loc[genome, 'Seq'] = str(result.seq)\n",
    "                # convert to seqrecord\n",
    "                record = SeqRecord(Seq(result.seq), id=genome[0], description=genome[2])\n",
    "                records.append(record)\n",
    "                # exit loop once sequence is found\n",
    "                break\n",
    "        \n",
    "    # Write the records to a FASTA file\n",
    "    with open(\"check_\" + name + \".fasta\", \"w\") as output_handle:\n",
    "        SeqIO.write(records, output_handle, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export in genomic-centric organization\n",
    "nif = pd.read_feather(f'nif_final_{date}.feather')\n",
    "nif.reset_index(inplace = True)\n",
    "\n",
    "# load GTDB metadata\n",
    "GTDB_metadata = pd.read_csv('GTDB_metadata.gz', sep = '\\t', usecols=['accession', 'gtdb_taxonomy', 'ncbi_taxonomy', \n",
    "                                                'ncbi_taxonomy_unfiltered', 'ncbi_country', 'ncbi_isolation_source'])\n",
    "\n",
    "# def for sorting gene order\n",
    "def sort_order(g):\n",
    "    order = ['H', 'D', 'K', 'B', 'E', 'N']\n",
    "    return order.index(g)\n",
    "\n",
    "# get gene set (i.e. HDKEN)\n",
    "nif['Gene'] = nif['Gene'].str.replace('nif', '')\n",
    "nif['Gene set'] = nif[['GenomeID','contig','Gene']].groupby(['GenomeID','contig'])['Gene'].transform(lambda x: ''.join(sorted(x, key=sort_order)))\n",
    "# get contig positions (i.e. 1,2,3)\n",
    "nif['index'] = nif['Hit'].str.split('_').str[-1]\n",
    "nif['Position'] = nif[['GenomeID','contig','index']].groupby(['GenomeID','contig'])['index'].transform(lambda x: ','.join(sorted(x)))\n",
    "\n",
    "# group by genome, contig and save\n",
    "nif = nif[['GenomeID', 'contig', 'Gene set', 'Position', 'GTDB']]\n",
    "nif = nif.groupby(['GenomeID','contig']).first()\n",
    "nif.reset_index(level=['GenomeID','contig'], inplace = True)\n",
    "\n",
    "#merge dataframes to include accession, metadata, and sequences (for filtering)\n",
    "nif = pd.merge(nif, GTDB_metadata, left_on = 'GenomeID', right_on = 'accession', how = 'left').drop(columns = ['accession', 'gtdb_taxonomy'])\n",
    "\n",
    "nif['An/aerobic'] = ''\n",
    "nif['Thermophile'] = ''\n",
    "nif['Culture'] = ''\n",
    "nif['Reference'] = ''\n",
    "nif.sort_index(inplace = True)\n",
    "nif.set_index(['GenomeID', 'contig'], inplace = True)\n",
    "\n",
    "#nif.to_csv(f'nif_genomes_{date}.csv') #--> do not overwrite! contains hand typped info!!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Assessing operon organization (what neighbors nifHDKEN clusters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# operon organization\n",
    "\n",
    "from cluster_pos import cluster_pos\n",
    "\n",
    "# grab checked nifHDKENB\n",
    "nif = pd.read_feather(f'nif_final_{date}.feather')\n",
    "\n",
    "#multi-index to cluster by genome, contig\n",
    "nif.reset_index(inplace = True)\n",
    "nif.set_index(['GenomeID', 'contig'], inplace = True)\n",
    "nif.sort_index(inplace = True)\n",
    "nif.drop_duplicates(inplace = True)\n",
    "\n",
    "# iterate through each genome and contig\n",
    "for genome in nif.index.get_level_values(0).unique(): # iterate through each genome\n",
    "    for contig in nif.loc[genome].index.get_level_values(0).unique(): # iterate through each contig\n",
    "\n",
    "        tmp = nif.loc[(genome, contig)]\n",
    "\n",
    "        # get clusters\n",
    "        pos_clusters = cluster_pos(tmp.Hit.unique(), 40)\n",
    "        \n",
    "        # get positions within +/-12 genes of center \n",
    "        for cl in pos_clusters:\n",
    "            middle = cl[len(cl)//2] # center of cluster\n",
    "            pos = [middle+num for num in range(-12,13) if middle+num > 0]\n",
    "            acc = [contig + '_' + str(p) for p in pos] # get acc\n",
    "\n",
    "            # write list items into a .txt file\n",
    "            os.system(\"rm seqs.txt\")\n",
    "            with open(\"seqs.txt\", \"w\") as f:\n",
    "                for item in acc:\n",
    "                    f.write(f\"{item}\\n\")\n",
    "\n",
    "            # save subsets as fasta\n",
    "            try:\n",
    "                file = glob.glob(f\"*/{genome}_protein.faa\")[0]\n",
    "                os.system(f\"seqtk subseq {file} seqs.txt > operon_org/{genome}_{contig}_operon.fasta\")\n",
    "            except:\n",
    "                continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot operon-org of annotated genes\n",
    "import requests\n",
    "\n",
    "gene_abv = {'K00532': 'hydA', 'K07486':'tns', 'K07485':'tns', 'K07487':'tns', 'K13819': 'nifU'} # store ko2gene\n",
    "gene_data = pd.DataFrame(columns = ['genome', 'contig', 'query_id', 'gene', 'ko_number', 'start', 'end', 'orientation'])\n",
    "\n",
    "def ko2gene(ko):\n",
    "    try:\n",
    "        return gene_abv[ko]\n",
    "    except:\n",
    "        url = f\"https://rest.kegg.jp/get/ko:{ko}\"\n",
    "        r = requests.get(url)\n",
    "        for line in r.text.split(\"\\n\"):\n",
    "            if line.startswith(\"SYMBOL\"):\n",
    "                gene = line.split()[-1]\n",
    "                gene_abv[ko] = gene\n",
    "                return gene\n",
    "\n",
    "# grab annotation files\n",
    "for file in glob.glob(f\"../operon-org/microbeannotator/annotation_results/*\"):\n",
    "    # convet ko_number to gene abv\n",
    "    annot = pd.read_csv(file, sep = '\\t')\n",
    "    annot['gene'] = annot['ko_number'].apply(ko2gene)\n",
    "    \n",
    "    # grab genome and contig from file name\n",
    "    s = r'([A-Z_]+[A-Z0-9.]+)_([A-Z_]*[A-Z0-9.]+)'\n",
    "    genome = re.search(s, file).group(1)\n",
    "    annot['genome'] = genome\n",
    "    contig = re.search(s, file).group(2)\n",
    "    annot['contig'] = contig\n",
    "\n",
    "    # grab start, end, orientation of each gene\n",
    "    input = glob.glob(f\"../operon-org/input-fastas/{genome}_{contig}_operon.fasta\")[0]\n",
    "    for result in SeqIO.parse(input, \"fasta\"):\n",
    "        start = int(result.description.split('# ')[1])\n",
    "        end = int(result.description.split('# ')[2])\n",
    "        orientation = int(result.description.split('# ')[3])\n",
    "        try:\n",
    "            annot.loc[annot['query_id'] == result.id, 'start'] = start\n",
    "            annot.loc[annot['query_id'] == result.id, 'end'] = end\n",
    "            annot.loc[annot['query_id'] == result.id, 'orientation'] = orientation\n",
    "        except:\n",
    "            continue\n",
    "    \n",
    "    gene_data = pd.concat([gene_data, annot[['genome', 'contig', 'query_id', 'gene', 'ko_number', 'start', 'end', 'orientation']]])\n",
    "    gene_data.to_csv('operon-org-plot-data.csv')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# cool genomes\n",
    "\n",
    "to_plot = [('RS_GCF_027922225.1', 'NZ_BSFG01000004.1'),\n",
    "           ('RS_GCF_000214355.1', 'NC_015577.1'),\n",
    "           ('RS_GCF_000025525.1', 'NC_013887.1'),\n",
    "           ('RS_GCF_000017625.1', 'NC_009712.1'),\n",
    "           ('RS_GCF_000007345.1', 'NC_003552.1'), # https://www.pnas.org/doi/10.1073/pnas.0602278103\n",
    "           ('RS_GCF_000327045.1', 'NC_019940.1'),\n",
    "           ('RS_GCF_001742305.1', 'NZ_MJAT01000037.1'),\n",
    "           ('GB_GCA_031275645.1', 'JAISOL010000070.1'),\n",
    "           ('GB_GCA_001184205.1', 'LBMP01000004.1'),\n",
    "           ('RS_GCF_000016665.1', 'NC_009523.1'),\n",
    "           ('RS_GCF_000146505.1', 'NC_017448.1')] \n",
    "\n",
    "thermo = ['RS_GCF_001687335.1',\n",
    " 'RS_GCF_002898515.1',\n",
    " 'RS_GCF_002093115.1',\n",
    " 'RS_GCF_011207455.1',\n",
    " 'RS_GCF_027925145.1',\n",
    " 'GB_GCA_903896745.1',\n",
    " 'RS_GCF_000195755.1',\n",
    " 'RS_GCF_000266945.1',\n",
    " 'GB_GCA_030065275.1',\n",
    " 'RS_GCF_009688965.1',\n",
    " #'GB_GCA_947259715.1',\n",
    "# 'RS_GCF_000711225.1',\n",
    " 'GB_GCA_002423105.1',\n",
    " 'GB_GCA_025355965.1',\n",
    " #'GB_GCA_030154185.1',\n",
    " 'GB_GCA_015231595.1',\n",
    " 'GB_GCA_003232385.1',\n",
    " 'GB_GCA_023512345.1',\n",
    " 'RS_GCF_030913705.1',\n",
    " 'GB_GCA_003249225.1',\n",
    " 'GB_GCA_003245715.1']\n",
    "\n",
    "\n",
    "# RS_GCF_027922225.1 Azotobacter vinelandii\n",
    "    # NZ_BSFG01000061.1 anf\n",
    "    # NZ_BSFG01000004.1 nif\n",
    "    # NZ_BSFG01000002.1 vnf\n",
    "# RS_GCF_000179635.2 Ruminococcus albus 7\n",
    "# RS_GCF_000013085.1 Rhodospirillum rubrum --> needs nif, anf --> also has MarDK\n",
    "# RS_GCF_001941975.1 Sporomusa sphaeroides (1 nif cluster, 1 unknown cluster--should not be returned in search results)\n",
    "# RS_GCF_000214355.1 Leadbetter azotonutria ZAS-9\n",
    "# RS_GCF_000214375.1 Treponema primitia ZAS-2\n",
    "# RS_GCF_001027545.1 Endomicrobium proavitum\n",
    "# RS_GCF_000297095.1 Treponema primitia ZAS-1 (3 gene clusters--1 is potentially not nitrogenase but unknown fxn)\n",
    "# RS_GCF_000016665.1 Roseiflexus sp. RS-1 (1 nif cluster, 1 bch cluster--should not be returned in search results)\n",
    "# RS_GCF_001045675.1 Megasphaera cerevisiae (1 nif cluster, 1 unknown fxn (ruminant guts)-- should not be returned in search results)\n",
    "# GB_GCA_016177775.1 Geobacter sp. likely only 1 true nif, 1 ambiguous (JACOUQ010000015.1)\n",
    "# RS_GCF_008830225.1 multiple copies of nifH,N (NZ_WAIE01000001.1_498 likely false--~12 away from center)\n",
    "# RS_GCF_009617575.1 Duplicate nifH needs to be dropped (NZ_WIXJ01000009.1_11)\n",
    "# GB_GCA_004347685.1 Duplicate nifH (SLTJ01000030.1_8), very close to true genes...how to remove?\n",
    "# RS_GCF_000025525.1 Methanocladococcus sp. strain FS406-22\n",
    "# GCF_000021865.1 Rhodobacter capsulatus -- missing! needs nif, anf, vnf\n",
    "# RS_GCF_000007345.1 Methanosarcina acetivorans C2A, 1 nif, 1 anf, 1 vnf all on same contig\n",
    "# RS_GCF_000017625.1 Methanoregula boonei\n",
    "# GB_GCA_005843815.1 Chlorobium sp005843815 -- appears to have Bch + nif adjacent but only has one nifH/bchL SZYA01000008.1_8' -- COOL\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gene_data' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mgene_data\u001b[49m\u001b[38;5;241m.\u001b[39mloc[(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mRS_GCF_000016665.1\u001b[39m\u001b[38;5;124m'\u001b[39m)]\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m#gene_data[gene_data.gene == 'K07138'].index #016278505.1, 030653335.1, GB_GCA_031270875.1, ; 023443455.1; GB_GCA_900079695.1, \u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;66;03m#'Methanolinea_A sp018262855|sediment|GB_GCA_018262855.1|JADFDO010000016.1_17'\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m#'Methanoregula sp028679185|freshwater_wetland_soil|GB_GCA_028679185.1|JAJRBZ010000082.1_21'\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     10\u001b[0m \n\u001b[1;32m     11\u001b[0m \u001b[38;5;66;03m#NFS1 = Cysteine desulfurase,  gene supplies inorganic sulfur to these clusters by removing the sulfur from cysteine, creating alanine in the process\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'gene_data' is not defined"
     ]
    }
   ],
   "source": [
    "gene_data.loc[('RS_GCF_000016665.1')]\n",
    "#gene_data[gene_data.gene == 'K07138'].index #016278505.1, 030653335.1, GB_GCA_031270875.1, ; 023443455.1; GB_GCA_900079695.1, \n",
    "#'Methanolinea_A sp018262855|sediment|GB_GCA_018262855.1|JADFDO010000016.1_17'\n",
    "#'Methanoregula sp028679185|freshwater_wetland_soil|GB_GCA_028679185.1|JAJRBZ010000082.1_21'\n",
    "#'W589 sp011171075|estuarine_anaerobic_sediment|GB_GCA_011171075.1|DUFD01000045.1_67'\n",
    "#'Marinilabilia sp016278715|brine_fluids|RS_GCF_016278715.1|NZ_JADQBH010000117.1_7'\n",
    "\n",
    "#'Rhodopseudomonas palustris_C|water|GB_GCA_000504425.1|KI632513.1_1533'\n",
    "#'Methanoregula sp030653335|groundwater|GB_GCA_030653335.1|JAURZP010000061.1_17'\n",
    "\n",
    "#NFS1 = Cysteine desulfurase,  gene supplies inorganic sulfur to these clusters by removing the sulfur from cysteine, creating alanine in the process\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_data.loc[('GB_GCA_001184205.1')]\n",
    "#gene_data[gene_data.gene == 'K07138'].index #016278505.1, 030653335.1, GB_GCA_031270875.1, ; 023443455.1; GB_GCA_900079695.1, \n",
    "#'Methanolinea_A sp018262855|sediment|GB_GCA_018262855.1|JADFDO010000016.1_17'\n",
    "#'Methanoregula sp028679185|freshwater_wetland_soil|GB_GCA_028679185.1|JAJRBZ010000082.1_21'\n",
    "#'W589 sp011171075|estuarine_anaerobic_sediment|GB_GCA_011171075.1|DUFD01000045.1_67'\n",
    "#'Marinilabilia sp016278715|brine_fluids|RS_GCF_016278715.1|NZ_JADQBH010000117.1_7'\n",
    "\n",
    "#'Rhodopseudomonas palustris_C|water|GB_GCA_000504425.1|KI632513.1_1533'\n",
    "#'Methanoregula sp030653335|groundwater|GB_GCA_030653335.1|JAURZP010000061.1_17'\n",
    "\n",
    "#NFS1 = Cysteine desulfurase, \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_data.gene.value_counts().head(40)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'GB_GCA_020832415.1'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m/opt/anaconda3/envs/pro-pred/lib/python3.12/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'GB_GCA_020832415.1'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 16\u001b[0m\n\u001b[1;32m     13\u001b[0m \u001b[38;5;66;03m# gv.set_scale_xticks(labelsize=20)\u001b[39;00m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m genome \u001b[38;5;129;01min\u001b[39;00m genome_list:\n\u001b[0;32m---> 16\u001b[0m     genes \u001b[38;5;241m=\u001b[39m \u001b[43mgene_data\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloc\u001b[49m\u001b[43m[\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgenome\u001b[49m\u001b[43m)\u001b[49m\u001b[43m]\u001b[49m\u001b[38;5;241m.\u001b[39mgene\u001b[38;5;241m.\u001b[39mto_list()\n\u001b[1;32m     17\u001b[0m     starts \u001b[38;5;241m=\u001b[39m gene_data\u001b[38;5;241m.\u001b[39mloc[(genome)]\u001b[38;5;241m.\u001b[39mstart\u001b[38;5;241m.\u001b[39mto_list()\n\u001b[1;32m     18\u001b[0m     ends \u001b[38;5;241m=\u001b[39m gene_data\u001b[38;5;241m.\u001b[39mloc[(genome)]\u001b[38;5;241m.\u001b[39mend\u001b[38;5;241m.\u001b[39mto_list()\n",
      "File \u001b[0;32m/opt/anaconda3/envs/pro-pred/lib/python3.12/site-packages/pandas/core/indexing.py:1191\u001b[0m, in \u001b[0;36m_LocationIndexer.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   1189\u001b[0m maybe_callable \u001b[38;5;241m=\u001b[39m com\u001b[38;5;241m.\u001b[39mapply_if_callable(key, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mobj)\n\u001b[1;32m   1190\u001b[0m maybe_callable \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_deprecated_callable_usage(key, maybe_callable)\n\u001b[0;32m-> 1191\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_getitem_axis\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmaybe_callable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/pro-pred/lib/python3.12/site-packages/pandas/core/indexing.py:1431\u001b[0m, in \u001b[0;36m_LocIndexer._getitem_axis\u001b[0;34m(self, key, axis)\u001b[0m\n\u001b[1;32m   1429\u001b[0m \u001b[38;5;66;03m# fall thru to straight lookup\u001b[39;00m\n\u001b[1;32m   1430\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_validate_key(key, axis)\n\u001b[0;32m-> 1431\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_label\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/pro-pred/lib/python3.12/site-packages/pandas/core/indexing.py:1381\u001b[0m, in \u001b[0;36m_LocIndexer._get_label\u001b[0;34m(self, label, axis)\u001b[0m\n\u001b[1;32m   1379\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_get_label\u001b[39m(\u001b[38;5;28mself\u001b[39m, label, axis: AxisInt):\n\u001b[1;32m   1380\u001b[0m     \u001b[38;5;66;03m# GH#5567 this will fail if the label is not present in the axis.\u001b[39;00m\n\u001b[0;32m-> 1381\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mxs\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlabel\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43maxis\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43maxis\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/pro-pred/lib/python3.12/site-packages/pandas/core/generic.py:4293\u001b[0m, in \u001b[0;36mNDFrame.xs\u001b[0;34m(self, key, axis, level, drop_level)\u001b[0m\n\u001b[1;32m   4290\u001b[0m     index \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mindex\n\u001b[1;32m   4292\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(index, MultiIndex):\n\u001b[0;32m-> 4293\u001b[0m     loc, new_index \u001b[38;5;241m=\u001b[39m \u001b[43mindex\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_loc_level\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4294\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m drop_level:\n\u001b[1;32m   4295\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m lib\u001b[38;5;241m.\u001b[39mis_integer(loc):\n\u001b[1;32m   4296\u001b[0m             \u001b[38;5;66;03m# Slice index must be an integer or None\u001b[39;00m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/pro-pred/lib/python3.12/site-packages/pandas/core/indexes/multi.py:3290\u001b[0m, in \u001b[0;36mMultiIndex._get_loc_level\u001b[0;34m(self, key, level)\u001b[0m\n\u001b[1;32m   3288\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m indexer, maybe_mi_droplevels(indexer, ilevels)\n\u001b[1;32m   3289\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 3290\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_level_indexer\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mlevel\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mlevel\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3291\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m   3292\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(key, \u001b[38;5;28mstr\u001b[39m)\n\u001b[1;32m   3293\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlevels[level]\u001b[38;5;241m.\u001b[39m_supports_partial_string_indexing\n\u001b[1;32m   3294\u001b[0m     ):\n\u001b[1;32m   3295\u001b[0m         \u001b[38;5;66;03m# check to see if we did an exact lookup vs sliced\u001b[39;00m\n\u001b[1;32m   3296\u001b[0m         check \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mlevels[level]\u001b[38;5;241m.\u001b[39mget_loc(key)\n",
      "File \u001b[0;32m/opt/anaconda3/envs/pro-pred/lib/python3.12/site-packages/pandas/core/indexes/multi.py:3391\u001b[0m, in \u001b[0;36mMultiIndex._get_level_indexer\u001b[0;34m(self, key, level, indexer)\u001b[0m\n\u001b[1;32m   3388\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mslice\u001b[39m(i, j, step)\n\u001b[1;32m   3390\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 3391\u001b[0m     idx \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_loc_single_level_index\u001b[49m\u001b[43m(\u001b[49m\u001b[43mlevel_index\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3393\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m level \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m0\u001b[39m \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_lexsort_depth \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0\u001b[39m:\n\u001b[1;32m   3394\u001b[0m         \u001b[38;5;66;03m# Desired level is not sorted\u001b[39;00m\n\u001b[1;32m   3395\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(idx, \u001b[38;5;28mslice\u001b[39m):\n\u001b[1;32m   3396\u001b[0m             \u001b[38;5;66;03m# test_get_loc_partial_timestamp_multiindex\u001b[39;00m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/pro-pred/lib/python3.12/site-packages/pandas/core/indexes/multi.py:2980\u001b[0m, in \u001b[0;36mMultiIndex._get_loc_single_level_index\u001b[0;34m(self, level_index, key)\u001b[0m\n\u001b[1;32m   2978\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m\n\u001b[1;32m   2979\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m-> 2980\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mlevel_index\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/anaconda3/envs/pro-pred/lib/python3.12/site-packages/pandas/core/indexes/base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3808\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3809\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3810\u001b[0m     ):\n\u001b[1;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'GB_GCA_020832415.1'"
     ]
    }
   ],
   "source": [
    "# plot\n",
    "from pygenomeviz import GenomeViz\n",
    "\n",
    "gene_data = pd.read_csv('../operon-org/operon-org-plot-data.csv', index_col=[1,2])\n",
    "\n",
    "hydA = gene_data.loc[gene_data[gene_data.gene == 'hybC'].index]\n",
    "modA = gene_data.loc[gene_data[gene_data.gene == 'modA'].index]\n",
    "modAB = modA.loc[modA[modA.gene == 'modB'].index]\n",
    "genome_list = ['GB_GCA_020832415.1']\n",
    "\n",
    "gv = GenomeViz(feature_track_ratio=0.7)\n",
    "gv.set_scale_bar(labelsize=20)\n",
    "# gv.set_scale_xticks(labelsize=20)\n",
    "\n",
    "for genome in genome_list:\n",
    "    genes = gene_data.loc[(genome)].gene.to_list()\n",
    "    starts = gene_data.loc[(genome)].start.to_list()\n",
    "    ends = gene_data.loc[(genome)].end.to_list()\n",
    "    orientations = gene_data.loc[(genome)].orientation.to_list()\n",
    "\n",
    "    track = gv.add_feature_track(genome, (int(min(starts)), int(max(ends))))\n",
    "    for idx, gene in enumerate(genes):\n",
    "        if gene == 'nifH' or gene == 'vnfH' or gene == 'anfH' or gene == 'cfbC':\n",
    "            color = '#156082'\n",
    "        elif gene == 'nifD' or gene == 'vnfD' or gene == 'anfD' or gene == 'cfbD':\n",
    "            color = '#76C0AF'\n",
    "        elif gene == 'nifK' or gene == 'vnfK' or gene == 'anfK':\n",
    "            color = '#98A2C2'\n",
    "        elif gene == 'nifB':\n",
    "            color = '#C85D2B'\n",
    "        elif gene == 'nifE' or gene == 'vnfE' or gene == 'anfE':\n",
    "            color = 'orange'\n",
    "        elif gene == 'nifN' or gene == 'vnfN' or gene == 'anfN':\n",
    "            color = 'pink'\n",
    "        elif gene == 'nifA':\n",
    "            color = '#C8B2A6'\n",
    "        elif gene == 'hydA' or gene == 'hybC':\n",
    "            color = '#D5A86E'\n",
    "        elif gene == 'modA' or gene == 'modB' or gene == 'modC':\n",
    "            color = '#D5A86E'\n",
    "        else:\n",
    "            color = '#FFFFFF'\n",
    "            \n",
    "        track.add_feature(\n",
    "            int(starts[idx]),\n",
    "            int(ends[idx]),\n",
    "            int(orientations[idx]),\n",
    "            plotstyle='bigarrow',\n",
    "            arrow_shaft_ratio=0.7,\n",
    "            fc = color,\n",
    "            lw = 1,\n",
    "            label = gene,\n",
    "            text_kws=dict(rotation=0, vpos=\"bottom\", hpos=\"center\"))\n",
    "        \n",
    "    #track.add_sublabel()#int(min(starts)), int(max(ends))), fontsize=20)\n",
    "\n",
    "gv.savefig(\"../operon-org/Rhodospirillum_rubrum.png\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "OLD Below"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read fasta file merged_nifH.faa and save as df\n",
    "\n",
    "# nifH = list(SeqIO.parse(\"clustered_nifH_rep_seq.fasta\", \"fasta\"))\n",
    "# nifH = pd.DataFrame([(record.id, str(record.seq)) for record in nifH], columns = ['GenomeID', 'Seq'])\n",
    "\n",
    "nifD_list = list(SeqIO.parse(\"ko_nifD.fasta\", \"fasta\"))\n",
    "nifD_list = pd.DataFrame([(record.id, str(record.seq)) for record in nifD_list], columns = ['GenomeID', 'Seq'])\n",
    "nifD_list = nifD_list.GenomeID\n",
    "\n",
    "# nifD = list(SeqIO.parse(\"clustered_nifD_rep_seq.fasta\", \"fasta\"))\n",
    "# nifD = pd.DataFrame([(record.id, str(record.seq)) for record in nifD], columns = ['GenomeID', 'Seq'])\n",
    "\n",
    "# nifK = list(SeqIO.parse(\"clustered_nifK_rep_seq.fasta\", \"fasta\"))\n",
    "# nifK = pd.DataFrame([(record.id, str(record.seq)) for record in nifK], columns = ['GenomeID', 'Seq'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "metadata": {},
   "outputs": [],
   "source": [
    "# final\n",
    "# CDD conserved features and error tolerance\n",
    "\n",
    "nifD_FeMoCo = (r'V[A-Z]{25}R[A-Z]{97,99}H[A-Z]{29,38}Y[A-Z]I[A-Z]{43}C[A-Z]{79,81}VGG[A-Z]R[A-Z]{21,23}F[A-Z]{60}([A-Z]{50,52})?H', 5)\n",
    "# to reduce computation time, remove non-indentity residues (except Cys as it is involved in holding MoFe) from the pattern\n",
    "nifD_FeMoCo_mod = (r'V[A-Z]{25}R[A-Z]{97,99}H[A-Z]{29,38}Y[A-Z]I[A-Z]{43}C[A-Z]{106,110}F[A-Z]{60}([A-Z]{50,52})?H', 10)\n",
    "# for all other nifD check their length!\n",
    "nifD_Pcluster = (r'C[A-Z]Y[A-Z]{20}P[A-Z]GC[A-Z]{64,65}C[A-Z]{30,31}G', 5)\n",
    "\n",
    "nifK_Pcluster = (r'C[A-Z]P[A-Z]{19}S[A-Z]GC[A-Z]{2}YF[A-Z]{51,53}TC[A-Z]{30,35}S', 7)\n",
    "\n",
    "nifH_active = (r'KGGIGKS[A-Z]{24,36}KAD', 4)\n",
    "nifH_FeS = (r'C[A-Z]{29,37}C[A-Z]G',2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 274,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "typical nifD done\n",
      "nifD done\n"
     ]
    }
   ],
   "source": [
    "# check seeds seq\n",
    "\n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "import numpy as np\n",
    "\n",
    "from check_seq import check_seq\n",
    "from check_seq import conditional_check_seq\n",
    "\n",
    "#nifH\n",
    "\n",
    "# #pass constant arguments to check_seq\n",
    "# feature = nifH_active\n",
    "# filled_check_seq = partial(check_seq, feature = feature)\n",
    "# # parallelize iterables with Pool\n",
    "# with Pool(4) as p:\n",
    "#     results = p.map(filled_check_seq, nifH.Seq)\n",
    "\n",
    "# nifH.loc[:, 'match1'] = np.array(results)\n",
    "\n",
    "# # match FeS cluster binding site (binds Fe4-S4 cluster)\n",
    "# feature = nifH_FeS\n",
    "# filled_check_seq = partial(check_seq, feature = feature)\n",
    "# # parallelize iterables with Pool\n",
    "# with Pool(4) as p:\n",
    "#     results = p.map(filled_check_seq, nifH.Seq)\n",
    "# nifH.loc[:, 'match2'] = np.array(results)\n",
    "\n",
    "# print('nifH done')\n",
    "\n",
    "# match nifD MoFeCo binding site\n",
    "feature = nifD_FeMoCo\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, nifD.Seq)\n",
    "nifD.loc[:, 'match1'] = np.array(results)\n",
    "\n",
    "# match nifD MoFeCo binding site\n",
    "feature = nifD_Pcluster\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, nifD.Seq)\n",
    "nifD.loc[:, 'match2'] = np.array(results)\n",
    "\n",
    "print('typical nifD done')\n",
    "\n",
    "# match nifD MoFeCo binding site\n",
    "feature = nifD_FeMoCo\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "\n",
    "# Use Pool to apply the function\n",
    "with Pool(4) as p:\n",
    "    results = p.starmap(conditional_check_seq, zip(nifD.Seq, nifD.match1))\n",
    "\n",
    "# match nifD MoFeCo binding site\n",
    "feature = nifD_FeMoCo_mod\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, nifD.Seq)\n",
    "nifD.match2 = results\n",
    "\n",
    "print('nifD done')\n",
    "\n",
    "# # match nifD MoFeCo binding site --> all nifK pass\n",
    "# feature = nifK_Pcluster\n",
    "# filled_check_seq = partial(check_seq, feature = feature)\n",
    "# # parallelize iterables with Pool\n",
    "# with Pool(4) as p:\n",
    "#     results = p.map(filled_check_seq, nifK.Seq)\n",
    "\n",
    "\n",
    "# nifK.loc[:, 'match1'] = np.array(results)\n",
    "\n",
    "# print('nifK done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<regex.Match object; span=(73, 192), match='WGPIKDMIHISHGPVGCGYWSWSGRRNYYIGTTGIDTFGTMHFTSDFQERDIVFGGDKKLVKLIQELDVLFPLNRGVSIQSECPIGLIGDDIEAVARKTSKEIGKPVVPVRCEGFRGVS', fuzzy_counts=(4, 0, 1)>\n"
     ]
    }
   ],
   "source": [
    "# print rows where nifH.match1 is false\n",
    "# at 5, all but 2 rows fail rsin:B6N60_04474, dou:BMF77_01967\n",
    "# failed_nifH = nifH.loc[nifH.match1 != True]\n",
    "\n",
    "# failed = list(failed_nifH.Seq)\n",
    "\n",
    "nifH_active = (r'KGGIGKS[A-Z]{24,36}KAD', 5)\n",
    "\n",
    "nifK_Pcluster = (r'C[A-Z]P[A-Z]{19}S[A-Z]GC[A-Z]{2}YF[A-Z]{51,53}TC[A-Z]{30,35}S', 6)\n",
    "\n",
    "nifD_FeMoCo = (r'V[A-Z]{25}R[A-Z]{97,99}H[A-Z]{29,38}Y[A-Z]I[A-Z]{43}C[A-Z]{79,81}VGG[A-Z]R[A-Z]{21,23}F[A-Z]{60}([A-Z]{50,52})?H', 6)\n",
    "# to reduce computation time, remove non-indentity residues (except Cys as it is involved in holding MoFe) from the pattern\n",
    "nifD_FeMoCo_mod = (r'V[A-Z]{25}R[A-Z]{97,99}H[A-Z]{29,38}Y[A-Z]I[A-Z]{43}C[A-Z]{106,110}F[A-Z]{60}([A-Z]{50,52})?H', 10)\n",
    "\n",
    "seq = 'MTPPENQNIIEERKELIKEVLDAYPDKAKKKREKHLSVYEEGKSDCGVKSNIKSLPGVMTARGCAYAGSKGVVWGPIKDMIHISHGPVGCGYWSWSGRRNYYIGTTGIDTFGTMHFTSDFQERDIVFGGDKKLVKLIQELDVLFPLNRGVSIQSECPIGLIGDDIEAVARKTSKEIGKPVVPVRCEGFRGVSQSLGHHIANDMVRDWVFTRADQAKKDGTLQFEGTPYDVAIIGDYNIGGDAWASRILLEEIGLRVVAQWSGDGTINEMLMTPNVKMNLIHCYRSMNYISRHMEEAYGIPCLNYDNVTMTDENIRQIAFYGKGGIGKSTTSQNTLAAMAEMGQRILIVGCDPKADSTRLMLHSKAQTTVLHLAAERGAVEDIEIEEVMLTGFRNVRCVESGGPEPGVGCAGRGIITAINFLEENGAYQDLDFVSYDVLGDVVCGGFAMPIREGKAQEIYIVTSGEMMAMYAANNIARGVLKYAHTGGVRLGGLICNSRNTDREIELIETLAKRLNTQMIHYVPRDNIVQHAELRRMTVNEYAPESNQANEYRILAQKIIDNKNLAIPTPIEMEELEELLIEFGILESDENAAKLVGKTAAEAPVS'\n",
    "\n",
    "fuzzy_pattern = f'({nifK_Pcluster[0]}){{e<={nifK_Pcluster[1]}}}'\n",
    "\n",
    "match = regex.search(fuzzy_pattern, seq, regex.BESTMATCH)\n",
    "\n",
    "print(match)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 335,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "nifH done\n",
      "typical nifD done\n",
      "nifK done\n"
     ]
    }
   ],
   "source": [
    "# similar to final but in python file\n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "import numpy as np\n",
    "\n",
    "from check_seq import check_seq\n",
    "\n",
    "# nifH\n",
    "\n",
    "# pass constant arguments to check_seq\n",
    "feature = nifH_active\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, nifH.Seq)\n",
    "\n",
    "nifH.loc[:, 'match1'] = np.array(results)\n",
    "\n",
    "# match FeS cluster binding site (binds Fe4-S4 cluster)\n",
    "feature = nifH_FeS\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, nifH.Seq)\n",
    "nifH.loc[:, 'match2'] = np.array(results)\n",
    "\n",
    "print('nifH done')\n",
    "\n",
    "# match nifD MoFeCo binding site\n",
    "feature = nifD_FeMoCo\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, nifD.Seq)\n",
    "nifD.loc[:, 'match1'] = np.array(results)\n",
    "\n",
    "# match nifD MoFeCo binding site\n",
    "feature = nifD_Pcluster\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, nifD.Seq)\n",
    "nifD.loc[:, 'match2'] = np.array(results)\n",
    "\n",
    "print('typical nifD done')\n",
    "\n",
    "# match nifD MoFeCo binding site\n",
    "feature = nifD_FeMoCo\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "\n",
    "# Use Pool to apply the function\n",
    "with Pool(4) as p:\n",
    "    results = p.starmap(conditional_check_seq, zip(nifD.Seq, nifD.match1))\n",
    "\n",
    "# match nifD MoFeCo binding site\n",
    "feature = nifD_FeMoCo_mod\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, nifD.Seq)\n",
    "nifD.match2 = results\n",
    "\n",
    "print('nifD done')\n",
    "\n",
    "# match nifD MoFeCo binding site\n",
    "feature = nifK_Pcluster\n",
    "filled_check_seq = partial(check_seq, feature = feature)\n",
    "# parallelize iterables with Pool\n",
    "with Pool(4) as p:\n",
    "    results = p.map(filled_check_seq, nifK.Seq)\n",
    "\n",
    "\n",
    "nifK.loc[:, 'match1'] = np.array(results)\n",
    "nifK = nifK.drop(columns = 'match2')\n",
    "\n",
    "print('nifK done')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What do I do with nifH and nifD seq that fail CDD criteria? (usually they are too short and only contain a portion of the gene)?\n",
    "    - instead of hmmsearch with all NFixDB onlny use initial seeds seq?\n",
    "    - def remove them from nifH (sequences that are long enough dont have correct residues)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# finalized checks, e such that all seqs in clustered_nif_seq.fasta pass -- KEEP!!!\n",
    "\n",
    "nifD_FeMoCo = r'V[A-Z]{25}R[A-Z]{97,99}H[A-Z]{29,38}Y[A-Z]I[A-Z]{43}C[A-Z]{79,81}VGG[A-Z]R[A-Z]{21,23}F[A-Z]{60}([A-Z]{50,52})?H' # e<=4\n",
    "# to reduce computation time, remove non-indentity residues (except Cys as it is involved in holding MoFe) from the pattern\n",
    "nifD_FeMoCo_mod = r'V[A-Z]{25}R[A-Z]{97,99}H[A-Z]{29,38}Y[A-Z]I[A-Z]{43}C[A-Z]{106,110}F[A-Z]{60}([A-Z]{50,52})?H' #e<=8\n",
    "# for all other nifD check their length!\n",
    "\n",
    "nifD_Pcluster = r'C[A-Z]Y[A-Z]{20}P[A-Z]GC[A-Z]{64,65}C[A-Z]{30,31}G' # e<=5\n",
    "\n",
    "nifK_Pcluster = r'C[A-Z]P[A-Z]{19}S[A-Z]GC[A-Z]{2}YF[A-Z]{51,53}TC[A-Z]{30,35}S' # e<=8\n",
    "\n",
    "nifH_active = r'KGGIGKS[A-Z]{24,36}KAD' # e<=2\n",
    "nifH_FeS = r'C[A-Z]{29,37}C[A-Z]G' # e<=2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import SeqIO\n",
    "\n",
    "# grab sequences from same genomes and string together before clustering\n",
    "\n",
    "# join df on genomeID\n",
    "\n",
    "nif = nifH.merge(nifD, on = 'GenomeID', how = 'inner', suffixes = ('_H', '_D'))\n",
    "nif = nif.merge(nifK, on = 'GenomeID', how = 'inner', suffixes = ('_K', '_K'))\n",
    "nif['Full Seq'] = nif.Seq_H + nif.Seq_D + nif.Seq\n",
    "\n",
    "\n",
    "# Convert the DataFrame to SeqRecord objects\n",
    "records = []\n",
    "for index, row in nif.iterrows():\n",
    "    record = SeqRecord(Seq(row['Full Seq']), id=index, description=\"nifHDK\")\n",
    "    records.append(record)\n",
    "\n",
    "# Write the records to a FASTA file\n",
    "with open(\"filteredhits.fasta\", \"w\") as output_handle:\n",
    "    SeqIO.write(records, output_handle, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now I need to cluster the hits by sequence similarity before predicting their structures\n",
    "file = 'clustered_filteredhits_30_rep_seq.fasta'\n",
    "\n",
    "# get list of ids from fasta file\n",
    "ids = []\n",
    "for record in SeqIO.parse(file, \"fasta\"):\n",
    "    ids.append(record.id)\n",
    "\n",
    "clustered_nif = nif[nif.index.isin(ids)]\n",
    "\n",
    "clustered_nif = clustered_nif[['Seq_H', 'Seq_D', 'Seq']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Extra functions!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# determine if MarDK is annotated by my pipeline\n",
    "mar = pd.read_csv('../../abb6310_tables6.csv')\n",
    "mar_genomes = mar.loc[1:, 'Unnamed: 0']\n",
    "mar_contigs = mar.loc[1:, 'Unnamed: 1']\n",
    "\n",
    "nif_genomes = pd.read_csv(f'../hmmer/nif_genomes_{date}.csv')\n",
    "nif_genomes.GenomeID = nif_genomes.GenomeID.apply(lambda x:x[3:])\n",
    "nif_genomes.contig = nif_genomes.contig.apply(lambda x:'NZ_' + x.split('_')[-1])\n",
    "\n",
    "nif_mar = nif_genomes[nif_genomes.GenomeID.isin(mar_genomes) & nif_genomes.contig.isin(mar_contigs)]\n",
    "mar_nif = mar[mar['Unnamed: 0'].isin(nif_mar.GenomeID)]\n",
    "\n",
    "# MarDK\n",
    "## GCA_900100835.1, NZ_FNCM01000001.1 : 99852-101219, 101212-102807, 102822-103679\n",
    "## GCA_900184925.1, NZ_FXYJ02000005.1 : a lot going on in this region --> unclear (some nif proteins match marDK)\n",
    "## GCA_900119535.1, NZ_FPJT01000003.1 : 191972-193573, 196359-197213, 193576-194919\n",
    "## GCA_003990955.1, NZ_BIFV01000001.1 : also a lot going on here, but does not seem to be marDK, in fact, MarDK annotation seems to be incorrect based on genomic distance of proteins\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list = pd.read_csv('../GroupIVA-Members.csv')\n",
    "table = pd.read_csv(f'nif_genomes_{date}.csv')\n",
    "\n",
    "table = table[table.GenomeID.isin(list.Accession.to_list())]\n",
    "#sort table by genome ID order of list\n",
    "table['GenomeID'] = pd.Categorical(table['GenomeID'], categories=list.Accession.dropna().to_list(), ordered=True)\n",
    "table = table.sort_values('GenomeID')\n",
    "\n",
    "table.to_clipboard(index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "phylum\n",
      "p__Pseudomonadota      7029\n",
      "p__Bacillota_A         5453\n",
      "p__Desulfobacterota    2859\n",
      "p__Bacteroidota        1311\n",
      "p__Cyanobacteriota     1198\n",
      "                       ... \n",
      "p__Nitrospirota_B         3\n",
      "p__JACRDZ01               3\n",
      "p__SZUA-182               3\n",
      "p__UBA3054                3\n",
      "p__Asgardarchaeota        3\n",
      "Name: count, Length: 62, dtype: int64\n",
      "class\n",
      "c__Clostridia             5383\n",
      "c__Gammaproteobacteria    3961\n",
      "c__Alphaproteobacteria    2983\n",
      "c__Cyanobacteriia         1190\n",
      "c__Bacteroidia             861\n",
      "                          ... \n",
      "c__Limnochordia              3\n",
      "c__Methanomethylicia         3\n",
      "c__SURF-8                    3\n",
      "c__SZUA-365                  3\n",
      "c__JACRHL01                  3\n",
      "Name: count, Length: 144, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# taxa counter\n",
    "nif = pd.read_feather(f'nif_final_{date}.feather')\n",
    "nif['phylum'] = nif['GTDB'].str.split(';').str[1]\n",
    "nif['class'] = nif['GTDB'].str.split(';').str[2]\n",
    "nif['order'] = nif['GTDB'].str.split(';').str[3]\n",
    "nif['family'] = nif['GTDB'].str.split(';').str[4]\n",
    "\n",
    "phyla = nif['phylum'].value_counts()\n",
    "print(phyla)\n",
    "\n",
    "classes = nif['class'].value_counts()\n",
    "print(classes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# clean up newick file nodes = Species | Gene | GenomeID | Hit\n",
    "import Bio.Phylo.NewickIO\n",
    "\n",
    "def replace_char(string):\n",
    "    # Replace the unwanted characters with an empty string\n",
    "    return string.replace(' ', '_').replace(':', '_').replace('(', '_').replace(')', '_').replace(',', '_').replace(';', '_')\n",
    "\n",
    "# export in genomic-centric organization\n",
    "nif = pd.read_feather(f'nif_final_{date}.feather')\n",
    "nif.reset_index(inplace = True)\n",
    "nif.set_index(['GenomeID', 'Hit'], inplace = True)\n",
    "\n",
    "# supposedly prevents decimal truncation by bio.phlyo...but doesn't work\n",
    "Bio.Phylo.NewickIO._format_branch_length = lambda bl: f\"{bl:.10f}\" if bl is not None else ''\n",
    "\n",
    "# load GTDB metadata --> comment out bc it takes a while to load\n",
    "GTDB_metadata = pd.read_csv('GTDB_metadata.gz', sep = '\\t', usecols=['accession', 'gtdb_taxonomy', 'ncbi_taxonomy', \n",
    "                                                'ncbi_taxonomy_unfiltered', 'ncbi_country', 'ncbi_isolation_source'])\n",
    "\n",
    "# change names from treefiles to only be GTDB genomes\n",
    "for gene in 'DK':\n",
    "    file = f'../trees/nif{gene}_noOut_{date}/clustered_nif{gene}_noOut_rep_seq.trim.treefile'\n",
    "    tree = Phylo.read(file, \"newick\")\n",
    "    envs = []\n",
    "    # Extract and print node names\n",
    "    for clade in tree.get_terminals():\n",
    "        if clade.name.startswith('WP_'):\n",
    "            new_name = clade.name\n",
    "        else:\n",
    "            genome = clade.name.split('.')[0] + '.' + clade.name.split('.')[1][0]\n",
    "            hit = '.'.join(clade.name.split('.')[1:])[2:]\n",
    "            species = nif.loc[(genome, hit), 'GTDB'].split('__')[-1]\n",
    "            gene_name = nif.loc[(genome, hit), 'Gene']\n",
    "            env = GTDB_metadata.loc[GTDB_metadata.accession == genome, 'ncbi_isolation_source'].values[0]\n",
    "            env = replace_char(env)\n",
    "            new_name = f\"{species}|{gene_name}|{env}|{genome}|{hit}\"\n",
    "        clade.name = new_name\n",
    "\n",
    "    # Save the modified tree to a new file\n",
    "    Phylo.write(tree, f'../trees/nif{gene}_noOut_{date}/nif{gene}_env.treefile', \"newick\")\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get putative diazotroph genome names from LAMG results \n",
    "filteredhits_LAMG = list(SeqIO.parse(\"filteredhits_LAMG.fasta\", \"fasta\"))\n",
    "diazotrophs = [rec.name for rec in filteredhits_LAMG] \n",
    "\n",
    "# save as text file\n",
    "with open(\"diazotrophs_LAMG.txt\", \"w\") as f:\n",
    "    for item in diazotrophs:\n",
    "        f.write(\"%s\\n\" % item)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# grab LAMG nifH seq \n",
    "nifH = pd.read_feather('filtered_nifH_LAMG.feather')\n",
    "\n",
    "# Convert the DataFrame to SeqRecord objects\n",
    "records = []\n",
    "for index, row in nifH.iterrows():\n",
    "    record = SeqRecord(Seq(row['Seq_H']), id=index, description=\"nifH_hmm\")\n",
    "    records.append(record)\n",
    "\n",
    "# Write the records to a FASTA file\n",
    "with open(\"filtered_nifH_LAMG.fasta\", \"w\") as output_handle:\n",
    "    SeqIO.write(records, output_handle, \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Axes: >"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAigAAAGdCAYAAAA44ojeAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8hTgPZAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAs8UlEQVR4nO3dcVRcdX7//9cIwxAQ2JAow6xo0MVtFbSWaAxuN3EDk5OKqc05Zruxu9ndtBsb5UhJTpqY3zlO9ueCmx6TbEm1us1JUnM49OxRrKeJhsm3ik05/kQ0R8Btao/ZaFyQsy0CERxmyef3h1/uOiEJTBjufCDPxzlzkvnc99z7uW+G4TWfmQGPMcYIAADAIlckewIAAADnIqAAAADrEFAAAIB1CCgAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKyTmuwJXIqzZ8/q17/+tbKysuTxeJI9HQAAMAnGGA0ODioQCOiKKy6+RjIjA8qvf/1rFRQUJHsaAADgEnz00Ue65pprLlozIwNKVlaWpC9OMDs7e8L6aDSq5uZmBYNBeb3e6Z7eZY9+u4deu4t+u4deu8utfg8MDKigoMD5OX4xMzKgjL2sk52dPemAkpGRoezsbO7oLqDf7qHX7qLf7qHX7nK735N5ewZvkgUAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAABgnbgCyoIFC+TxeMZdHnroIUmSMUahUEiBQEBz5szR0qVL1dXVFbOPSCSiqqoqzZ8/X5mZmVq5cqVOnz6duDMCAAAzXlwBpa2tTd3d3c4lHA5Lku6//35J0o4dO7Rz507t2bNHbW1t8vv9qqio0ODgoLOP6upqNTU1qbGxUceOHdOZM2dUWVmp0dHRBJ4WAAD2WrDlkBZsOZTsaVgtroBy1VVXye/3O5d//dd/1Q033KAlS5bIGKPdu3dr27ZtWrVqlYqLi3XgwAENDQ2poaFBktTf36+9e/fqySefVHl5uW677TYdPHhQHR0dOnr06LScIAAAmHlSL/WGIyMjOnjwoGpqauTxePTBBx+op6dHwWDQqfH5fFqyZIlaW1u1fv16tbe3KxqNxtQEAgEVFxertbVVy5cvP++xIpGIIpGIc31gYECSFI1GFY1GJ5zrWM1kajF19Ns99Npd9Ns9s73XvhQjyZ7zc6vf8ez/kgPKiy++qE8//VTf//73JUk9PT2SpLy8vJi6vLw8nTp1yqlJS0vT3Llzx9WM3f586urqtH379nHjzc3NysjImPScx16Sgjvot3votbvot3tma6933PHFv4cPH07uRM4x3f0eGhqadO0lB5S9e/dqxYoVCgQCMeMejyfmujFm3Ni5JqrZunWrampqnOsDAwMqKChQMBhUdnb2hHONRqMKh8OqqKiQ1+udsB5TQ7/dQ6/dRb/dM9t7XRw6IknqDJ3/lQO3udXvsVdAJuOSAsqpU6d09OhRvfDCC86Y3++X9MUqSX5+vjPe29vrrKr4/X6NjIyor68vZhWlt7dXZWVlFzyez+eTz+cbN+71euNqZLz1mBr67R567S767Z7Z2uvI6BdPym07t+nudzz7vqTfg7Jv3z5dffXVuueee5yxwsJC+f3+mOWhkZERtbS0OOGjtLRUXq83pqa7u1udnZ0XDSgAAODyEvcKytmzZ7Vv3z6tXbtWqam/u7nH41F1dbVqa2tVVFSkoqIi1dbWKiMjQ2vWrJEk5eTkaN26ddq4caPmzZun3Nxcbdq0SSUlJSovL0/cWQEAgBkt7oBy9OhRffjhh/rhD384btvmzZs1PDysDRs2qK+vT4sWLVJzc7OysrKcml27dik1NVWrV6/W8PCwli1bpv379yslJWVqZwIAAGaNuANKMBiUMea82zwej0KhkEKh0AVvn56ervr6etXX18d7aAAAcJngb/EAAADrEFAAAIB1CCgAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAFlqw5ZAWbDmU7GkkDQEFAABYJzXZEwAA4HL15RWSXz1xTxJnYh9WUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAADWIaAAAADrEFAAAIB1CCgAAMA6BBQAAGCd1GRPAACAy8WCLYeSPYUZgxUUAABgHQIKAACwDgEFAABYh4ACAACsE3dA+fjjj/Xnf/7nmjdvnjIyMvQHf/AHam9vd7YbYxQKhRQIBDRnzhwtXbpUXV1dMfuIRCKqqqrS/PnzlZmZqZUrV+r06dNTPxsAADArxBVQ+vr6dNddd8nr9erll1/We++9pyeffFJf+cpXnJodO3Zo586d2rNnj9ra2uT3+1VRUaHBwUGnprq6Wk1NTWpsbNSxY8d05swZVVZWanR0NGEnBgAAZq64Pmb805/+VAUFBdq3b58ztmDBAuf/xhjt3r1b27Zt06pVqyRJBw4cUF5enhoaGrR+/Xr19/dr7969eu6551ReXi5JOnjwoAoKCnT06FEtX748AacFAABmsrgCyksvvaTly5fr/vvvV0tLi7761a9qw4YN+su//EtJ0smTJ9XT06NgMOjcxufzacmSJWptbdX69evV3t6uaDQaUxMIBFRcXKzW1tbzBpRIJKJIJOJcHxgYkCRFo1FFo9EJ5z1WM5laTB39dg+9dhf9ds9s7bUvxVxw27nnOlbrRg/c6nc8+/cYYy7crXOkp6dLkmpqanT//ffrzTffVHV1tZ555hl973vfU2trq+666y59/PHHCgQCzu1+9KMf6dSpUzpy5IgaGhr0gx/8ICZwSFIwGFRhYaGeeeaZcccNhULavn37uPGGhgZlZGRM+mQBAEDyDA0Nac2aNerv71d2dvZFa+NaQTl79qwWLlyo2tpaSdJtt92mrq4uPf300/re977n1Hk8npjbGWPGjZ3rYjVbt25VTU2Nc31gYEAFBQUKBoMTnqD0RWILh8OqqKiQ1+udsB5TQ7/dQ6/dRb/dM1t7XRw6csFtnaHYVxDGas8dnw5u9XvsFZDJiCug5Ofn66abbooZ+/3f/309//zzkiS/3y9J6unpUX5+vlPT29urvLw8p2ZkZER9fX2aO3duTE1ZWdl5j+vz+eTz+caNe73euBoZbz2mhn67h167i367Z7b1OjJ64Sfr557nWK2b5z/d/Y5n33F9iueuu+7SiRMnYsb+67/+S9ddd50kqbCwUH6/X+Fw2Nk+MjKilpYWJ3yUlpbK6/XG1HR3d6uzs/OCAQUAAFxe4lpB+eu//muVlZWptrZWq1ev1ptvvqlnn31Wzz77rKQvXtqprq5WbW2tioqKVFRUpNraWmVkZGjNmjWSpJycHK1bt04bN27UvHnzlJubq02bNqmkpMT5VA8AALi8xRVQbr/9djU1NWnr1q368Y9/rMLCQu3evVsPPPCAU7N582YNDw9rw4YN6uvr06JFi9Tc3KysrCynZteuXUpNTdXq1as1PDysZcuWaf/+/UpJSUncmQEAgBkrroAiSZWVlaqsrLzgdo/Ho1AopFAodMGa9PR01dfXq76+Pt7DAwCAywB/iwcAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQDAAgu2HNKCLYeSPQ1rEFAAAIB1CCgAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAADWIaAAAADrEFAAAIB1CCgAAMA6BBQAAGAdAgoAALBOXAElFArJ4/HEXPx+v7PdGKNQKKRAIKA5c+Zo6dKl6urqitlHJBJRVVWV5s+fr8zMTK1cuVKnT59OzNkAAIBZIe4VlJtvvlnd3d3OpaOjw9m2Y8cO7dy5U3v27FFbW5v8fr8qKio0ODjo1FRXV6upqUmNjY06duyYzpw5o8rKSo2OjibmjAAAwIyXGvcNUlNjVk3GGGO0e/dubdu2TatWrZIkHThwQHl5eWpoaND69evV39+vvXv36rnnnlN5ebkk6eDBgyooKNDRo0e1fPnyKZ4OAACYDeIOKO+//74CgYB8Pp8WLVqk2tpaXX/99Tp58qR6enoUDAadWp/PpyVLlqi1tVXr169Xe3u7otFoTE0gEFBxcbFaW1svGFAikYgikYhzfWBgQJIUjUYVjUYnnPNYzWRqMXX02z302l302z2ztde+FDNhzdg5j9W60QO3+h3P/j3GmIm79X+9/PLLGhoa0o033qhPPvlEjz/+uP7zP/9TXV1dOnHihO666y59/PHHCgQCzm1+9KMf6dSpUzpy5IgaGhr0gx/8ICZsSFIwGFRhYaGeeeaZ8x43FApp+/bt48YbGhqUkZEx2ekDAIAkGhoa0po1a9Tf36/s7OyL1sa1grJixQrn/yUlJVq8eLFuuOEGHThwQHfeeackyePxxNzGGDNu7FwT1WzdulU1NTXO9YGBARUUFCgYDE54gtIXiS0cDquiokJer3fCekwN/XYPvXYX/XbPbO11cejIhDWdoeUxtWPXp5Nb/R57BWQy4n6J58syMzNVUlKi999/X/fdd58kqaenR/n5+U5Nb2+v8vLyJEl+v18jIyPq6+vT3LlzY2rKysoueByfzyefzzdu3Ov1xtXIeOsxNfTbPfTaXfTbPbOt15HRiz9hl+Sc71itm+c/3f2OZ99T+j0okUhEv/zlL5Wfn6/CwkL5/X6Fw2Fn+8jIiFpaWpzwUVpaKq/XG1PT3d2tzs7OiwYUAABweYlrBWXTpk269957de2116q3t1ePP/64BgYGtHbtWnk8HlVXV6u2tlZFRUUqKipSbW2tMjIytGbNGklSTk6O1q1bp40bN2revHnKzc3Vpk2bVFJS4nyqBwAAIK6Acvr0aX3nO9/Rb37zG1111VW688479cYbb+i6666TJG3evFnDw8PasGGD+vr6tGjRIjU3NysrK8vZx65du5SamqrVq1dreHhYy5Yt0/79+5WSkpLYMwMAADNWXAGlsbHxots9Ho9CoZBCodAFa9LT01VfX6/6+vp4Dg0AAC4j/C0eAABgHQIKAACwDgEFAABYh4ACAACsM6Vf1AYAACa2YMuhZE9hxmEFBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAADWIaAAAADrEFAAAIB1CCgAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYJ3UZE8AAABc2IIth5z//+qJe5I4E3exggIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1plSQKmrq5PH41F1dbUzZoxRKBRSIBDQnDlztHTpUnV1dcXcLhKJqKqqSvPnz1dmZqZWrlyp06dPT2UqAABgFrnkgNLW1qZnn31Wt9xyS8z4jh07tHPnTu3Zs0dtbW3y+/2qqKjQ4OCgU1NdXa2mpiY1Njbq2LFjOnPmjCorKzU6OnrpZwIAAGaNSwooZ86c0QMPPKCf//znmjt3rjNujNHu3bu1bds2rVq1SsXFxTpw4ICGhobU0NAgServ79fevXv15JNPqry8XLfddpsOHjyojo4OHT16NDFnBQAAZrRL+mvGDz30kO655x6Vl5fr8ccfd8ZPnjypnp4eBYNBZ8zn82nJkiVqbW3V+vXr1d7ermg0GlMTCARUXFys1tZWLV++fNzxIpGIIpGIc31gYECSFI1GFY1GJ5zvWM1kajF19Ns99Npd9Ns9s63XvhQz6dqxcz7fbaarH271O579xx1QGhsb9fbbb6utrW3ctp6eHklSXl5ezHheXp5OnTrl1KSlpcWsvIzVjN3+XHV1ddq+ffu48ebmZmVkZEx67uFweNK1mDr67R567S767Z7Z0usdd0y+9vDhwxe8zdi26TLd/R4aGpp0bVwB5aOPPtIjjzyi5uZmpaenX7DO4/HEXDfGjBs718Vqtm7dqpqaGuf6wMCACgoKFAwGlZ2dPeG8o9GowuGwKioq5PV6J6zH1NBv99Brd9Fv98zkXheHjkiSOkPLx41NxtjtznebL+8zkdzq99grIJMRV0Bpb29Xb2+vSktLnbHR0VG9/vrr2rNnj06cOCHpi1WS/Px8p6a3t9dZVfH7/RoZGVFfX1/MKkpvb6/KysrOe1yfzyefzzdu3Ov1xtXIeOsxNfTbPfTaXfTbPTOx15HRL55sf3neY2OTMXa7891munsx3f2OZ99xvUl22bJl6ujo0PHjx53LwoUL9cADD+j48eO6/vrr5ff7Y5aIRkZG1NLS4oSP0tJSeb3emJru7m51dnZeMKAAAIDLS1wrKFlZWSouLo4Zy8zM1Lx585zx6upq1dbWqqioSEVFRaqtrVVGRobWrFkjScrJydG6deu0ceNGzZs3T7m5udq0aZNKSkpUXl6eoNMCAAAz2SV9iudiNm/erOHhYW3YsEF9fX1atGiRmpublZWV5dTs2rVLqampWr16tYaHh7Vs2TLt379fKSkpiZ4OAACYgaYcUF577bWY6x6PR6FQSKFQ6IK3SU9PV319verr66d6eAAAMAvxt3gAAIB1CCgAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAGCGWLDlkBZsOZTsabiCgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAww1wOv/I+NdkTAAAAvzPbg8dksYICAACsQ0ABAADWIaAAAADrEFAAAIB14gooTz/9tG655RZlZ2crOztbixcv1ssvv+xsN8YoFAopEAhozpw5Wrp0qbq6umL2EYlEVFVVpfnz5yszM1MrV67U6dOnE3M2AABgVogroFxzzTV64okn9NZbb+mtt97St771Lf3Jn/yJE0J27NihnTt3as+ePWpra5Pf71dFRYUGBwedfVRXV6upqUmNjY06duyYzpw5o8rKSo2Ojib2zAAAwIwVV0C599579cd//Me68cYbdeONN+onP/mJrrzySr3xxhsyxmj37t3atm2bVq1apeLiYh04cEBDQ0NqaGiQJPX392vv3r168sknVV5erttuu00HDx5UR0eHjh49Oi0nCAAAZp5L/j0oo6Oj+sUvfqHPPvtMixcv1smTJ9XT06NgMOjU+Hw+LVmyRK2trVq/fr3a29sVjUZjagKBgIqLi9Xa2qrly5ef91iRSESRSMS5PjAwIEmKRqOKRqMTznWsZjK1mDr67R567S767Z6Z3GtfipEUO/exsURLVH/c6nc8+487oHR0dGjx4sX6/PPPdeWVV6qpqUk33XSTWltbJUl5eXkx9Xl5eTp16pQkqaenR2lpaZo7d+64mp6engses66uTtu3bx833tzcrIyMjEnPPRwOT7oWU0e/3UOv3UW/3TMTe73jji/+PXz48LixRPvyMRJhuvs9NDQ06dq4A8rXv/51HT9+XJ9++qmef/55rV27Vi0tLc52j8cTU2+MGTd2rolqtm7dqpqaGuf6wMCACgoKFAwGlZ2dPeGco9GowuGwKioq5PV6J6zH1NBv99Brd9Fv98zkXheHjkiSOkPLx40l2pePMRVu9XvsFZDJiDugpKWl6Wtf+5okaeHChWpra9PPfvYz/c3f/I2kL1ZJ8vPznfre3l5nVcXv92tkZER9fX0xqyi9vb0qKyu74DF9Pp98Pt+4ca/XG1cj463H1NBv99Brd9Fv98zEXkdGv3jC/eV5j40lWqJ7M939jmffU/49KMYYRSIRFRYWyu/3xywPjYyMqKWlxQkfpaWl8nq9MTXd3d3q7Oy8aEABAACXl7hWUB599FGtWLFCBQUFGhwcVGNjo1577TW98sor8ng8qq6uVm1trYqKilRUVKTa2lplZGRozZo1kqScnBytW7dOGzdu1Lx585Sbm6tNmzappKRE5eXl03KCAABg5okroHzyySf67ne/q+7ubuXk5OiWW27RK6+8ooqKCknS5s2bNTw8rA0bNqivr0+LFi1Sc3OzsrKynH3s2rVLqampWr16tYaHh7Vs2TLt379fKSkpiT0zAAAwY8UVUPbu3XvR7R6PR6FQSKFQ6II16enpqq+vV319fTyHBgAAlxH+Fg8AALDOJf+iNgAAcGELthxK9hRmNFZQAACAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAADWIaAAAADrEFAAAIB1CCgAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAADWIaAAAADrEFAAAIB1CCgAAMA6BBQAAGAdAgoAALBOXAGlrq5Ot99+u7KysnT11Vfrvvvu04kTJ2JqjDEKhUIKBAKaM2eOli5dqq6urpiaSCSiqqoqzZ8/X5mZmVq5cqVOnz499bMBAACzQlwBpaWlRQ899JDeeOMNhcNh/fa3v1UwGNRnn33m1OzYsUM7d+7Unj171NbWJr/fr4qKCg0ODjo11dXVampqUmNjo44dO6YzZ86osrJSo6OjiTszAAAwY6XGU/zKK6/EXN+3b5+uvvpqtbe365vf/KaMMdq9e7e2bdumVatWSZIOHDigvLw8NTQ0aP369erv79fevXv13HPPqby8XJJ08OBBFRQU6OjRo1q+fHmCTg0AAMxUU3oPSn9/vyQpNzdXknTy5En19PQoGAw6NT6fT0uWLFFra6skqb29XdFoNKYmEAiouLjYqQEAAJe3uFZQvswYo5qaGn3jG99QcXGxJKmnp0eSlJeXF1Obl5enU6dOOTVpaWmaO3fuuJqx258rEokoEok41wcGBiRJ0WhU0Wh0wrmO1UymFlNHv91Dr91Fv90zk3vtSzGuHStR/XGr3/Hs/5IDysMPP6x3331Xx44dG7fN4/HEXDfGjBs718Vq6urqtH379nHjzc3NysjImPScw+HwpGsxdfTbPfTaXfTbPTOx1zvucO9Yhw8fTuj+prvfQ0NDk669pIBSVVWll156Sa+//rquueYaZ9zv90v6YpUkPz/fGe/t7XVWVfx+v0ZGRtTX1xezitLb26uysrLzHm/r1q2qqalxrg8MDKigoEDBYFDZ2dkTzjcajSocDquiokJerze+k0Xc6Ld76LW76Ld7ZnKvi0NHXDtWZygx79t0q99jr4BMRlwBxRijqqoqNTU16bXXXlNhYWHM9sLCQvn9foXDYd12222SpJGREbW0tOinP/2pJKm0tFRer1fhcFirV6+WJHV3d6uzs1M7duw473F9Pp98Pt+4ca/XG1cj463H1NBv99Brd9Fv98zEXkdGL/6KQSIlujfT3e949h1XQHnooYfU0NCgf/mXf1FWVpbznpGcnBzNmTNHHo9H1dXVqq2tVVFRkYqKilRbW6uMjAytWbPGqV23bp02btyoefPmKTc3V5s2bVJJSYnzqR4AAHB5iyugPP3005KkpUuXxozv27dP3//+9yVJmzdv1vDwsDZs2KC+vj4tWrRIzc3NysrKcup37dql1NRUrV69WsPDw1q2bJn279+vlJSUqZ0NAACYFeJ+iWciHo9HoVBIoVDogjXp6emqr69XfX19PIcHAACXCf4WDwAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAABgHQIKAACwDgEFAABYh4ACAACsQ0ABAADWIaAAAADrEFAAAIB1CCgAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA6xBQAACAdQgoAADAOgQUAABgHQIKAACwDgEFAIApWLDlkBZsOZTsacw6BBQAAGAdAgoAALBOarInAADAbMDLPInFCgoAALAOAQUAgFlgtr1ZN+6A8vrrr+vee+9VIBCQx+PRiy++GLPdGKNQKKRAIKA5c+Zo6dKl6urqiqmJRCKqqqrS/PnzlZmZqZUrV+r06dNTOhEAADB7xB1QPvvsM916663as2fPebfv2LFDO3fu1J49e9TW1ia/36+KigoNDg46NdXV1WpqalJjY6OOHTumM2fOqLKyUqOjo5d+JgAAYNaI+02yK1as0IoVK867zRij3bt3a9u2bVq1apUk6cCBA8rLy1NDQ4PWr1+v/v5+7d27V88995zKy8slSQcPHlRBQYGOHj2q5cuXT+F0AADAbJDQT/GcPHlSPT09CgaDzpjP59OSJUvU2tqq9evXq729XdFoNKYmEAiouLhYra2t5w0okUhEkUjEuT4wMCBJikajikajE85rrGYytZg6+u0eeu0u+u2emdRrX4pJ2rG/3J+xeVxKz9zqdzz7T2hA6enpkSTl5eXFjOfl5enUqVNOTVpamubOnTuuZuz256qrq9P27dvHjTc3NysjI2PS8wuHw5OuxdTRb/fQa3fRb/fMhF7vuCN5xz58+PC4eXx5LF7T3e+hoaFJ107L70HxeDwx140x48bOdbGarVu3qqamxrk+MDCggoICBYNBZWdnTzifaDSqcDisiooKeb3eSZwBpoJ+u4deu4t+u2cm9bo4dCTZU4jRGYr/rRJu9XvsFZDJSGhA8fv9kr5YJcnPz3fGe3t7nVUVv9+vkZER9fX1xayi9Pb2qqys7Lz79fl88vl848a9Xm9cjYy3HlNDv91Dr91Fv90zE3odGb34E3C3TaVf093vePad0N+DUlhYKL/fH7NENDIyopaWFid8lJaWyuv1xtR0d3ers7PzggEFAABcXuJeQTlz5oz++7//27l+8uRJHT9+XLm5ubr22mtVXV2t2tpaFRUVqaioSLW1tcrIyNCaNWskSTk5OVq3bp02btyoefPmKTc3V5s2bVJJSYnzqR4AAHB5izugvPXWW7r77rud62PvDVm7dq3279+vzZs3a3h4WBs2bFBfX58WLVqk5uZmZWVlObfZtWuXUlNTtXr1ag0PD2vZsmXav3+/UlJSEnBKAABgpos7oCxdulTGXPgjVR6PR6FQSKFQ6II16enpqq+vV319fbyHBwAAlwH+Fg8AALDOtHzMGACA2Ww2/VE+W7GCAgAArENAAQAA1iGgAAAwgQVbDvGyjst4DwoAAJNESHEPKygAAMA6BBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWIeAAgAArENAAQAA1iGgAAAA66QmewIAANhkwZZDzv9/9cQ9SZzJ5Y0VFAAAYB0CCia0YMuhmGcUADARHjcwVQQUAABgHQIKAACwDgEFAABYh0/x4JLxTncAsx3vo0keVlAAAIB1CCgAAMwis+UTVAQUAABgHd6DgoQYS+u8FwXATDUbVh1mEwIKgIThjdPTgycAuBwRUGYZNx7IeJYBAJhuBBTE4Jmau1hxOD/6cvka+9q///8GkzwTJBtvkgUwLWbLJwkAJAcrKIAlLsfVq8vxnPE7k/n6F4eOKDLq4WXryxABZYabzDf4pfwQ4Bt2etFf4Hf4fsD5EFBmkOl4tskDg31m4qrCxe5H526bTJiOd9tM6lUinO89Oue739hwX5qO9xPFc1429CBZbL1PTFZSA8pTTz2lv/3bv1V3d7duvvlm7d69W3/0R3+UzCm5LhmrG5d6e8IMpkui7lvnfj/Z8mbbi32fu/nJu3hWWicbgi60n6kqDh3RjjsSv98v4zHNbkkLKP/8z/+s6upqPfXUU7rrrrv0zDPPaMWKFXrvvfd07bXXJmtaCRXPg+XFku5Ujp1MxaEjOvGTynHjtiR4W+YxFWMP4mOv00sXvm+db9tkas53X5pKz+K9byb6vjyVJwW+FOP0e+y+fSnP5pMtWfOw4WuJmSNpAWXnzp1at26d/uIv/kKStHv3bh05ckRPP/206urqkjUtSVNf1ZjKA9V0rm7Y8uB4rsks3bv5THiqx0rUg+bFgsHvfmAmd262/GBI1P1/KueV6FWg6ThGoh5fEh1ap2oyT/BsffxLFtu+h8/HY4wxbh90ZGREGRkZ+sUvfqE//dM/dcYfeeQRHT9+XC0tLTH1kUhEkUjEud7f369rr71WJ0+eVFZW1oTHi0ajevXVV3X33XfL6/VOWL+o7v9Ikv6/rcsmXYvf8V1h9P/cdlaPv3OFImc9U9rX2NfgYn1OxtfpYseczvvEuf1IZK9nk3juN/F8vei3e+i1O8a+D+L9OXmpBgcHVVhYqE8//VQ5OTkXLzZJ8PHHHxtJ5j/+4z9ixn/yk5+YG2+8cVz9Y489ZiRx4cKFCxcuXGbB5aOPPpowKyT1TbIeT2wqNsaMG5OkrVu3qqamxrl+9uxZ/e///q/mzZt33vpzDQwMqKCgQB999JGys7OnPnFcFP12D712F/12D712l1v9NsZocHBQgUBgwtqkBJT58+crJSVFPT09MeO9vb3Ky8sbV+/z+eTz+WLGvvKVr8R93OzsbO7oLqLf7qHX7qLf7qHX7nKj3xO+tPN/JeVX3aelpam0tFThcDhmPBwOq6ysLBlTAgAAFknaSzw1NTX67ne/q4ULF2rx4sV69tln9eGHH+rBBx9M1pQAAIAlkhZQvv3tb+t//ud/9OMf/1jd3d0qLi7W4cOHdd111yX8WD6fT4899ti4l4kwPei3e+i1u+i3e+i1u2zsd1I+ZgwAAHAxSXkPCgAAwMUQUAAAgHUIKAAAwDoEFAAAYJ1ZH1CeeuopFRYWKj09XaWlpfr3f//3ZE9pRnr99dd17733KhAIyOPx6MUXX4zZboxRKBRSIBDQnDlztHTpUnV1dcXURCIRVVVVaf78+crMzNTKlSt1+vRpF89iZqirq9Ptt9+urKwsXX311brvvvt04sSJmBr6nThPP/20brnlFucXVC1evFgvv/yys51eT5+6ujp5PB5VV1c7Y/Q7cUKhkDweT8zF7/c7263v9VT/ro7NGhsbjdfrNT//+c/Ne++9Zx555BGTmZlpTp06leypzTiHDx8227ZtM88//7yRZJqammK2P/HEEyYrK8s8//zzpqOjw3z72982+fn5ZmBgwKl58MEHzVe/+lUTDofN22+/be6++25z6623mt/+9rcun43dli9fbvbt22c6OzvN8ePHzT333GOuvfZac+bMGaeGfifOSy+9ZA4dOmROnDhhTpw4YR599FHj9XpNZ2enMYZeT5c333zTLFiwwNxyyy3mkUceccbpd+I89thj5uabbzbd3d3Opbe319lue69ndUC54447zIMPPhgz9nu/93tmy5YtSZrR7HBuQDl79qzx+/3miSeecMY+//xzk5OTY/7hH/7BGGPMp59+arxer2lsbHRqPv74Y3PFFVeYV155xbW5z0S9vb1GkmlpaTHG0G83zJ071/zjP/4jvZ4mg4ODpqioyITDYbNkyRInoNDvxHrsscfMrbfeet5tM6HXs/YlnpGREbW3tysYDMaMB4NBtba2JmlWs9PJkyfV09MT02ufz6clS5Y4vW5vb1c0Go2pCQQCKi4u5usxgf7+fklSbm6uJPo9nUZHR9XY2KjPPvtMixcvptfT5KGHHtI999yj8vLymHH6nXjvv/++AoGACgsL9Wd/9mf64IMPJM2MXif1rxlPp9/85jcaHR0d98cH8/Lyxv2RQkzNWD/P1+tTp045NWlpaZo7d+64Gr4eF2aMUU1Njb7xjW+ouLhYEv2eDh0dHVq8eLE+//xzXXnllWpqatJNN93kPAjT68RpbGzU22+/rba2tnHbuG8n1qJFi/RP//RPuvHGG/XJJ5/o8ccfV1lZmbq6umZEr2dtQBnj8Xhirhtjxo0hMS6l13w9Lu7hhx/Wu+++q2PHjo3bRr8T5+tf/7qOHz+uTz/9VM8//7zWrl2rlpYWZzu9ToyPPvpIjzzyiJqbm5Wenn7BOvqdGCtWrHD+X1JSosWLF+uGG27QgQMHdOedd0qyu9ez9iWe+fPnKyUlZVzK6+3tHZcYMTVj7wq/WK/9fr9GRkbU19d3wRrEqqqq0ksvvaRXX31V11xzjTNOvxMvLS1NX/va17Rw4ULV1dXp1ltv1c9+9jN6nWDt7e3q7e1VaWmpUlNTlZqaqpaWFv3d3/2dUlNTnX7R7+mRmZmpkpISvf/++zPivj1rA0paWppKS0sVDodjxsPhsMrKypI0q9mpsLBQfr8/ptcjIyNqaWlxel1aWiqv1xtT093drc7OTr4e5zDG6OGHH9YLL7ygf/u3f1NhYWHMdvo9/YwxikQi9DrBli1bpo6ODh0/fty5LFy4UA888ICOHz+u66+/nn5Po0gkol/+8pfKz8+fGfftaX8bbhKNfcx479695r333jPV1dUmMzPT/OpXv0r21GacwcFB884775h33nnHSDI7d+4077zzjvOR7SeeeMLk5OSYF154wXR0dJjvfOc75/242jXXXGOOHj1q3n77bfOtb32Ljwaex1/91V+ZnJwc89prr8V8PHBoaMipod+Js3XrVvP666+bkydPmnfffdc8+uij5oorrjDNzc3GGHo93b78KR5j6Hcibdy40bz22mvmgw8+MG+88YaprKw0WVlZzs9A23s9qwOKMcb8/d//vbnuuutMWlqa+cM//EPno5qIz6uvvmokjbusXbvWGPPFR9Yee+wx4/f7jc/nM9/85jdNR0dHzD6Gh4fNww8/bHJzc82cOXNMZWWl+fDDD5NwNnY7X58lmX379jk19DtxfvjDHzqPEVdddZVZtmyZE06ModfT7dyAQr8TZ+z3mni9XhMIBMyqVatMV1eXs932XnuMMWb612kAAAAmb9a+BwUAAMxcBBQAAGAdAgoAALAOAQUAAFiHgAIAAKxDQAEAANYhoAAAAOsQUAAAgHUIKAAAwDoEFAAAYB0CCgAAsA4BBQAAWOf/B/qXPPfwBhNMAAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# how many hits have short aln len?\n",
    "bac_tophits = pd.read_feather('bacteria_tophits.feather')\n",
    "arch_tophits = pd.read_feather('archaea_tophits.feather')\n",
    "tophits = pd.concat([bac_tophits, arch_tophits])\n",
    "\n",
    "# 200bp good cutoff for nifH aln len (based on hist)\n",
    "\n",
    "nifH_tophits = tophits[tophits.index.get_level_values(1) =='nifH']\n",
    "tophits_nifH200 = nifH_tophits[nifH_tophits['Alignment Length'] > 200]\n",
    "\n",
    "filtered_accessions = tophits_nifH200.index.get_level_values(0).to_list()\n",
    "\n",
    "tophits_aln_filtered = tophits.loc[filtered_accessions]\n",
    "\n",
    "nifK_tophits = tophits_aln_filtered[tophits_aln_filtered.index.get_level_values(1) =='nifK'] \n",
    "nifK_tophits['Alignment Length'].hist(bins=200)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GenomeID</th>\n",
       "      <th>gtdb_taxonomy</th>\n",
       "      <th>ncbi_country</th>\n",
       "      <th>ncbi_isolation_source</th>\n",
       "      <th>ncbi_taxonomy</th>\n",
       "      <th>ncbi_taxonomy_unfiltered</th>\n",
       "      <th>Seq_nifD</th>\n",
       "      <th>Seq_nifK</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GB_GCA_903822005.1</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...</td>\n",
       "      <td>none</td>\n",
       "      <td>lake water</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...</td>\n",
       "      <td>MSTAIDDTRKIIDEVLEAYPEKVRKARARHLAPNDPAGSCASCKIQ...</td>\n",
       "      <td>MSKLKIKDHSELFLEPEYVDLFERKKEFEDACPDEEVKRVLDWTKT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GB_GCA_937889015.1</td>\n",
       "      <td>d__Bacteria;p__Planctomycetota;c__Planctomycet...</td>\n",
       "      <td>none</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>d__Bacteria;p__Planctomycetota;c__Planctomycet...</td>\n",
       "      <td>d__Bacteria;x__PVC group;p__Planctomycetota;c_...</td>\n",
       "      <td>MAPELIDPERVKQELLQAYPAKVGRKRAGQMVVNRKDTPGESPEIV...</td>\n",
       "      <td>MNLLKGAVRGSQENPTVLLRHTTPNVVERQALTINPAKTCQPIGAM...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GB_GCA_937924475.1</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...</td>\n",
       "      <td>none</td>\n",
       "      <td>Mariane sediment</td>\n",
       "      <td>d__Bacteria;p__;c__;o__;f__;g__;s__</td>\n",
       "      <td>d__Bacteria;x__environmental samples;s__uncult...</td>\n",
       "      <td>MNTYTINKKNEEEIRHHRVCRSRGGESCAFDGAMIVLQPIADAVHL...</td>\n",
       "      <td>MFAISTTCMAEVIGDDLNAFIKKAKQQGVIPEQLPTPYSNTSSFTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GB_GCA_002339325.1</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...</td>\n",
       "      <td>South Africa:Carltonville</td>\n",
       "      <td>water</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...</td>\n",
       "      <td>MTRIIEQDTQRIVDEVLEVYPEKTKKERKKHLLANDPECSSSTCGL...</td>\n",
       "      <td>MKVQDHVKLYRREDYEKLFRDKREFECPYPEEEVKKVEEYTKTEEY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GB_GCA_020725335.1</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...</td>\n",
       "      <td>Russia: Borehole near Beyelii Yar in Tomsk</td>\n",
       "      <td>groundwater filtered on 0.22 m filters</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...</td>\n",
       "      <td>d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...</td>\n",
       "      <td>MPRIIEQDTEKLIEEVLEIYPEKTRKERRKHLISNDPECTTSTCQL...</td>\n",
       "      <td>MKIMDHVELYKSEDYKELFENKKQFECPYSEEEVKKIEEYTKTEEY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1978</th>\n",
       "      <td>GB_GCA_026418275.1</td>\n",
       "      <td>d__Bacteria;p__Bacillota_A;c__Clostridia;o__Ac...</td>\n",
       "      <td>Malaysia: Sungkai; Perak</td>\n",
       "      <td>alkaline hot spring water</td>\n",
       "      <td>d__Bacteria;p__Bacillota;c__Clostridia;o__;f__...</td>\n",
       "      <td>d__Bacteria;x__Terrabacteria group;p__Bacillot...</td>\n",
       "      <td>MPINIQNIEAATRENRLGAITGYHGTIKDLVSKATCCSLKNRERCF...</td>\n",
       "      <td>MTQLIEQLRHVCSLGALQSVLAIHRAVPILHAGPGCGQKLWSATGV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1981</th>\n",
       "      <td>GB_GCA_026419705.1</td>\n",
       "      <td>d__Bacteria;p__Cyanobacteriota;c__Cyanobacteri...</td>\n",
       "      <td>Malaysia: Sungkai; Perak</td>\n",
       "      <td>alkaline hot spring water</td>\n",
       "      <td>d__Bacteria;p__Cyanobacteriota;c__Cyanophyceae...</td>\n",
       "      <td>d__Bacteria;x__Terrabacteria group;x__Cyanobac...</td>\n",
       "      <td>MIITQGKINELLSEPGCEHNHHTHGDKKNKACKQQAQPGAAQGGCA...</td>\n",
       "      <td>GCVAYFRTHLSRHYKEPFSAVSSSMTEDAAVFGGLNNMIEGLQVSY...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1983</th>\n",
       "      <td>GB_GCA_029266015.1</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...</td>\n",
       "      <td>China:Yunnan province</td>\n",
       "      <td>native outdoor hot spring sediment</td>\n",
       "      <td>d__Bacteria;p__Bacteroidota;c__;o__;f__;g__;s__</td>\n",
       "      <td>d__Bacteria;x__FCB group;x__Bacteroidota/Chlor...</td>\n",
       "      <td>MKTKVTMDPERWKAEVLTKYPSKVAKKRAKSIVVNEGDEVKQIQAN...</td>\n",
       "      <td>MLLRHTTDKIIEREALTVNPAKTCQPVGAMYAALGIHGCLPHSHGS...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1984</th>\n",
       "      <td>GB_GCA_030389965.1</td>\n",
       "      <td>d__Bacteria;p__Chloroflexota;c__Chloroflexia;o...</td>\n",
       "      <td>Canada: near Kenora; Ontario; Lake 227; IISD-E...</td>\n",
       "      <td>Single colony of Candidatus Chlorohelix alloph...</td>\n",
       "      <td>d__Bacteria;p__Chloroflexota;c__;o__;f__;g__;s__</td>\n",
       "      <td>d__Bacteria;x__Terrabacteria group;p__Chlorofl...</td>\n",
       "      <td>MAVLKCDATIPERDKHILIKGEGQDGKCSIACNSATTPGDMTERGC...</td>\n",
       "      <td>MSTVVKQERAVSINPLRACAPIGAMLATFGIHGALTINHGSQGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1985</th>\n",
       "      <td>GB_GCA_020049135.1</td>\n",
       "      <td>d__Bacteria;p__Verrucomicrobiota;c__Verrucomic...</td>\n",
       "      <td>USA: northwestern Lake Huron</td>\n",
       "      <td>purple microbial (photo- and chemosynthetic) m...</td>\n",
       "      <td>d__Bacteria;p__;c__;o__;f__;g__;s__</td>\n",
       "      <td>d__Bacteria;x__unclassified Bacteria;s__bacterium</td>\n",
       "      <td>MPTDIEMKNPDGIASSIPAEKIKEEMLKAYPAKTAKKRGKQILSND...</td>\n",
       "      <td>MNTVLEQEEEKEMKHCTATRNACKLCTPLGACLAFRGVEGCIPFLH...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1061 rows Ã— 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                GenomeID                                      gtdb_taxonomy  \\\n",
       "0     GB_GCA_903822005.1  d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...   \n",
       "1     GB_GCA_937889015.1  d__Bacteria;p__Planctomycetota;c__Planctomycet...   \n",
       "2     GB_GCA_937924475.1  d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...   \n",
       "3     GB_GCA_002339325.1  d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...   \n",
       "4     GB_GCA_020725335.1  d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...   \n",
       "...                  ...                                                ...   \n",
       "1978  GB_GCA_026418275.1  d__Bacteria;p__Bacillota_A;c__Clostridia;o__Ac...   \n",
       "1981  GB_GCA_026419705.1  d__Bacteria;p__Cyanobacteriota;c__Cyanobacteri...   \n",
       "1983  GB_GCA_029266015.1  d__Bacteria;p__Bacteroidota;c__Bacteroidia;o__...   \n",
       "1984  GB_GCA_030389965.1  d__Bacteria;p__Chloroflexota;c__Chloroflexia;o...   \n",
       "1985  GB_GCA_020049135.1  d__Bacteria;p__Verrucomicrobiota;c__Verrucomic...   \n",
       "\n",
       "                                           ncbi_country  \\\n",
       "0                                                  none   \n",
       "1                                                  none   \n",
       "2                                                  none   \n",
       "3                             South Africa:Carltonville   \n",
       "4            Russia: Borehole near Beyelii Yar in Tomsk   \n",
       "...                                                 ...   \n",
       "1978                           Malaysia: Sungkai; Perak   \n",
       "1981                           Malaysia: Sungkai; Perak   \n",
       "1983                              China:Yunnan province   \n",
       "1984  Canada: near Kenora; Ontario; Lake 227; IISD-E...   \n",
       "1985                       USA: northwestern Lake Huron   \n",
       "\n",
       "                                  ncbi_isolation_source  \\\n",
       "0                                            lake water   \n",
       "1                                            metagenome   \n",
       "2                                      Mariane sediment   \n",
       "3                                                 water   \n",
       "4                groundwater filtered on 0.22 m filters   \n",
       "...                                                 ...   \n",
       "1978                          alkaline hot spring water   \n",
       "1981                          alkaline hot spring water   \n",
       "1983                 native outdoor hot spring sediment   \n",
       "1984  Single colony of Candidatus Chlorohelix alloph...   \n",
       "1985  purple microbial (photo- and chemosynthetic) m...   \n",
       "\n",
       "                                          ncbi_taxonomy  \\\n",
       "0     d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...   \n",
       "1     d__Bacteria;p__Planctomycetota;c__Planctomycet...   \n",
       "2                   d__Bacteria;p__;c__;o__;f__;g__;s__   \n",
       "3     d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...   \n",
       "4     d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...   \n",
       "...                                                 ...   \n",
       "1978  d__Bacteria;p__Bacillota;c__Clostridia;o__;f__...   \n",
       "1981  d__Bacteria;p__Cyanobacteriota;c__Cyanophyceae...   \n",
       "1983    d__Bacteria;p__Bacteroidota;c__;o__;f__;g__;s__   \n",
       "1984   d__Bacteria;p__Chloroflexota;c__;o__;f__;g__;s__   \n",
       "1985                d__Bacteria;p__;c__;o__;f__;g__;s__   \n",
       "\n",
       "                               ncbi_taxonomy_unfiltered  \\\n",
       "0     d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...   \n",
       "1     d__Bacteria;x__PVC group;p__Planctomycetota;c_...   \n",
       "2     d__Bacteria;x__environmental samples;s__uncult...   \n",
       "3     d__Bacteria;p__Nitrospirota;c__Nitrospiria;o__...   \n",
       "4     d__Bacteria;p__Nitrospirota;c__Thermodesulfovi...   \n",
       "...                                                 ...   \n",
       "1978  d__Bacteria;x__Terrabacteria group;p__Bacillot...   \n",
       "1981  d__Bacteria;x__Terrabacteria group;x__Cyanobac...   \n",
       "1983  d__Bacteria;x__FCB group;x__Bacteroidota/Chlor...   \n",
       "1984  d__Bacteria;x__Terrabacteria group;p__Chlorofl...   \n",
       "1985  d__Bacteria;x__unclassified Bacteria;s__bacterium   \n",
       "\n",
       "                                               Seq_nifD  \\\n",
       "0     MSTAIDDTRKIIDEVLEAYPEKVRKARARHLAPNDPAGSCASCKIQ...   \n",
       "1     MAPELIDPERVKQELLQAYPAKVGRKRAGQMVVNRKDTPGESPEIV...   \n",
       "2     MNTYTINKKNEEEIRHHRVCRSRGGESCAFDGAMIVLQPIADAVHL...   \n",
       "3     MTRIIEQDTQRIVDEVLEVYPEKTKKERKKHLLANDPECSSSTCGL...   \n",
       "4     MPRIIEQDTEKLIEEVLEIYPEKTRKERRKHLISNDPECTTSTCQL...   \n",
       "...                                                 ...   \n",
       "1978  MPINIQNIEAATRENRLGAITGYHGTIKDLVSKATCCSLKNRERCF...   \n",
       "1981  MIITQGKINELLSEPGCEHNHHTHGDKKNKACKQQAQPGAAQGGCA...   \n",
       "1983  MKTKVTMDPERWKAEVLTKYPSKVAKKRAKSIVVNEGDEVKQIQAN...   \n",
       "1984  MAVLKCDATIPERDKHILIKGEGQDGKCSIACNSATTPGDMTERGC...   \n",
       "1985  MPTDIEMKNPDGIASSIPAEKIKEEMLKAYPAKTAKKRGKQILSND...   \n",
       "\n",
       "                                               Seq_nifK  \n",
       "0     MSKLKIKDHSELFLEPEYVDLFERKKEFEDACPDEEVKRVLDWTKT...  \n",
       "1     MNLLKGAVRGSQENPTVLLRHTTPNVVERQALTINPAKTCQPIGAM...  \n",
       "2     MFAISTTCMAEVIGDDLNAFIKKAKQQGVIPEQLPTPYSNTSSFTG...  \n",
       "3     MKVQDHVKLYRREDYEKLFRDKREFECPYPEEEVKKVEEYTKTEEY...  \n",
       "4     MKIMDHVELYKSEDYKELFENKKQFECPYSEEEVKKIEEYTKTEEY...  \n",
       "...                                                 ...  \n",
       "1978  MTQLIEQLRHVCSLGALQSVLAIHRAVPILHAGPGCGQKLWSATGV...  \n",
       "1981  GCVAYFRTHLSRHYKEPFSAVSSSMTEDAAVFGGLNNMIEGLQVSY...  \n",
       "1983  MLLRHTTDKIIEREALTVNPAKTCQPVGAMYAALGIHGCLPHSHGS...  \n",
       "1984  MSTVVKQERAVSINPLRACAPIGAMLATFGIHGALTINHGSQGCAT...  \n",
       "1985  MNTVLEQEEEKEMKHCTATRNACKLCTPLGACLAFRGVEGCIPFLH...  \n",
       "\n",
       "[1061 rows x 8 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get thermophiles (for Mikhail)\n",
    "\n",
    "# make sure seqs are in pruned tree (i.e. from nitrogenase clades)\n",
    "tree = Phylo.read(\"Nitrogenase_clades.txt\", \"newick\")\n",
    "\n",
    "# Extract and print node names\n",
    "nodes = []\n",
    "for clade in tree.get_terminals():\n",
    "    nodes.append(clade.name)\n",
    "\n",
    "# uncluster seqs (using tsv files)\n",
    "clusters = pd.read_csv('clustered_nifH_hmm_90_cluster.tsv', sep='\\t', header=None, index_col = 0)\n",
    "diazotrophs = clusters[clusters.index.isin(nodes)][1].to_list()\n",
    "\n",
    "# load GTDB metadata\n",
    "GTDB_metadata = pd.read_csv('GTDB_metadata.gz', sep = '\\t', usecols=['accession', 'gtdb_taxonomy', 'ncbi_taxonomy', \n",
    "                                                'ncbi_taxonomy_unfiltered', 'ncbi_country', 'ncbi_isolation_source'])\n",
    "\n",
    "# get nifD seqs\n",
    "nifD_seqs = list(SeqIO.parse(\"filtered_nifD_LAMG.fasta\", \"fasta\"))\n",
    "nifD_df = pd.DataFrame([(record.id, str(record.seq)) for record in nifD_seqs], columns = ['GenomeID', 'Seq'])\n",
    "\n",
    "# get nifK seqs\n",
    "nifK_seqs = list(SeqIO.parse(\"filtered_nifK_LAMG.fasta\", \"fasta\"))\n",
    "nifK_df = pd.DataFrame([(record.id, str(record.seq)) for record in nifK_seqs], columns = ['GenomeID', 'Seq'])\n",
    "\n",
    "# merge dataframes to include accession, metadata, and sequences (for filtering)\n",
    "diazotrophs = pd.DataFrame(diazotrophs, columns = ['GenomeID'])\n",
    "diazotrophs = pd.merge(diazotrophs, GTDB_metadata, left_on = 'GenomeID', right_on = 'accession', how = 'left').drop(columns = 'accession')\n",
    "diazotrophs = pd.merge(diazotrophs, nifD_df, on = 'GenomeID', how = 'left')\n",
    "diazotrophs = pd.merge(diazotrophs, nifK_df, on = 'GenomeID', how = 'left', suffixes=('_nifD', '_nifK'))\n",
    "\n",
    "# keep species w/ \"therm\" or \"heat\" in taxonomy or isolation source\n",
    "thermophiles = diazotrophs[diazotrophs.gtdb_taxonomy.str.contains('therm', case = False)]\n",
    "thermophiles =  pd.concat([thermophiles, diazotrophs[diazotrophs.ncbi_taxonomy.str.contains('therm', case = False)]], ignore_index=True)\n",
    "thermophiles =  pd.concat([thermophiles, diazotrophs[diazotrophs.ncbi_taxonomy_unfiltered.str.contains('therm', case = False)]], ignore_index=True)\n",
    "thermophiles =  pd.concat([thermophiles, diazotrophs[diazotrophs.ncbi_country.str.contains('therm', case = False)]], ignore_index=True)\n",
    "thermophiles =  pd.concat([thermophiles, diazotrophs[diazotrophs.ncbi_country.str.contains('hot', case = False)]], ignore_index=True)\n",
    "thermophiles =  pd.concat([thermophiles, diazotrophs[diazotrophs.ncbi_isolation_source.str.contains('therm', case = False)]], ignore_index=True)\n",
    "thermophiles =  pd.concat([thermophiles, diazotrophs[diazotrophs.ncbi_isolation_source.str.contains('hot', case = False)]], ignore_index=True)\n",
    "\n",
    "# remove duplicates\n",
    "thermophiles = thermophiles.drop_duplicates(subset = 'GenomeID')\n",
    "\n",
    "# export to csv\n",
    "thermophiles.to_csv('thermophiles_LAMG.csv', index = False, sep='\\t')\n",
    "\n",
    "thermophiles\n",
    "\n",
    "# awk '{print \">\" $1 \"|\" $2; print $NF}' thermophiles_LAMG.csv > thermo_nifK.fasta\n",
    "# awk '{print \">\" $1 \"|\" $2; print $(NF-1)}' thermophiles_LAMG.csv > thermo_nifD.fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "metadata": {},
   "outputs": [],
   "source": [
    "# information scoring\n",
    "\n",
    "# open aln file\n",
    "file = f\"../../thermo_nifD.aln\"\n",
    "\n",
    "# store in dataframe        \n",
    "results = list(SeqIO.parse(file, \"fasta\"))\n",
    "acc = [result.id.split('|')[0] for result in results]\n",
    "seqs = [list(str(result.seq)) for result in results]\n",
    "des = [[result.description.split('|')[-1]] for result in results]\n",
    "\n",
    "pssm = pd.DataFrame(seqs, index = acc)\n",
    "\n",
    "# get counts, frequency, entropy --> score\n",
    "score = pssm.apply(pd.Series.value_counts, axis = 0) # count\n",
    "score = score.apply(lambda x: x/1062, axis = 0) # frequency\n",
    "score.fillna(0, inplace = True) # replace Nan w/ 0\n",
    "score.loc['information'] = np.log2(22) + score.map(lambda x: x * np.log2(x) if x > 0 else 0).sum() # entropy, 0*log0 = 0\n",
    "score.iloc[:-1, :] = score.apply(lambda x: x*x.loc['information'], axis = 0).iloc[:-1,:] # individual score (closer to 4.39 = more conserved)\n",
    "\n",
    "# find sites of interest\n",
    "score.loc['azo'] = seqs[0]\n",
    "score.loc['max'] = score.iloc[1:-2, :].idxmax()\n",
    "sites = score.where(score.loc['max'] == score.loc['azo'], inplace=False)\n",
    "score.loc['azo-match'] = score.apply(lambda x: x.loc['max'] == x.loc['azo'])\n",
    "\n",
    "# sites of interest\n",
    "# sites = score[~score.loc['azo-match']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>...</th>\n",
       "      <th>1837</th>\n",
       "      <th>1838</th>\n",
       "      <th>1839</th>\n",
       "      <th>1840</th>\n",
       "      <th>1841</th>\n",
       "      <th>1842</th>\n",
       "      <th>1843</th>\n",
       "      <th>1844</th>\n",
       "      <th>1845</th>\n",
       "      <th>1846</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>-</th>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>...</td>\n",
       "      <td>4.330477</td>\n",
       "      <td>4.336893</td>\n",
       "      <td>4.332668</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "      <td>4.444419</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.028815</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004119</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004119</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>E</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004123</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>F</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>G</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.008233</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.037103</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>I</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004116</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>K</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>L</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>M</th>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>N</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.032948</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>R</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>S</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>V</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.004189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>W</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>X</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>information</th>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>...</td>\n",
       "      <td>4.371642</td>\n",
       "      <td>4.378118</td>\n",
       "      <td>4.373853</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "      <td>4.448608</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>azo</th>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>M</td>\n",
       "      <td>K</td>\n",
       "      <td>K</td>\n",
       "      <td>R</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>F</td>\n",
       "      <td>G</td>\n",
       "      <td>S</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>H</td>\n",
       "      <td>N</td>\n",
       "      <td>E</td>\n",
       "      <td>R</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>H</td>\n",
       "      <td>D</td>\n",
       "      <td>V</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>azo-match</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26 rows Ã— 1847 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 0         1         2         3         4         5     \\\n",
       "-            4.444419  4.444419  4.444419  4.444419  4.444419  4.444419   \n",
       "A                 0.0       0.0       0.0       0.0       0.0  0.004189   \n",
       "C                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "D                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "E                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "F                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "G                 0.0       0.0       0.0       0.0  0.004189       0.0   \n",
       "H                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "I                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "K                 0.0  0.004189  0.004189       0.0       0.0       0.0   \n",
       "L                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "M            0.004189       0.0       0.0       0.0       0.0       0.0   \n",
       "N                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "P                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "Q                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "R                 0.0       0.0       0.0  0.004189       0.0       0.0   \n",
       "S                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "T                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "V                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "W                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "X                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "Y                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "information  4.448608  4.448608  4.448608  4.448608  4.448608  4.448608   \n",
       "azo                 -         -         -         -         -         -   \n",
       "max                 M         K         K         R         G         A   \n",
       "azo-match       False     False     False     False     False     False   \n",
       "\n",
       "                 6         7         8         9     ...      1837      1838  \\\n",
       "-            4.444419  4.444419  4.444419  4.444419  ...  4.330477  4.336893   \n",
       "A                 0.0       0.0       0.0       0.0  ...  0.028815       0.0   \n",
       "C                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "D                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "E                 0.0       0.0       0.0       0.0  ...       0.0  0.004123   \n",
       "F                 0.0  0.004189       0.0       0.0  ...       0.0       0.0   \n",
       "G                 0.0       0.0  0.004189       0.0  ...  0.008233       0.0   \n",
       "H                 0.0       0.0       0.0       0.0  ...       0.0  0.037103   \n",
       "I                 0.0       0.0       0.0       0.0  ...  0.004116       0.0   \n",
       "K                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "L                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "M                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "N                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "P                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "Q                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "R                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "S                 0.0       0.0       0.0  0.004189  ...       0.0       0.0   \n",
       "T                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "V            0.004189       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "W                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "X                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "Y                 0.0       0.0       0.0       0.0  ...       0.0       0.0   \n",
       "information  4.448608  4.448608  4.448608  4.448608  ...  4.371642  4.378118   \n",
       "azo                 -         -         -         -  ...         -         -   \n",
       "max                 V         F         G         S  ...         A         H   \n",
       "azo-match       False     False     False     False  ...     False     False   \n",
       "\n",
       "                 1839      1840      1841      1842      1843      1844  \\\n",
       "-            4.332668  4.444419  4.444419  4.444419  4.444419  4.444419   \n",
       "A                 0.0       0.0       0.0  0.004189       0.0       0.0   \n",
       "C            0.004119       0.0       0.0       0.0       0.0       0.0   \n",
       "D            0.004119       0.0       0.0       0.0       0.0       0.0   \n",
       "E                 0.0  0.004189       0.0       0.0       0.0       0.0   \n",
       "F                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "G                 0.0       0.0       0.0       0.0  0.004189       0.0   \n",
       "H                 0.0       0.0       0.0       0.0       0.0  0.004189   \n",
       "I                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "K                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "L                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "M                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "N            0.032948       0.0       0.0       0.0       0.0       0.0   \n",
       "P                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "Q                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "R                 0.0       0.0  0.004189       0.0       0.0       0.0   \n",
       "S                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "T                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "V                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "W                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "X                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "Y                 0.0       0.0       0.0       0.0       0.0       0.0   \n",
       "information  4.373853  4.448608  4.448608  4.448608  4.448608  4.448608   \n",
       "azo                 -         -         -         -         -         -   \n",
       "max                 N         E         R         A         G         H   \n",
       "azo-match       False     False     False     False     False     False   \n",
       "\n",
       "                 1845      1846  \n",
       "-            4.444419  4.444419  \n",
       "A                 0.0       0.0  \n",
       "C                 0.0       0.0  \n",
       "D            0.004189       0.0  \n",
       "E                 0.0       0.0  \n",
       "F                 0.0       0.0  \n",
       "G                 0.0       0.0  \n",
       "H                 0.0       0.0  \n",
       "I                 0.0       0.0  \n",
       "K                 0.0       0.0  \n",
       "L                 0.0       0.0  \n",
       "M                 0.0       0.0  \n",
       "N                 0.0       0.0  \n",
       "P                 0.0       0.0  \n",
       "Q                 0.0       0.0  \n",
       "R                 0.0       0.0  \n",
       "S                 0.0       0.0  \n",
       "T                 0.0       0.0  \n",
       "V                 0.0  0.004189  \n",
       "W                 0.0       0.0  \n",
       "X                 0.0       0.0  \n",
       "Y                 0.0       0.0  \n",
       "information  4.448608  4.448608  \n",
       "azo                 -         -  \n",
       "max                 D         V  \n",
       "azo-match       False     False  \n",
       "\n",
       "[26 rows x 1847 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 393,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>172</th>\n",
       "      <th>174</th>\n",
       "      <th>308</th>\n",
       "      <th>475</th>\n",
       "      <th>485</th>\n",
       "      <th>488</th>\n",
       "      <th>495</th>\n",
       "      <th>502</th>\n",
       "      <th>527</th>\n",
       "      <th>569</th>\n",
       "      <th>...</th>\n",
       "      <th>1083</th>\n",
       "      <th>1126</th>\n",
       "      <th>1205</th>\n",
       "      <th>1220</th>\n",
       "      <th>1244</th>\n",
       "      <th>1271</th>\n",
       "      <th>1282</th>\n",
       "      <th>1286</th>\n",
       "      <th>1306</th>\n",
       "      <th>1314</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Conservation</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>9</td>\n",
       "      <td>9</td>\n",
       "      <td>6</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>9</td>\n",
       "      <td>10</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>9</td>\n",
       "      <td>7</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Quality</th>\n",
       "      <td>1135.2317</td>\n",
       "      <td>1278.8102</td>\n",
       "      <td>1664.9602</td>\n",
       "      <td>1139.5281</td>\n",
       "      <td>1677.4182</td>\n",
       "      <td>2033.7096</td>\n",
       "      <td>2126.2568</td>\n",
       "      <td>1651.0927</td>\n",
       "      <td>1581.4888</td>\n",
       "      <td>1248.7866</td>\n",
       "      <td>...</td>\n",
       "      <td>1086.1302</td>\n",
       "      <td>1352.8774</td>\n",
       "      <td>1797.6267</td>\n",
       "      <td>1080.2826</td>\n",
       "      <td>878.3229</td>\n",
       "      <td>205.22469</td>\n",
       "      <td>2041.5193</td>\n",
       "      <td>1996.0206</td>\n",
       "      <td>1603.496</td>\n",
       "      <td>1140.915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Consensus#</th>\n",
       "      <td>61.393597</td>\n",
       "      <td>54.613937</td>\n",
       "      <td>65.53672</td>\n",
       "      <td>50.282486</td>\n",
       "      <td>54.613937</td>\n",
       "      <td>58.568737</td>\n",
       "      <td>56.026367</td>\n",
       "      <td>90.20715</td>\n",
       "      <td>71.56309</td>\n",
       "      <td>56.40301</td>\n",
       "      <td>...</td>\n",
       "      <td>50.47081</td>\n",
       "      <td>67.3258</td>\n",
       "      <td>79.472694</td>\n",
       "      <td>56.967983</td>\n",
       "      <td>55.461395</td>\n",
       "      <td>53.483994</td>\n",
       "      <td>59.60452</td>\n",
       "      <td>53.107346</td>\n",
       "      <td>63.276836</td>\n",
       "      <td>52.259888</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Consensus</th>\n",
       "      <td>K 61.39%; N 8.66%; D 8.19%; E 2.64%; R 1.69%;...</td>\n",
       "      <td>A 54.61%; K 11.49%; S 8.19%; E 4.90%; R 3.30%...</td>\n",
       "      <td>T 65.54%; S 18.46%; A 3.48%; V 3.48%; R 0.75%...</td>\n",
       "      <td>C 50.28%; A 31.54%; S 12.71%; M 0.47%; V 0.47...</td>\n",
       "      <td>R 54.61%; K 28.06%; S 4.14%; A 3.11%; G 2.35%...</td>\n",
       "      <td>V 58.57%; I 27.50%; L 7.91%; A 2.73%; F 0.47%...</td>\n",
       "      <td>I 56.03%; V 39.64%; A 1.41%; R 0.47%; S 0.47%...</td>\n",
       "      <td>W 90.21%; R 4.24%; S 0.94%; H 0.66%; Y 0.56%;...</td>\n",
       "      <td>T 71.56%; G 8.47%; S 8.10%; A 3.20%; Y 2.35%;...</td>\n",
       "      <td>N 56.40%; T 13.28%; S 12.62%; Q 7.82%; V 1.22...</td>\n",
       "      <td>...</td>\n",
       "      <td>A 50.47%; K 1.51%; S 1.51%; E 0.94%; R 0.85%;...</td>\n",
       "      <td>G 67.33%; A 14.60%; I 4.90%; D 3.77%; S 1.51%...</td>\n",
       "      <td>Y 79.47%; F 14.78%; P 1.98%; V 0.38%; L 0.28%...</td>\n",
       "      <td>Y 56.97%; W 35.40%; H 2.26%; Q 1.41%; D 0.94%...</td>\n",
       "      <td>A 55.46%; H 33.99%; I 1.69%; V 1.51%; T 1.32%...</td>\n",
       "      <td>N 53.48%; I 21.19%; V 13.84%; T 2.17%; R 1.69...</td>\n",
       "      <td>I 59.60%; M 19.96%; V 12.05%; L 6.97%; W 0.19...</td>\n",
       "      <td>V 53.11%; I 22.13%; L 17.23%; M 1.79%; F 1.69...</td>\n",
       "      <td>S 63.28%; N 10.08%; A 9.32%; T 8.38%; C 1.41%...</td>\n",
       "      <td>K 52.26%; R 11.02%; T 7.34%; E 3.95%; Q 3.30%...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Occupancy</th>\n",
       "      <td>938</td>\n",
       "      <td>1018</td>\n",
       "      <td>984</td>\n",
       "      <td>1027</td>\n",
       "      <td>1040</td>\n",
       "      <td>1048</td>\n",
       "      <td>1054</td>\n",
       "      <td>1050</td>\n",
       "      <td>1042</td>\n",
       "      <td>1033</td>\n",
       "      <td>...</td>\n",
       "      <td>610</td>\n",
       "      <td>1040</td>\n",
       "      <td>1045</td>\n",
       "      <td>1054</td>\n",
       "      <td>1050</td>\n",
       "      <td>1045</td>\n",
       "      <td>1052</td>\n",
       "      <td>1053</td>\n",
       "      <td>1034</td>\n",
       "      <td>962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>K</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>R</td>\n",
       "      <td>V</td>\n",
       "      <td>I</td>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>A</td>\n",
       "      <td>N</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>S</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>azo</th>\n",
       "      <td>D</td>\n",
       "      <td>N</td>\n",
       "      <td>S</td>\n",
       "      <td>S</td>\n",
       "      <td>K</td>\n",
       "      <td>I</td>\n",
       "      <td>V</td>\n",
       "      <td>R</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>...</td>\n",
       "      <td>-</td>\n",
       "      <td>S</td>\n",
       "      <td>F</td>\n",
       "      <td>W</td>\n",
       "      <td>H</td>\n",
       "      <td>I</td>\n",
       "      <td>M</td>\n",
       "      <td>L</td>\n",
       "      <td>N</td>\n",
       "      <td>Q</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>azo_match</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8 rows Ã— 59 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                           172   \\\n",
       "0                                                                 \n",
       "Conservation                                                  1   \n",
       "Quality                                               1135.2317   \n",
       "Consensus#                                            61.393597   \n",
       "Consensus      K 61.39%; N 8.66%; D 8.19%; E 2.64%; R 1.69%;...   \n",
       "Occupancy                                                   938   \n",
       "max                                                           K   \n",
       "azo                                                           D   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           174   \\\n",
       "0                                                                 \n",
       "Conservation                                                  0   \n",
       "Quality                                               1278.8102   \n",
       "Consensus#                                            54.613937   \n",
       "Consensus      A 54.61%; K 11.49%; S 8.19%; E 4.90%; R 3.30%...   \n",
       "Occupancy                                                  1018   \n",
       "max                                                           A   \n",
       "azo                                                           N   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           308   \\\n",
       "0                                                                 \n",
       "Conservation                                                  1   \n",
       "Quality                                               1664.9602   \n",
       "Consensus#                                             65.53672   \n",
       "Consensus      T 65.54%; S 18.46%; A 3.48%; V 3.48%; R 0.75%...   \n",
       "Occupancy                                                   984   \n",
       "max                                                           T   \n",
       "azo                                                           S   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           475   \\\n",
       "0                                                                 \n",
       "Conservation                                                  1   \n",
       "Quality                                               1139.5281   \n",
       "Consensus#                                            50.282486   \n",
       "Consensus      C 50.28%; A 31.54%; S 12.71%; M 0.47%; V 0.47...   \n",
       "Occupancy                                                  1027   \n",
       "max                                                           C   \n",
       "azo                                                           S   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           485   \\\n",
       "0                                                                 \n",
       "Conservation                                                  4   \n",
       "Quality                                               1677.4182   \n",
       "Consensus#                                            54.613937   \n",
       "Consensus      R 54.61%; K 28.06%; S 4.14%; A 3.11%; G 2.35%...   \n",
       "Occupancy                                                  1040   \n",
       "max                                                           R   \n",
       "azo                                                           K   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           488   \\\n",
       "0                                                                 \n",
       "Conservation                                                  9   \n",
       "Quality                                               2033.7096   \n",
       "Consensus#                                            58.568737   \n",
       "Consensus      V 58.57%; I 27.50%; L 7.91%; A 2.73%; F 0.47%...   \n",
       "Occupancy                                                  1048   \n",
       "max                                                           V   \n",
       "azo                                                           I   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           495   \\\n",
       "0                                                                 \n",
       "Conservation                                                  9   \n",
       "Quality                                               2126.2568   \n",
       "Consensus#                                            56.026367   \n",
       "Consensus      I 56.03%; V 39.64%; A 1.41%; R 0.47%; S 0.47%...   \n",
       "Occupancy                                                  1054   \n",
       "max                                                           I   \n",
       "azo                                                           V   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           502   \\\n",
       "0                                                                 \n",
       "Conservation                                                  6   \n",
       "Quality                                               1651.0927   \n",
       "Consensus#                                             90.20715   \n",
       "Consensus      W 90.21%; R 4.24%; S 0.94%; H 0.66%; Y 0.56%;...   \n",
       "Occupancy                                                  1050   \n",
       "max                                                           W   \n",
       "azo                                                           R   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           527   \\\n",
       "0                                                                 \n",
       "Conservation                                                  7   \n",
       "Quality                                               1581.4888   \n",
       "Consensus#                                             71.56309   \n",
       "Consensus      T 71.56%; G 8.47%; S 8.10%; A 3.20%; Y 2.35%;...   \n",
       "Occupancy                                                  1042   \n",
       "max                                                           T   \n",
       "azo                                                           G   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           569   ...  \\\n",
       "0                                                                ...   \n",
       "Conservation                                                  7  ...   \n",
       "Quality                                               1248.7866  ...   \n",
       "Consensus#                                             56.40301  ...   \n",
       "Consensus      N 56.40%; T 13.28%; S 12.62%; Q 7.82%; V 1.22...  ...   \n",
       "Occupancy                                                  1033  ...   \n",
       "max                                                           N  ...   \n",
       "azo                                                           A  ...   \n",
       "azo_match                                                  True  ...   \n",
       "\n",
       "                                                           1083  \\\n",
       "0                                                                 \n",
       "Conservation                                                  0   \n",
       "Quality                                               1086.1302   \n",
       "Consensus#                                             50.47081   \n",
       "Consensus      A 50.47%; K 1.51%; S 1.51%; E 0.94%; R 0.85%;...   \n",
       "Occupancy                                                   610   \n",
       "max                                                           A   \n",
       "azo                                                           -   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1126  \\\n",
       "0                                                                 \n",
       "Conservation                                                  3   \n",
       "Quality                                               1352.8774   \n",
       "Consensus#                                              67.3258   \n",
       "Consensus      G 67.33%; A 14.60%; I 4.90%; D 3.77%; S 1.51%...   \n",
       "Occupancy                                                  1040   \n",
       "max                                                           G   \n",
       "azo                                                           S   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1205  \\\n",
       "0                                                                 \n",
       "Conservation                                                  9   \n",
       "Quality                                               1797.6267   \n",
       "Consensus#                                            79.472694   \n",
       "Consensus      Y 79.47%; F 14.78%; P 1.98%; V 0.38%; L 0.28%...   \n",
       "Occupancy                                                  1045   \n",
       "max                                                           Y   \n",
       "azo                                                           F   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1220  \\\n",
       "0                                                                 \n",
       "Conservation                                                 10   \n",
       "Quality                                               1080.2826   \n",
       "Consensus#                                            56.967983   \n",
       "Consensus      Y 56.97%; W 35.40%; H 2.26%; Q 1.41%; D 0.94%...   \n",
       "Occupancy                                                  1054   \n",
       "max                                                           Y   \n",
       "azo                                                           W   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1244  \\\n",
       "0                                                                 \n",
       "Conservation                                                  4   \n",
       "Quality                                                878.3229   \n",
       "Consensus#                                            55.461395   \n",
       "Consensus      A 55.46%; H 33.99%; I 1.69%; V 1.51%; T 1.32%...   \n",
       "Occupancy                                                  1050   \n",
       "max                                                           A   \n",
       "azo                                                           H   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1271  \\\n",
       "0                                                                 \n",
       "Conservation                                                  6   \n",
       "Quality                                               205.22469   \n",
       "Consensus#                                            53.483994   \n",
       "Consensus      N 53.48%; I 21.19%; V 13.84%; T 2.17%; R 1.69...   \n",
       "Occupancy                                                  1045   \n",
       "max                                                           N   \n",
       "azo                                                           I   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1282  \\\n",
       "0                                                                 \n",
       "Conservation                                                  8   \n",
       "Quality                                               2041.5193   \n",
       "Consensus#                                             59.60452   \n",
       "Consensus      I 59.60%; M 19.96%; V 12.05%; L 6.97%; W 0.19...   \n",
       "Occupancy                                                  1052   \n",
       "max                                                           I   \n",
       "azo                                                           M   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1286  \\\n",
       "0                                                                 \n",
       "Conservation                                                  9   \n",
       "Quality                                               1996.0206   \n",
       "Consensus#                                            53.107346   \n",
       "Consensus      V 53.11%; I 22.13%; L 17.23%; M 1.79%; F 1.69...   \n",
       "Occupancy                                                  1053   \n",
       "max                                                           V   \n",
       "azo                                                           L   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1306  \\\n",
       "0                                                                 \n",
       "Conservation                                                  7   \n",
       "Quality                                                1603.496   \n",
       "Consensus#                                            63.276836   \n",
       "Consensus      S 63.28%; N 10.08%; A 9.32%; T 8.38%; C 1.41%...   \n",
       "Occupancy                                                  1034   \n",
       "max                                                           S   \n",
       "azo                                                           N   \n",
       "azo_match                                                  True   \n",
       "\n",
       "                                                           1314  \n",
       "0                                                                \n",
       "Conservation                                                  1  \n",
       "Quality                                                1140.915  \n",
       "Consensus#                                            52.259888  \n",
       "Consensus      K 52.26%; R 11.02%; T 7.34%; E 3.95%; Q 3.30%...  \n",
       "Occupancy                                                   962  \n",
       "max                                                           K  \n",
       "azo                                                           Q  \n",
       "azo_match                                                  True  \n",
       "\n",
       "[8 rows x 59 columns]"
      ]
     },
     "execution_count": 393,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = f\"thermo_nifD_features.csv\"\n",
    "\n",
    "features = pd.read_csv(file, header=None, index_col = 0)\n",
    "features.columns = range(len(features.columns))\n",
    "features = features.drop(features.columns[-1], axis=1)\n",
    "features.loc['max'] = features.loc['Consensus']\n",
    "features.loc['Quality'] = features.apply(lambda x: float(x.loc['Quality']), axis=0)\n",
    "features.loc['Consensus#'] = features.apply(lambda x: float(x.loc['Consensus#']), axis=0)\n",
    "features.loc['max'] = features.apply(lambda x: x.loc['Consensus'].strip()[0], axis=0)\n",
    "features.loc['azo'] = seqs[0]\n",
    "features.loc['azo_match'] = score.apply(lambda x: x.loc['max'] != x.loc['azo'])\n",
    "features = features.loc[:, features.loc['azo_match']]\n",
    "\n",
    "features = features.loc[:, features.loc['Consensus#'] >= 50]\n",
    "# features = features.loc[:, features.loc['azo'] = 'L']\n",
    "# features = features.loc[:, features.loc['azo'] == 'I']\n",
    "# # features = features.loc[:, features.loc['azo'] != 'V']\n",
    "features#.to_csv('thermo_nifD_features_filt.csv', index = False, sep=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 436,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>502</th>\n",
       "      <th>527</th>\n",
       "      <th>569</th>\n",
       "      <th>691</th>\n",
       "      <th>836</th>\n",
       "      <th>877</th>\n",
       "      <th>997</th>\n",
       "      <th>1314</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RS_GCF_001687335.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>Y</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_002898515.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_002093115.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_011207455.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>Y</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_027925145.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>L</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_903896745.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_000195755.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_000266945.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_030065275.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>L</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_009688965.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>I</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_947259715.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_000711225.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_002423105.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>Q</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_025355965.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>K</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>C</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_030154185.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>V</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_015231595.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>L</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_003232385.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_023512345.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>N</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RS_GCF_030913705.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_003249225.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>S</td>\n",
       "      <td>G</td>\n",
       "      <td>V</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GB_GCA_003245715.1</th>\n",
       "      <td>W</td>\n",
       "      <td>T</td>\n",
       "      <td>Q</td>\n",
       "      <td>G</td>\n",
       "      <td>L</td>\n",
       "      <td>A</td>\n",
       "      <td>V</td>\n",
       "      <td>K</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   502  527  569  691  836  877  997  1314\n",
       "RS_GCF_001687335.1    W    T    Y    G    V    A    V    K\n",
       "RS_GCF_002898515.1    W    T    S    G    V    C    V    K\n",
       "RS_GCF_002093115.1    W    T    S    G    I    C    V    K\n",
       "RS_GCF_011207455.1    W    T    Y    G    V    A    V    R\n",
       "RS_GCF_027925145.1    W    T    N    G    L    A    V    K\n",
       "GB_GCA_903896745.1    W    T    N    G    I    A    V    K\n",
       "RS_GCF_000195755.1    W    T    N    G    V    A    V    R\n",
       "RS_GCF_000266945.1    W    T    N    G    V    A    V    K\n",
       "GB_GCA_030065275.1    W    T    N    G    L    A    V    R\n",
       "RS_GCF_009688965.1    W    T    N    G    I    A    V    R\n",
       "GB_GCA_947259715.1    W    T    -    G    V    C    V    K\n",
       "RS_GCF_000711225.1    W    T    C    G    V    C    V    K\n",
       "GB_GCA_002423105.1    W    T    Q    G    V    C    V    K\n",
       "GB_GCA_025355965.1    W    T    K    G    V    C    V    K\n",
       "GB_GCA_030154185.1    W    T    V    G    V    A    V    K\n",
       "GB_GCA_015231595.1    W    T    N    G    L    G    V    K\n",
       "GB_GCA_003232385.1    W    T    S    G    V    A    V    K\n",
       "GB_GCA_023512345.1    W    T    N    G    V    V    V    K\n",
       "RS_GCF_030913705.1    W    T    C    G    V    A    V    K\n",
       "GB_GCA_003249225.1    W    T    S    G    V    A    V    R\n",
       "GB_GCA_003245715.1    W    T    Q    G    L    A    V    K"
      ]
     },
     "execution_count": 436,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# thermo to plot MSA\n",
    "\n",
    "file = f\"../Thermo/thermo_nifD.aln\"\n",
    "\n",
    "# store in dataframe        \n",
    "results = list(SeqIO.parse(file, \"fasta\"))\n",
    "acc = [result.id.split('|')[0] for result in results]\n",
    "seqs = [list(str(result.seq)) for result in results]\n",
    "des = [[result.description.split('|')[-1]] for result in results]\n",
    "\n",
    "pssm = pd.DataFrame(seqs, index = acc)\n",
    "pssm = pssm[pssm.loc[:,502] !='R']\n",
    "pssm = pssm[pssm.loc[:,527] =='T']\n",
    "pssm = pssm[(pssm.loc[:,569] !='A')]# | (pssm.loc[:,569] =='T')]\n",
    "pssm = pssm[pssm.loc[:,691] =='G']\n",
    "pssm = pssm[(pssm.loc[:,836] =='V') | (pssm.loc[:,836] =='L')| (pssm.loc[:,836] =='I')]\n",
    "pssm = pssm[(pssm.loc[:,877] !='S') & (pssm.loc[:,877] !='T')]\n",
    "pssm = pssm[pssm.loc[:,997] !='I']\n",
    "pssm = pssm[(pssm.loc[:,1314] =='K') | (pssm.loc[:,1314] =='R')]\n",
    "pssm = pssm[[502, 527, 569, 691, 836, 877, 997, 1314]]\n",
    "pssm = pssm.drop_duplicates()\n",
    "pssm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 447,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "21 16 39 36 33\n"
     ]
    }
   ],
   "source": [
    "file = f\"../Thermo/thermo_nifK.aln\"\n",
    "\n",
    "# store in dataframe        \n",
    "results = list(SeqIO.parse(file, \"fasta\"))\n",
    "seqs = [list(str(result.seq)) for result in results]\n",
    "protein = seqs[0]\n",
    "R = protein.count('R')\n",
    "H = protein.count('H')\n",
    "K = protein.count('K')\n",
    "D = protein.count('D')\n",
    "E = protein.count('E')\n",
    "\n",
    "print(R,H,K,D,E)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "5592\n",
      "5592\n"
     ]
    }
   ],
   "source": [
    "file = '../operon-org/structure.pickle'\n",
    "import pickle\n",
    "\n",
    "file = pickle.load(open(file, 'rb'))\n",
    "print(len(file))\n",
    "\n",
    "remove_list = ['GB_GCA_002898135.1_BDTQ01000234.1', \n",
    "             'GB_GCA_021797365.1_JAJZZU010000300.1',\n",
    "             'GB_GCA_009649835.1_PYCK01000054.1',\n",
    "             'GB_GCA_014380135.1_JACMKN010000012.1',\n",
    "             'GB_GCA_017889725.1_JAGGAE010000092.1',\n",
    "             'GB_GCA_902385025.1_CABMDX010000290.1',\n",
    "             'GB_GCA_027004315.1_JALTHV010000232.1',\n",
    "             'GB_GCA_031288575.1_JAITRD010000305.1',\n",
    "             'GB_GCA_937141995.1_CAKZQX010000080.1',\n",
    "             'GB_GCA_029778665.1_JAKOPX010000087.1',\n",
    "             'GB_GCA_031276475.1_JAISMZ010000161.1',\n",
    "             'RS_GCF_015731765.1_NZ_VRYY01000094.1',\n",
    "             'GB_GCA_013041765.1_JABDRL010000192.1',\n",
    "             'GB_GCA_021798485.1_JAJZGK010000148.1',\n",
    "             'GB_GCA_031290695.1_JAITNB010000026.1',\n",
    "             'GB_GCA_031278575.1_JAITKD010000016.1',\n",
    "             'GB_GCA_937886325.1_CALABV010000158.1',\n",
    "             'GB_GCA_937915285.1_CALDNG010000494.1',\n",
    "             'GB_GCA_021802595.1_JAJZRT010000330.1']\n",
    "\n",
    "for item in remove_list:\n",
    "    to_remove = '../operon-org/input-fastas/' + item + '_operon.fasta'\n",
    "    try:\n",
    "        file.pop(to_remove)\n",
    "    except:\n",
    "        pass\n",
    "\n",
    "pickle.dump(file, open('../operon-org/structure-update.pickle', 'wb'))\n",
    "print(len(file))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "parse_hmm",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
