{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import requests\n",
    "from multiprocessing import Pool\n",
    "import wget"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get KoFam sequences\n",
    "### nifH (K02588 in KEGG), nifD (K02586), and nifK (K02591)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "ko_dict = {'nifH':'K02588', 'nifD':'K02586', 'nifK':'K02591'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a file fasta file for each konumber to record fasta  sequences\n",
    "for ko in ko_dict:\n",
    "    with open('ko_{}_known.fasta'.format(ko), 'w') as f:\n",
    "        f.write('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pull \"known\" nifHDK seq from Dixon et al\n",
    "known_nif = pd.read_csv('Nitrogenase_Seqs.csv')\n",
    "\n",
    "known_nifH = known_nif.NifH.dropna().tolist()\n",
    "known_nifD = known_nif.NifD.dropna().tolist()\n",
    "known_nifK = known_nif.NifK.dropna().tolist()\n",
    "\n",
    "known_list = [known_nifH, known_nifD, known_nifK]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['chd:Calhy_2145', 'ckn:Calkro_2123', 'mfs:MFS40622_0031']\n",
      "['chd:Calhy_2140']\n",
      "['chd:Calhy_2139', 'ckn:Calkro_2119', 'mfs:MFS40622_0035']\n"
     ]
    }
   ],
   "source": [
    "for ko, known in zip(ko_dict, known_list):\n",
    "    # Get ko number\n",
    "    ko_number = ko_dict[ko]\n",
    "\n",
    "    # URL to fetch the list of genes associated with the KO number\n",
    "    url = f\"http://rest.kegg.jp/link/genes/{ko_number}\"\n",
    "\n",
    "    # Fetch the list of genes\n",
    "    response = requests.get(url)\n",
    "    genes = response.text.strip().split(\"\\n\")\n",
    "    genes = [gene.split(\"\\t\")[1] for gene in genes] # only keep gene name (KEGG locus tag)\n",
    "\n",
    "    # Only keep known genes (Dixon et al)\n",
    "    genes = [gene for gene in genes if gene.split(\":\")[1] in known]\n",
    "\n",
    "    # Saves genes in .fasta file\n",
    "    for gene in genes:\n",
    "\n",
    "        # fetch the sequence\n",
    "        sequence_url = f\"http://rest.kegg.jp/get/{gene}/aaseq\"\n",
    "        sequence_response = requests.get(sequence_url)\n",
    "\n",
    "        # Write the sequence to the file\n",
    "        with open('ko_{}_known.fasta'.format(ko), 'a') as f:\n",
    "            f.write(sequence_response.text)\n",
    "\n",
    "# takes about 2 min (or 20 min without filtering out unknown genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\n\\n- get the list of genomes from the two families from Table 1\\n- get the list of proteins from the two families\\n- download the proteins from the two families\\n- search for nifH,D,K proteins in the two families\\n- add the proteins to the fasta files\\n- make hmm profile\\n\\n'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add Anaeromyxobacteraceae and Geobacteraceae proteins\n",
    "\n",
    "'''\n",
    "\n",
    "- get the list of genomes from the two families from Table 1\n",
    "- get the list of proteins from the two families\n",
    "- download the proteins from the two families\n",
    "- search for nifH,D,K proteins in the two families\n",
    "- add the proteins to the fasta files\n",
    "- make hmm profile\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_018919385.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_016458275.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_016458235.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_016458305.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_014193515.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_014193585.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_023221575.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_014193675.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_008802365.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_008802355.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_008369015.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_004117875.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_004917075.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_004917065.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_004791675.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_020179575.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_020179605.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_018986915.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_019063845.1/?search=nif\n",
      "https://www.ncbi.nlm.nih.gov/datasets/gene/GCF_019063885.1/?search=nif\n"
     ]
    }
   ],
   "source": [
    "genomes = ['GCF_018919385.1', 'GCF_016458275.1', 'GCF_016458235.1', 'GCF_016458305.1', 'GCF_014193515.1', 'GCF_014193585.1', \n",
    "'GCF_023221575.1', 'GCF_014193675.1', 'GCF_008802365.1', 'GCF_008802355.1', 'GCF_008369015.1', 'GCF_004117875.1', 'GCF_004917075.1',\n",
    "'GCF_004917065.1', 'GCF_004791675.1', 'GCF_020179575.1', 'GCF_020179605.1', 'GCF_018986915.1', 'GCF_019063845.1', 'GCF_019063885.1']\n",
    "\n",
    "# GenBank accession numbers for genomes produced in this study JAKLOY010000000, BLXW01000000 manually included\n",
    "\n",
    "# make url\n",
    "for genome in genomes:\n",
    "    print(f\"https://www.ncbi.nlm.nih.gov/datasets/gene/{genome}/?search=nif\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "plotting",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
