{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import requests\n",
    "from multiprocessing import Pool\n",
    "from Bio import Entrez\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Get CDD sequences\n",
    "### nifH (cd02040 in CDD), nifD (cd01976), and nifK (cd01974)\n",
    "### nifB (cd00562), nifE (cd01968), vnfE (cd01972), nifN (cd03466, cd01966), and vnfN (cd01971)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load alignment data (downloaded from NCBI CDD website)\n",
    "# get accession numbers\n",
    "\n",
    "# gene_list = ['nifH', 'nifD', 'nifK']\n",
    "gene_list = ['nifB', 'nifE', 'nifN']\n",
    "acc_list = []\n",
    "\n",
    "for gene in gene_list:\n",
    "    aln_file = f'CDD_{gene}_alignment.faa'\n",
    "    acc = []\n",
    "\n",
    "    with open(aln_file, 'r') as f:\n",
    "        aln = f.read()\n",
    "\n",
    "        for line in aln.split('\\n'):\n",
    "            if line.startswith('>'):\n",
    "                acc.append(line.split('|')[1])\n",
    "\n",
    "    acc_list.append(acc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a file fasta file for each gene to record sequences\n",
    "for gene in gene_list:\n",
    "    with open(f'CDD_{gene}_seq.faa', 'w') as f:\n",
    "        f.write('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the protein sequences from NCBI using accession numbers\n",
    "\n",
    "Entrez.email = 'zshivji@caltech.edu'\n",
    "\n",
    "for i, gene in enumerate(gene_list):\n",
    "    for acc_no in acc_list[i]:\n",
    "        # Fetch the record from NCBI\n",
    "        handle = Entrez.efetch(db=\"protein\", id=acc_no, rettype=\"fasta\", retmode=\"text\")\n",
    "        seq_record = SeqIO.read(handle, \"fasta\")\n",
    "        handle.close()\n",
    "        \n",
    "        # Write the sequence to the output FASTA file\n",
    "        with open(f'CDD_{gene}_seq.faa', 'a') as f:\n",
    "            SeqIO.write(seq_record, f, \"fasta\")\n",
    "\n",
    "# takes about 1.5 min"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pro-pred",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
